<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Basics • rvat</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Basics">
<meta property="og:description" content="rvat">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rvat</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/basics.html">Basics</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/basics.html">Basics</a>
    </li>
    <li>
      <a href="../articles/association_testing.html">Association testing</a>
    </li>
    <li>
      <a href="../articles/explore_visualize_results.html">Visualization</a>
    </li>
    <li>
      <a href="../articles/cmdline.html">Command-line tool</a>
    </li>
  </ul>
</li>
<li>
  <a href="../articles/faq.html">F.A.Q.</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/kkenna/rvat" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Basics</h1>
            
      
      
      <div class="hidden name"><code>basics.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Libraries</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">rvat</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">rvatData</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SummarizedExperiment</span><span class="op">)</span></code></pre></div>
<div class="section level2">
<h2 id="example-data">Example data<a class="anchor" aria-label="anchor" href="#example-data"></a>
</h2>
<p>In these tutorials, we’ll use simulated example data that we included in the <code>rvatData</code> package.</p>
<p>Included in-memory datasets:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span>package <span class="op">=</span> <span class="st">"rvatData"</span><span class="op">)</span><span class="op">$</span><span class="va">results</span><span class="op">[</span>,<span class="st">"Item"</span><span class="op">]</span></code></pre></div>
<pre><code><span class="co">## [1] "GT"         "GTsmall"    "rvbresults"</span></code></pre>
<p>Included on-disk datasets, note that the paths to the on-disk datasets can be retrieved as below:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/rvatData/man/rvat_example.html" class="external-link">rvat_example</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] "rvatData.gdb"     "rvatData.pheno"   "rvatData.varinfo" "rvatData.vcf.gz"</span></code></pre>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/rvatData/man/rvat_example.html" class="external-link">rvat_example</a></span><span class="op">(</span><span class="st">"rvatData.gdb"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] "/Users/phop2/Library/R/3.6/library/rvatData/extdata/rvatData.gdb"</span></code></pre>
<p>Some output will be written during the tutorials, use a temporary directory for that:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">outdir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="setting-up-and-populating-a-gdb">Setting up and populating a gdb<a class="anchor" aria-label="anchor" href="#setting-up-and-populating-a-gdb"></a>
</h2>
<div class="section level3">
<h3 id="creating-a-gdb-based-on-a-vcf-file">Creating a gdb based on a VCF-file<a class="anchor" aria-label="anchor" href="#creating-a-gdb-based-on-a-vcf-file"></a>
</h3>
<p>The first step consists of setting up a so-called <code><a href="../reference/gdb.html">gdb()</a></code>, an SQLite-based data structure which forms the heart of all RVAT operation. Gdb files allow for much more rapid and memory efficient loading of sample genotype data into numeric matrices within R. They also allow for the upload and integration of additional variant and sample annotation data to allow for complex data</p>
<p>Here, we show how to create a <code><a href="../reference/gdb.html">gdb()</a></code> object from a VCF-file. Note that RVAT will assign a unique <code>VAR_id</code> to each variant, which allow rapid querying of variants, these IDs have no inherent meaning (i.e. they do not refer to chromosomal position or the like).</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">vcfpath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rvatData/man/rvat_example.html" class="external-link">rvat_example</a></span><span class="op">(</span><span class="st">"rvatData.vcf.gz"</span><span class="op">)</span>
<span class="fu"><a href="../reference/buildGdb.html">buildGdb</a></span><span class="op">(</span>vcf <span class="op">=</span> <span class="va">vcfpath</span>,
         output <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outdir</span>,<span class="st">"/rvat_tutorials.gdb"</span><span class="op">)</span>, 
         overWrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## 2023-05-25 15:52:45  Creating gdb tables</span></code></pre>
<pre><code><span class="co">## 25000 sample IDs detected</span></code></pre>
<pre><code><span class="co">## 2023-05-25 15:52:45  Parsing vcf records</span></code></pre>
<pre><code><span class="co">## 2023-05-25 15:53:05  Processing completed for 1000 records. Committing to db.</span></code></pre>
<pre><code><span class="co">## 2023-05-25 15:53:18  Processing completed for 1763 records. Committing to db.</span></code></pre>
<pre><code><span class="co">## 2023-05-25 15:53:18  Creating var table indexes</span></code></pre>
<pre><code><span class="co">## 2023-05-25 15:53:18  Creating SM table index</span></code></pre>
<pre><code><span class="co">## 2023-05-25 15:53:18  Creating dosage table index</span></code></pre>
<pre><code><span class="co">## 2023-05-25 15:53:18  Creating ranged var table</span></code></pre>
<pre><code><span class="co">## 2023-05-25 15:53:18  Complete</span></code></pre>
<pre><code><span class="co">## rvat gdb object</span>
<span class="co">##  Path: /var/folders/cl/wvc0rvjx4vd5rzt2_fhpmfth0000gp/T//RtmpX5ASAH/rvat_tutorials.gdb</span></code></pre>
<p>Connect to the gdb you just created using the <code><a href="../reference/gdb.html">gdb()</a></code> function:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gdb</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gdb.html">gdb</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outdir</span>,<span class="st">"/rvat_tutorials.gdb"</span><span class="op">)</span><span class="op">)</span>
<span class="va">gdb</span></code></pre></div>
<pre><code><span class="co">## rvat gdb object</span>
<span class="co">##  Path: /private/var/folders/cl/wvc0rvjx4vd5rzt2_fhpmfth0000gp/T/RtmpX5ASAH/rvat_tutorials.gdb</span></code></pre>
</div>
<div class="section level3">
<h3 id="import-variant-info">Import variant info<a class="anchor" aria-label="anchor" href="#import-variant-info"></a>
</h3>
<p>add variant info to the gdb using the <code><a href="../reference/uploadAnno.html">uploadAnno()</a></code> method, variants are mapped to the gdb <code>VAR_id</code> based on the CHROM,POS,REF,ALT fields.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># example file including info such as chromosomal location, annotated gene and variant impact.</span>
<span class="va">varinfopath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rvatData/man/rvat_example.html" class="external-link">rvat_example</a></span><span class="op">(</span><span class="st">"rvatData.varinfo"</span><span class="op">)</span>
<span class="va">varinfo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="va">varinfopath</span>, header <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>

<span class="co"># the `name` parameter specifies the name of the table in the gdb, </span>
<span class="co"># the variant info that should be important is specified using the `value` parameter</span>
<span class="fu"><a href="../reference/uploadAnno.html">uploadAnno</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">gdb</span>, name <span class="op">=</span> <span class="st">"varInfo"</span>,value <span class="op">=</span> <span class="va">varinfo</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Loading table 'varInfo' from interactive R session'</span></code></pre>
<pre><code><span class="co">## 20 fields detected (CHROM,POS,ID,REF,ALT,QUAL,FILTER,AC,AN,AF,gene_name,HighImpact,ModerateImpact,Synonymous,aaref,aapos,aaalt,CADDphred,PolyPhen,SIFT)</span></code></pre>
<p>The <code><a href="../reference/gdb.html">listAnno()</a></code> method lists all variant annotation tables in the gdb:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/gdb.html">listAnno</a></span><span class="op">(</span><span class="va">gdb</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##      name               value                     date</span>
<span class="co">## 1 varInfo interactive_session Thu May 25 15:53:19 2023</span></code></pre>
<p>The <code><a href="../reference/getAnno.html">getAnno()</a></code> method retrieves a given variant annotation table from the gdb:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">varinfo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getAnno.html">getAnno</a></span><span class="op">(</span><span class="va">gdb</span>, table <span class="op">=</span> <span class="st">"varinfo"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">head</a></span><span class="op">(</span><span class="va">varinfo</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##   VAR_id CHROM      POS          ID REF ALT     QUAL FILTER AC    AN      AF</span>
<span class="co">## 1      1  chr1 11013847 rs758344713   G   C   593.13     NA  1 49978 0.00002</span>
<span class="co">## 2      2  chr1 11013912        &lt;NA&gt;   A   G   722.13     NA  1 49996 0.00002</span>
<span class="co">## 3      3  chr1 11013928 rs755357622   C   T  2598.23     NA  1 49992 0.00002</span>
<span class="co">## 4      4  chr1 11013936        &lt;NA&gt;   A   C  4135.36     NA  1 49990 0.00002</span>
<span class="co">## 5      5  chr1 11013936        &lt;NA&gt;   A   G  4135.36     NA  1 49990 0.00002</span>
<span class="co">## 6      6  chr1 11016874  rs80356715   C   T 64910.10     NA 29 49998 0.00058</span>
<span class="co">##   gene_name HighImpact ModerateImpact Synonymous aaref aapos aaalt CADDphred</span>
<span class="co">## 1    TARDBP          0              0          1     .     .     .         .</span>
<span class="co">## 2    TARDBP          0              1          0     H    62     R      24.8</span>
<span class="co">## 3    TARDBP          0              0          1     .     .     .         .</span>
<span class="co">## 4    TARDBP          0              1          0     N    70     T      24.1</span>
<span class="co">## 5    TARDBP          0              1          0     N    70     S      22.6</span>
<span class="co">## 6    TARDBP          0              1          0     A    90     V      22.2</span>
<span class="co">##   PolyPhen SIFT</span>
<span class="co">## 1        .    .</span>
<span class="co">## 2        P    D</span>
<span class="co">## 3        .    .</span>
<span class="co">## 4        B    T</span>
<span class="co">## 5        B    T</span>
<span class="co">## 6        B    T</span></code></pre>
<p>a specific region can be selected as follows:</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">varinfo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getAnno.html">getAnno</a></span><span class="op">(</span><span class="va">gdb</span>, table <span class="op">=</span> <span class="st">"varinfo"</span>, ranges <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>CHROM<span class="op">=</span><span class="st">"chr1"</span>, start <span class="op">=</span> <span class="fl">11013847</span>, end <span class="op">=</span> <span class="fl">11016874</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">head</a></span><span class="op">(</span><span class="va">varinfo</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##   VAR_id CHROM      POS          ID REF ALT     QUAL FILTER AC    AN      AF</span>
<span class="co">## 1      1  chr1 11013847 rs758344713   G   C   593.13     NA  1 49978 0.00002</span>
<span class="co">## 2      2  chr1 11013912        &lt;NA&gt;   A   G   722.13     NA  1 49996 0.00002</span>
<span class="co">## 3      3  chr1 11013928 rs755357622   C   T  2598.23     NA  1 49992 0.00002</span>
<span class="co">## 4      4  chr1 11013936        &lt;NA&gt;   A   C  4135.36     NA  1 49990 0.00002</span>
<span class="co">## 5      5  chr1 11013936        &lt;NA&gt;   A   G  4135.36     NA  1 49990 0.00002</span>
<span class="co">## 6      6  chr1 11016874  rs80356715   C   T 64910.10     NA 29 49998 0.00058</span>
<span class="co">##   gene_name HighImpact ModerateImpact Synonymous aaref aapos aaalt CADDphred</span>
<span class="co">## 1    TARDBP          0              0          1     .     .     .         .</span>
<span class="co">## 2    TARDBP          0              1          0     H    62     R      24.8</span>
<span class="co">## 3    TARDBP          0              0          1     .     .     .         .</span>
<span class="co">## 4    TARDBP          0              1          0     N    70     T      24.1</span>
<span class="co">## 5    TARDBP          0              1          0     N    70     S      22.6</span>
<span class="co">## 6    TARDBP          0              1          0     A    90     V      22.2</span>
<span class="co">##   PolyPhen SIFT</span>
<span class="co">## 1        .    .</span>
<span class="co">## 2        P    D</span>
<span class="co">## 3        .    .</span>
<span class="co">## 4        B    T</span>
<span class="co">## 5        B    T</span>
<span class="co">## 6        B    T</span></code></pre>
</div>
<div class="section level3">
<h3 id="import-sample-info">Import sample info<a class="anchor" aria-label="anchor" href="#import-sample-info"></a>
</h3>
<p>add phenotype info to the gdb using the <code><a href="../reference/uploadCohort.html">uploadCohort()</a></code> method:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># First we'll read in the phenotype data</span>
<span class="va">phenopath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rvatData/man/rvat_example.html" class="external-link">rvat_example</a></span><span class="op">(</span><span class="st">"rvatData.pheno"</span><span class="op">)</span>
<span class="va">pheno</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span> <span class="va">phenopath</span>, header <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>

<span class="co"># Import into the gdb (the `IID` column is used)</span>
<span class="co"># the `name` parameter specifies the name of the table in the gdb, </span>
<span class="co"># the sample info that should be important is specified using the `value` parameter</span>
<span class="fu"><a href="../reference/uploadCohort.html">uploadCohort</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">gdb</span>, name <span class="op">=</span> <span class="st">"pheno"</span>, value <span class="op">=</span> <span class="va">pheno</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Loading cohort 'pheno' from interactive R session</span></code></pre>
<pre><code><span class="co">## 10 fields detected (IID,sex,pheno,pop,superPop,PC1,PC2,PC3,PC4,age)</span></code></pre>
<pre><code><span class="co">## 0 males, 0 females and 25000 unknown gender</span></code></pre>
<pre><code><span class="co">## Retained 25000 of 25000 uploaded samples that could be mapped to dosage matrix</span></code></pre>
<pre><code><span class="co">## [1] 1</span></code></pre>
<p>The <code><a href="../reference/gdb.html">listCohort()</a></code> method lists all uploaded sample info tables in the gdb:</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/gdb.html">listCohort</a></span><span class="op">(</span><span class="va">gdb</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##    name               value                     date</span>
<span class="co">## 1 pheno interactive_session Thu May 25 15:53:19 2023</span></code></pre>
<p>The <code><a href="../reference/getCohort.html">getCohort()</a></code> method retrieves a given sample info table from the gdb:</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pheno</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getCohort.html">getCohort</a></span><span class="op">(</span><span class="va">gdb</span>, cohort <span class="op">=</span> <span class="st">"pheno"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">head</a></span><span class="op">(</span><span class="va">pheno</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##    IID sex pheno pop superPop       PC1       PC2        PC3         PC4</span>
<span class="co">## 1 ALS1   0     1 PJL      SAS  22.58749  4.738025 38.9157564 -10.1667342</span>
<span class="co">## 2 ALS2   0     1 BEB      SAS  26.37370 -1.445903 31.9712989  -6.0222555</span>
<span class="co">## 3 ALS3   0     1 TSI      EUR  30.75496 48.866843 -5.4092404  14.9474404</span>
<span class="co">## 4 ALS4   0     1 GWD      AFR -87.48035 -6.193628 -0.9504131  -1.2029932</span>
<span class="co">## 5 ALS5   0     1 MSL      AFR -87.79310 -5.323925 -1.0353300  -0.7403234</span>
<span class="co">## 6 ALS6   0     1 YRI      AFR -86.64071 -3.418417 -1.3964977   1.0561607</span>
<span class="co">##        age</span>
<span class="co">## 1 66.01743</span>
<span class="co">## 2 66.84306</span>
<span class="co">## 3 54.35749</span>
<span class="co">## 4 65.09777</span>
<span class="co">## 5 58.94664</span>
<span class="co">## 6 63.13635</span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="genomatrix-objects">GenoMatrix objects<a class="anchor" aria-label="anchor" href="#genomatrix-objects"></a>
</h2>
<p>Central to <code>RVAT</code> is the <code><a href="../reference/genoMatrix.html">genoMatrix()</a></code> object. The <code><a href="../reference/genoMatrix.html">genoMatrix()</a></code> object is an extension of the BioConductor <a href="https://bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html" class="external-link">summarizedExperiment</a> class. A <code><a href="../reference/genoMatrix.html">genoMatrix()</a></code> contains a matrix of genotypes, variant info and sample info queried from a <code><a href="../reference/gdb.html">gdb()</a></code> object, various methods are implemented for genoMatrix objects, which we’ll explore in this section.</p>
<div class="section level3">
<h3 id="construct-a-genomatrix">Construct a genoMatrix<a class="anchor" aria-label="anchor" href="#construct-a-genomatrix"></a>
</h3>
<p>First, we’ll build a genoMatrix object using the <code><a href="../reference/getGT.html">getGT()</a></code> method. Variants to include can be specified in two main ways: 1) by directly passing the VAR_ids to the <code><a href="../reference/getGT.html">getGT()</a></code> method, 2) by supplying a <code>varSet</code>. Here we’ll start with the first method, varSets are discussed in the <code><a href="../articles/association_testing.html">vignette("association_testing")</a></code></p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># get varinfo</span>
<span class="va">varinfo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getAnno.html">getAnno</a></span><span class="op">(</span><span class="va">gdb</span>, table <span class="op">=</span> <span class="st">"varinfo"</span>, where <span class="op">=</span> <span class="st">"gene_name = 'SOD1' and ModerateImpact = 1"</span><span class="op">)</span>

<span class="va">GT</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getGT.html">getGT</a></span><span class="op">(</span><span class="va">gdb</span>, 
            VAR_id <span class="op">=</span> <span class="va">varinfo</span><span class="op">$</span><span class="va">VAR_id</span>,
            cohort <span class="op">=</span> <span class="st">"pheno"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Retrieved genotypes for 32 variants</span></code></pre>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">GT</span></code></pre></div>
<pre><code><span class="co">## rvat genoMatrix Object</span>
<span class="co">## class: genoMatrix </span>
<span class="co">## dim: 32 25000 </span>
<span class="co">## metadata(8): ploidyLevels m ... geneticModel imputeMethod</span>
<span class="co">## assays(1): GT</span>
<span class="co">## rownames(32): 1405 1406 ... 1453 1454</span>
<span class="co">## rowData names(3): ploidy w AF</span>
<span class="co">## colnames(25000): ALS1 ALS2 ... CTRL19999 CTRL20000</span>
<span class="co">## colData names(10): IID sex ... PC4 age</span></code></pre>
</div>
<div class="section level3">
<h3 id="basic-operations">Basic operations<a class="anchor" aria-label="anchor" href="#basic-operations"></a>
</h3>
<p>Also see <a href="https://bioconductor.org/packages/release/bioc/vignettes/SummarizedExperiment/inst/doc/SummarizedExperiment.html" class="external-link">this</a> tutorial on SummarizedExperiments objects, the <code><a href="../reference/genoMatrix.html">genoMatrix()</a></code> class inherits from <code>summarizedExperiment</code>, meaning that all summarizedExperiment methods also apply to the <code>genoMatrix</code> object. Here we first show some basic operations:</p>
<p>retrieve colData (i.e. sample info):</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">GT</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## DataFrame with 6 rows and 10 columns</span>
<span class="co">##              IID       sex     pheno         pop    superPop               PC1</span>
<span class="co">##      &lt;character&gt; &lt;numeric&gt; &lt;integer&gt; &lt;character&gt; &lt;character&gt;         &lt;numeric&gt;</span>
<span class="co">## ALS1        ALS1         0         1         PJL         SAS  22.5874880632767</span>
<span class="co">## ALS2        ALS2         0         1         BEB         SAS  26.3736999801643</span>
<span class="co">## ALS3        ALS3         0         1         TSI         EUR  30.7549638674014</span>
<span class="co">## ALS4        ALS4         0         1         GWD         AFR  -87.480352070988</span>
<span class="co">## ALS5        ALS5         0         1         MSL         AFR -87.7931031380612</span>
<span class="co">## ALS6        ALS6         0         1         YRI         AFR -86.6407098360629</span>
<span class="co">##                    PC2                PC3                PC4              age</span>
<span class="co">##              &lt;numeric&gt;          &lt;numeric&gt;          &lt;numeric&gt;        &lt;numeric&gt;</span>
<span class="co">## ALS1  4.73802540565646   38.9157563628611   -10.166734241451 66.0174297334819</span>
<span class="co">## ALS2 -1.44590323607503   31.9712988928678  -6.02225552041664 66.8430629274916</span>
<span class="co">## ALS3  48.8668428777611  -5.40924038256157   14.9474404479132 54.3574926927099</span>
<span class="co">## ALS4 -6.19362845092275 -0.950413141050929  -1.20299315583993 65.0977652481132</span>
<span class="co">## ALS5 -5.32392517636858  -1.03533000342774 -0.740323422020038 58.9466357442474</span>
<span class="co">## ALS6 -3.41841650878776  -1.39649766160714   1.05616066833012 63.1363470385964</span></code></pre>
<p>retrieve rowData (i.e. variant info):</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">GT</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## DataFrame with 6 rows and 3 columns</span>
<span class="co">##           ploidy         w        AF</span>
<span class="co">##      &lt;character&gt; &lt;numeric&gt; &lt;numeric&gt;</span>
<span class="co">## 1405     diploid         1   0.00014</span>
<span class="co">## 1406     diploid         1     2e-05</span>
<span class="co">## 1408     diploid         1     2e-05</span>
<span class="co">## 1409     diploid         1     2e-05</span>
<span class="co">## 1411     diploid         1   0.00012</span>
<span class="co">## 1412     diploid         1     2e-05</span></code></pre>
<p>show metadata:</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Annotated-class.html" class="external-link">metadata</a></span><span class="op">(</span><span class="va">GT</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## $ploidyLevels</span>
<span class="co">## [1] "diploid"</span>
<span class="co">## </span>
<span class="co">## $m</span>
<span class="co">## [1] 25000</span>
<span class="co">## </span>
<span class="co">## $nvar</span>
<span class="co">## [1] 32</span>
<span class="co">## </span>
<span class="co">## $varSetName</span>
<span class="co">## [1] "unnamed"</span>
<span class="co">## </span>
<span class="co">## $unit</span>
<span class="co">## [1] "unnamed"</span>
<span class="co">## </span>
<span class="co">## $cohort</span>
<span class="co">## [1] "pheno"</span>
<span class="co">## </span>
<span class="co">## $geneticModel</span>
<span class="co">## [1] "allelic"</span>
<span class="co">## </span>
<span class="co">## $imputeMethod</span>
<span class="co">## [1] "none"</span></code></pre>
<p>A <code>genoMatrix</code> object can be subsetted similarly as a dataframe:</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">GT</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></code></pre></div>
<pre><code><span class="co">## rvat genoMatrix Object</span>
<span class="co">## class: genoMatrix </span>
<span class="co">## dim: 5 5 </span>
<span class="co">## metadata(8): ploidyLevels m ... geneticModel imputeMethod</span>
<span class="co">## assays(1): GT</span>
<span class="co">## rownames(5): 1405 1406 1408 1409 1411</span>
<span class="co">## rowData names(3): ploidy w AF</span>
<span class="co">## colnames(5): ALS1 ALS2 ALS3 ALS4 ALS5</span>
<span class="co">## colData names(10): IID sex ... PC4 age</span></code></pre>
<p>subset subjects with pheno == 1</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">GT</span><span class="op">[</span>,<span class="va">GT</span><span class="op">$</span><span class="va">pheno</span> <span class="op">==</span> <span class="fl">1</span><span class="op">]</span></code></pre></div>
<pre><code><span class="co">## rvat genoMatrix Object</span>
<span class="co">## class: genoMatrix </span>
<span class="co">## dim: 32 5000 </span>
<span class="co">## metadata(8): ploidyLevels m ... geneticModel imputeMethod</span>
<span class="co">## assays(1): GT</span>
<span class="co">## rownames(32): 1405 1406 ... 1453 1454</span>
<span class="co">## rowData names(3): ploidy w AF</span>
<span class="co">## colnames(5000): ALS1 ALS2 ... ALS4999 ALS5000</span>
<span class="co">## colData names(10): IID sex ... PC4 age</span></code></pre>
<p>subset first 2 variants for subjects older than 70:</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">GT</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span>,<span class="va">GT</span><span class="op">$</span><span class="va">age</span> <span class="op">&gt;</span> <span class="fl">70</span><span class="op">]</span></code></pre></div>
<pre><code><span class="co">## rvat genoMatrix Object</span>
<span class="co">## class: genoMatrix </span>
<span class="co">## dim: 2 3910 </span>
<span class="co">## metadata(8): ploidyLevels m ... geneticModel imputeMethod</span>
<span class="co">## assays(1): GT</span>
<span class="co">## rownames(2): 1405 1406</span>
<span class="co">## rowData names(3): ploidy w AF</span>
<span class="co">## colnames(3910): ALS17 ALS31 ... CTRL19990 CTRL19995</span>
<span class="co">## colData names(10): IID sex ... PC4 age</span></code></pre>
<p>Although you’ll usually not work directly on the genotypes, they can be retrieved as followingL</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">GT</span><span class="op">)</span><span class="op">$</span><span class="va">GT</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></code></pre></div>
<pre><code><span class="co">##      ALS1 ALS2 ALS3 ALS4 ALS5</span>
<span class="co">## 1405    0    0    0    0    0</span>
<span class="co">## 1406    0    0    0    0    0</span>
<span class="co">## 1408    0    0    0    0    0</span>
<span class="co">## 1409    0    0    0    0    0</span>
<span class="co">## 1411    0    0    0    0    0</span></code></pre>
</div>
<div class="section level3">
<h3 id="rvat-methods">RVAT methods<a class="anchor" aria-label="anchor" href="#rvat-methods"></a>
</h3>
<p>return the variant allele frequencies using the <code><a href="../reference/genoMatrix.html">getAF()</a></code> method:</p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/genoMatrix.html">getAF</a></span><span class="op">(</span><span class="va">GT</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##    1405    1406    1408    1409    1411    1412    1414    1415    1417    1418 </span>
<span class="co">## 0.00014 0.00002 0.00002 0.00002 0.00012 0.00002 0.00004 0.00002 0.00002 0.00002 </span>
<span class="co">##    1419    1420    1421    1424    1425    1426    1429    1430    1431    1432 </span>
<span class="co">## 0.00002 0.00002 0.00002 0.00002 0.00002 0.00004 0.00002 0.00006 0.00002 0.00002 </span>
<span class="co">##    1433    1434    1435    1439    1440    1441    1447    1448    1451    1452 </span>
<span class="co">## 0.00170 0.00004 0.00002 0.00004 0.00002 0.00026 0.00002 0.00004 0.00006 0.00002 </span>
<span class="co">##    1453    1454 </span>
<span class="co">## 0.00002 0.00002</span></code></pre>
<p>reurn the alternate allele counts:</p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/genoMatrix.html">getAC</a></span><span class="op">(</span><span class="va">GT</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## 1405 1406 1408 1409 1411 1412 1414 1415 1417 1418 1419 1420 1421 1424 1425 1426 </span>
<span class="co">##    7    1    1    1    6    1    2    1    1    1    1    1    1    1    1    2 </span>
<span class="co">## 1429 1430 1431 1432 1433 1434 1435 1439 1440 1441 1447 1448 1451 1452 1453 1454 </span>
<span class="co">##    1    3    1    1   85    2    1    2    1   13    1    2    3    1    1    1</span></code></pre>
<p>Flip genotype dosages such that all GT values represent minor allele counts (note that in this case that was already the case) using the <code><a href="../reference/genoMatrix.html">flipToMinor()</a></code> method:</p>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">GT</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/genoMatrix.html">flipToMinor</a></span><span class="op">(</span><span class="va">GT</span><span class="op">)</span></code></pre></div>
<p>Return a variant summary using the <code><a href="../reference/summariseGeno.html">summariseGeno()</a></code> method:</p>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sumgeno</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/summariseGeno.html">summariseGeno</a></span><span class="op">(</span><span class="va">GT</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sumgeno</span><span class="op">)</span> </code></pre></div>
<pre><code><span class="co">##      VAR_id      AF callRate geno0 geno1 geno2 hweP</span>
<span class="co">## 1405   1405 0.00014        1 24993     7     0    1</span>
<span class="co">## 1406   1406 0.00002        1 24999     1     0    1</span>
<span class="co">## 1408   1408 0.00002        1 24999     1     0    1</span>
<span class="co">## 1409   1409 0.00002        1 24999     1     0    1</span>
<span class="co">## 1411   1411 0.00012        1 24994     6     0    1</span>
<span class="co">## 1412   1412 0.00002        1 24999     1     0    1</span></code></pre>
<p>Meanimpute genotype dosages using the <code><a href="../reference/recode.html">recode()</a></code> method:</p>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/recode.html">recode</a></span><span class="op">(</span><span class="va">GT</span>, imputeMethod <span class="op">=</span> <span class="st">"meanImpute"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## rvat genoMatrix Object</span>
<span class="co">## class: genoMatrix </span>
<span class="co">## dim: 32 25000 </span>
<span class="co">## metadata(8): ploidyLevels m ... geneticModel imputeMethod</span>
<span class="co">## assays(1): GT</span>
<span class="co">## rownames(32): 1405 1406 ... 1453 1454</span>
<span class="co">## rowData names(3): ploidy w AF</span>
<span class="co">## colnames(25000): ALS1 ALS2 ... CTRL19999 CTRL20000</span>
<span class="co">## colData names(10): IID sex ... PC4 age</span></code></pre>
<p>Apply dominant model using the <code><a href="../reference/recode.html">recode()</a></code> method:</p>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/recode.html">recode</a></span><span class="op">(</span><span class="va">GT</span>, geneticModel <span class="op">=</span> <span class="st">"dominant"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## rvat genoMatrix Object</span>
<span class="co">## class: genoMatrix </span>
<span class="co">## dim: 32 25000 </span>
<span class="co">## metadata(8): ploidyLevels m ... geneticModel imputeMethod</span>
<span class="co">## assays(1): GT</span>
<span class="co">## rownames(32): 1405 1406 ... 1453 1454</span>
<span class="co">## rowData names(3): ploidy w AF</span>
<span class="co">## colnames(25000): ALS1 ALS2 ... CTRL19999 CTRL20000</span>
<span class="co">## colData names(10): IID sex ... PC4 age</span></code></pre>
<p>Aggregate genotype counts across variants, creating a single aggregate score per individual (i.e. a burden score):</p>
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">GT</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/aggregate-methods.html" class="external-link">aggregate</a></span><span class="op">(</span><span class="va">GT</span>, returnGT <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">GT</span><span class="op">$</span><span class="va">aggregate</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. </span>
<span class="co">## 0.00000 0.00000 0.00000 0.00592 0.00000 2.00000</span></code></pre>
<p>Retrieve IDs of carriers per variants:</p>
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/genoMatrix.html">getCarriers</a></span><span class="op">(</span><span class="va">GT</span>, colDataFields <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pop"</span>,<span class="st">"age"</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,<span class="op">]</span> </code></pre></div>
<pre><code><span class="co">##    VAR_id      IID genotype pop      age</span>
<span class="co">## 1    1405   ALS179        1 LWK 67.18991</span>
<span class="co">## 2    1405   ALS562        1 GIH 59.16450</span>
<span class="co">## 3    1405  ALS2001        1 ACB 71.93994</span>
<span class="co">## 4    1405  ALS2475        1 CLM 59.71758</span>
<span class="co">## 5    1405  ALS3109        1 IBS 68.99075</span>
<span class="co">## 6    1405  ALS3298        1 MSL 55.14870</span>
<span class="co">## 7    1405  ALS4948        1 PEL 66.82802</span>
<span class="co">## 8    1406  ALS1849        1 STU 65.67560</span>
<span class="co">## 9    1408 CTRL7179        1 IBS 73.71754</span>
<span class="co">## 10   1409  ALS3087        1 STU 71.16354</span></code></pre>
</div>
<div class="section level3">
<h3 id="association-testing">Association testing<a class="anchor" aria-label="anchor" href="#association-testing"></a>
</h3>
<p>RVAT provides an unified interface to various rare variant collapsing methods, while allowing for MAF weighting, filtering etc. There is a dedicated tutorial for association testing <code><a href="../articles/association_testing.html">vignette("association_testing")</a></code> , but we include a brief primer here:</p>
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">burden_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/assocTest.html">assocTest</a></span><span class="op">(</span><span class="va">GT</span>,
                            pheno <span class="op">=</span> <span class="st">"pheno"</span>,
                            covar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PC1"</span>, <span class="st">"PC2"</span>, <span class="st">"PC3"</span>, <span class="st">"PC4"</span><span class="op">)</span>,
                            test <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"firth"</span>, <span class="st">"skat"</span>, <span class="st">"acatv"</span><span class="op">)</span>,
                            name <span class="op">=</span> <span class="st">"example"</span><span class="op">)</span> <span class="co"># optional name of the analysis</span></code></pre></div>
<pre><code><span class="co">## Keeping 25000/25000 available samples for analysis.</span></code></pre>
<pre><code><span class="co">## 32/32 variants are retained for analysis</span></code></pre>
<pre><code><span class="co">## Warning in .rvat_ACAT(pvals[is.keep], weights.transformed[is.keep]): There are</span>
<span class="co">## p-values that are exactly 0!</span></code></pre>
<div class="sourceCode" id="cb79"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">burden_results</span></code></pre></div>
<pre><code><span class="co">## rvbResult with 3 rows and 23 columns</span>
<span class="co">##   cohort varSetName    name        unit pheno           covar geneticModel</span>
<span class="co">##    &lt;Rle&gt;      &lt;Rle&gt;   &lt;Rle&gt; &lt;character&gt; &lt;Rle&gt;           &lt;Rle&gt;        &lt;Rle&gt;</span>
<span class="co">## 1  pheno    unnamed example     unnamed pheno PC1,PC2,PC3,PC4      allelic</span>
<span class="co">## 2  pheno    unnamed example     unnamed pheno PC1,PC2,PC3,PC4      allelic</span>
<span class="co">## 3  pheno    unnamed example     unnamed pheno PC1,PC2,PC3,PC4      allelic</span>
<span class="co">##   MAFweight  test      nvar caseCarriers ctrlCarriers meanCaseScore</span>
<span class="co">##       &lt;Rle&gt; &lt;Rle&gt; &lt;numeric&gt;    &lt;numeric&gt;    &lt;numeric&gt;     &lt;numeric&gt;</span>
<span class="co">## 1         1 firth        32           91           41        0.0214</span>
<span class="co">## 2         1  skat        32           91           41        0.0214</span>
<span class="co">## 3         1 acatv        32           91           41        0.0214</span>
<span class="co">##   meanCtrlScore     caseN     ctrlN caseCallRate ctrlCallRate</span>
<span class="co">##       &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;    &lt;numeric&gt;    &lt;numeric&gt;</span>
<span class="co">## 1       0.00205      5000     20000            1            1</span>
<span class="co">## 2       0.00205      5000     20000            1            1</span>
<span class="co">## 3       0.00205      5000     20000            1            1</span>
<span class="co">##                      P           effect          effectSE    effectCIlower</span>
<span class="co">##              &lt;numeric&gt;        &lt;numeric&gt;         &lt;numeric&gt;        &lt;numeric&gt;</span>
<span class="co">## 1                    0 7.95799525968347 0.518105731692591 5.60394582370392</span>
<span class="co">## 2 9.51162575421858e-17               NA                NA               NA</span>
<span class="co">## 3                    0               NA                NA               NA</span>
<span class="co">##     effectCIupper</span>
<span class="co">##         &lt;numeric&gt;</span>
<span class="co">## 1 11.545132041146</span>
<span class="co">## 2              NA</span>
<span class="co">## 3              NA</span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="more-on-the-gdb">More on the gdb<a class="anchor" aria-label="anchor" href="#more-on-the-gdb"></a>
</h2>
<div class="section level3">
<h3 id="subsetting-a-gdb">Subsetting a gdb<a class="anchor" aria-label="anchor" href="#subsetting-a-gdb"></a>
</h3>
<p>A gdb can be subsetted, retaining queried variants. This can be useful to downsize a (big) gdb for targeted local analyses.</p>
<div class="sourceCode" id="cb81"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/subsetGdb.html">subsetGdb</a></span><span class="op">(</span><span class="va">gdb</span>, 
          output <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"/rvat_tutorials_subset.gdb"</span><span class="op">)</span>,
          where <span class="op">=</span> <span class="st">"gene_name = 'ABCA4'"</span>,
          intersect <span class="op">=</span> <span class="st">"varinfo"</span>, 
          overWrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Output gdb creation complete</span></code></pre>
<div class="sourceCode" id="cb83"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gdb_subset</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gdb.html">gdb</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"/rvat_tutorials_subset.gdb"</span><span class="op">)</span><span class="op">)</span>
<span class="va">anno</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getAnno.html">getAnno</a></span><span class="op">(</span><span class="va">gdb_subset</span>,table <span class="op">=</span> <span class="st">"varinfo"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedArray-utils.html" class="external-link">table</a></span><span class="op">(</span><span class="va">anno</span><span class="op">$</span><span class="va">gene_name</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## </span>
<span class="co">## ABCA4 </span>
<span class="co">##   646</span></code></pre>
</div>
<div class="section level3">
<h3 id="using-rsqlite">Using RSQLite<a class="anchor" aria-label="anchor" href="#using-rsqlite"></a>
</h3>
<p>Although several convenience methods exist to query and update a <code>gdb</code> (such as <code><a href="../reference/getGT.html">getGT()</a></code>, <code><a href="../reference/getAnno.html">getAnno()</a></code>, <code><a href="../reference/getCohort.html">getCohort()</a></code>, <code><a href="../reference/uploadAnno.html">uploadAnno()</a></code> and <code><a href="../reference/uploadCohort.html">uploadCohort()</a></code>), users can also query the gdb as any other SQLite database. Either from the command-line (as demonstrated in <code><a href="../articles/cmdline.html">vignette("cmdline")</a></code>), or by connecting to the database within R using <a href="https://cran.r-project.org/web/packages/RSQLite/index.html" class="external-link">RSQLite</a> for example.</p>
<p>An example:</p>
<div class="sourceCode" id="cb85"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># connect to the gdb</span>
<span class="va">gdb</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gdb.html">gdb</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"/rvat_tutorials.gdb"</span><span class="op">)</span><span class="op">)</span>

<span class="co"># select high impact SOD1 variants</span>
<span class="va">SOD1_highimpact</span> <span class="op">&lt;-</span> <span class="fu">RSQLite</span><span class="fu">::</span><span class="fu">dbGetQuery</span><span class="op">(</span><span class="va">gdb</span>,
                              <span class="st">"select distinct VAR_id from varinfo where gene_name = 'SOD1' and HighImpact = 1;"</span><span class="op">)</span></code></pre></div>
<p>More sophisticated examples will be included in our workflow examples.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Kevin Kenna, Paul Hop.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
