<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Getting started • rvat</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Getting started">
<meta property="og:description" content="rvat">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rvat</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.3.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/basics.html">Getting started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/basics.html">Getting started</a>
    </li>
    <li>
      <a href="../articles/association_testing.html">Association testing</a>
    </li>
    <li>
      <a href="../articles/explore_visualize_results.html">Visualization</a>
    </li>
    <li>
      <a href="../articles/variant_annotation.html">Variant annotation</a>
    </li>
    <li>
      <a href="../articles/GSA.html">Gene set analysis</a>
    </li>
    <li>
      <a href="../articles/spatial_clustering.html">Spatial clustering</a>
    </li>
    <li>
      <a href="../articles/cmdline.html">Command-line tool</a>
    </li>
  </ul>
</li>
<li>
  <a href="../articles/changelog.html">Changelog</a>
</li>
<li>
  <a href="../articles/faq.html">F.A.Q.</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/KennaLab/rvat" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Getting started</h1>
            
      
      
      <div class="hidden name"><code>basics.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Libraries</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">rvat</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">rvatData</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/SummarizedExperiment" class="external-link">SummarizedExperiment</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="example-data">Example data<a class="anchor" aria-label="anchor" href="#example-data"></a>
</h2>
<p>In these tutorials, we’ll use simulated example data that we included
in the <a href="https://github.com/kennalab/rvatData" class="external-link">rvatData</a>
package.</p>
<p>Included in-memory datasets:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span>package <span class="op">=</span> <span class="st">"rvatData"</span><span class="op">)</span><span class="op">$</span><span class="va">results</span><span class="op">[</span>,<span class="st">"Item"</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "GT"         "GTsmall"    "rvbresults"</span></span></code></pre>
<p>Included on-disk datasets, note that the paths to the on-disk
datasets can be retrieved as below:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/rvatData/man/rvat_example.html" class="external-link">rvat_example</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "rvatData_varsetfile.txt.gz" "rvatData.gdb"              </span></span>
<span><span class="co">## [3] "rvatData.pheno"             "rvatData.varinfo"          </span></span>
<span><span class="co">## [5] "rvatData.vcf.gz"</span></span></code></pre>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gdbpath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rvatData/man/rvat_example.html" class="external-link">rvat_example</a></span><span class="op">(</span><span class="st">"rvatData.gdb"</span><span class="op">)</span></span></code></pre></div>
<p>Some output will be written during the tutorials, use a temporary
directory for that:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">outdir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="setting-up-and-populating-a-gdb">Setting up and populating a gdb<a class="anchor" aria-label="anchor" href="#setting-up-and-populating-a-gdb"></a>
</h2>
<div class="section level3">
<h3 id="creating-a-gdb-based-on-a-vcf-file">Creating a gdb based on a VCF-file<a class="anchor" aria-label="anchor" href="#creating-a-gdb-based-on-a-vcf-file"></a>
</h3>
<p>The first step consists of setting up a so-called <code><a href="../reference/gdb.html">gdb()</a></code>,
an SQLite-based data structure which forms the heart of all RVAT
operation. Gdb files allow for much more rapid and memory efficient
loading of sample genotype data into numeric matrices within R. They
also allow for the upload and integration of additional variant and
sample annotation data to allow for complex data</p>
<p>Here, we show how to create a <code><a href="../reference/gdb.html">gdb()</a></code> object from a
VCF-file. Note that RVAT will assign a unique <code>VAR_id</code> to
each variant, which allow rapid querying of variants, these IDs have no
inherent meaning (i.e. they do not refer to chromosomal position or the
like). Note that providing a genome build is optional, this will be
stored in the gdb metadata, and will be used in downstream analyses to
assign ploidy on sex chromosomes (diploid, XnonPAR, YnonPAR).</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vcfpath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rvatData/man/rvat_example.html" class="external-link">rvat_example</a></span><span class="op">(</span><span class="st">"rvatData.vcf.gz"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/buildGdb.html">buildGdb</a></span><span class="op">(</span>vcf <span class="op">=</span> <span class="va">vcfpath</span>,</span>
<span>         output <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outdir</span>,<span class="st">"/rvat_tutorials.gdb"</span><span class="op">)</span>, </span>
<span>         overWrite <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>         genomeBuild <span class="op">=</span> <span class="st">"GRCh38"</span><span class="op">)</span></span></code></pre></div>
<p>Connect to the gdb you just created using the <code><a href="../reference/gdb.html">gdb()</a></code>
function:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gdb</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gdb.html">gdb</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outdir</span>,<span class="st">"/rvat_tutorials.gdb"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="import-variant-info">Import variant info<a class="anchor" aria-label="anchor" href="#import-variant-info"></a>
</h3>
<p>add variant info to the gdb using the <code><a href="../reference/uploadAnno.html">uploadAnno()</a></code>
method, variants are mapped to the gdb <code>VAR_id</code> based on the
CHROM,POS,REF,ALT fields.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># example file including info such as chromosomal location, annotated gene and variant impact.</span></span>
<span><span class="va">varinfopath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rvatData/man/rvat_example.html" class="external-link">rvat_example</a></span><span class="op">(</span><span class="st">"rvatData.varinfo"</span><span class="op">)</span></span>
<span><span class="va">varinfo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="va">varinfopath</span>, header <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># the `name` parameter specifies the name of the table in the gdb, </span></span>
<span><span class="co"># the variant info that should be imported is specified using the `value` parameter</span></span>
<span><span class="fu"><a href="../reference/uploadAnno.html">uploadAnno</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">gdb</span>, name <span class="op">=</span> <span class="st">"varInfo"</span>,value <span class="op">=</span> <span class="va">varinfo</span><span class="op">)</span></span></code></pre></div>
<p>The <code><a href="../reference/gdb.html">listAnno()</a></code> method lists all variant annotation
tables in the gdb:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/gdb.html">listAnno</a></span><span class="op">(</span><span class="va">gdb</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##      name               value                     date</span></span>
<span><span class="co">## 1 varInfo interactive_session Wed Feb 12 12:38:49 2025</span></span></code></pre>
<p>The <code><a href="../reference/getAnno.html">getAnno()</a></code> method retrieves a given variant
annotation table from the gdb:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">varinfo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getAnno.html">getAnno</a></span><span class="op">(</span><span class="va">gdb</span>, table <span class="op">=</span> <span class="st">"varinfo"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">varinfo</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   VAR_id CHROM      POS          ID REF ALT     QUAL FILTER AC    AN</span></span>
<span><span class="co">## 1      1  chr1 11013912        &lt;NA&gt;   A   G   722.13     NA  1 45484</span></span>
<span><span class="co">## 2      2  chr1 11013928 rs755357622   C   T  2598.23     NA  1 49194</span></span>
<span><span class="co">## 3      3  chr1 11013936        &lt;NA&gt;   A   C  4135.36     NA  1 50000</span></span>
<span><span class="co">## 4      4  chr1 11013936        &lt;NA&gt;   A   G  4135.36     NA  1 42636</span></span>
<span><span class="co">## 5      5  chr1 11013952        &lt;NA&gt;   C   G   517.13     NA  1 49574</span></span>
<span><span class="co">## 6      6  chr1 11016874  rs80356715   C   T 64910.10     NA 32 49766</span></span>
<span><span class="co">##             AF gene_name HighImpact ModerateImpact Synonymous CADDphred</span></span>
<span><span class="co">## 1 0.0000219858    TARDBP          0              1          0      24.8</span></span>
<span><span class="co">## 2 0.0000203277    TARDBP          0              0          1         .</span></span>
<span><span class="co">## 3 0.0000200000    TARDBP          0              1          0      24.1</span></span>
<span><span class="co">## 4 0.0000234544    TARDBP          0              1          0      22.6</span></span>
<span><span class="co">## 5 0.0000201719    TARDBP          0              0          1         .</span></span>
<span><span class="co">## 6 0.0006430093    TARDBP          0              1          0      22.2</span></span>
<span><span class="co">##   PolyPhen SIFT</span></span>
<span><span class="co">## 1        P    D</span></span>
<span><span class="co">## 2        .    .</span></span>
<span><span class="co">## 3        B    T</span></span>
<span><span class="co">## 4        B    T</span></span>
<span><span class="co">## 5        .    .</span></span>
<span><span class="co">## 6        B    T</span></span></code></pre>
<p>a specific region can be selected as follows:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">varinfo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getAnno.html">getAnno</a></span><span class="op">(</span><span class="va">gdb</span>, table <span class="op">=</span> <span class="st">"varinfo"</span>, ranges <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>CHROM<span class="op">=</span><span class="st">"chr1"</span>, start <span class="op">=</span> <span class="fl">11013847</span>, end <span class="op">=</span> <span class="fl">11016874</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">varinfo</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   VAR_id CHROM      POS          ID REF ALT     QUAL FILTER AC    AN</span></span>
<span><span class="co">## 1      1  chr1 11013912        &lt;NA&gt;   A   G   722.13     NA  1 45484</span></span>
<span><span class="co">## 2      2  chr1 11013928 rs755357622   C   T  2598.23     NA  1 49194</span></span>
<span><span class="co">## 3      3  chr1 11013936        &lt;NA&gt;   A   C  4135.36     NA  1 50000</span></span>
<span><span class="co">## 4      4  chr1 11013936        &lt;NA&gt;   A   G  4135.36     NA  1 42636</span></span>
<span><span class="co">## 5      5  chr1 11013952        &lt;NA&gt;   C   G   517.13     NA  1 49574</span></span>
<span><span class="co">## 6      6  chr1 11016874  rs80356715   C   T 64910.10     NA 32 49766</span></span>
<span><span class="co">##             AF gene_name HighImpact ModerateImpact Synonymous CADDphred</span></span>
<span><span class="co">## 1 0.0000219858    TARDBP          0              1          0      24.8</span></span>
<span><span class="co">## 2 0.0000203277    TARDBP          0              0          1         .</span></span>
<span><span class="co">## 3 0.0000200000    TARDBP          0              1          0      24.1</span></span>
<span><span class="co">## 4 0.0000234544    TARDBP          0              1          0      22.6</span></span>
<span><span class="co">## 5 0.0000201719    TARDBP          0              0          1         .</span></span>
<span><span class="co">## 6 0.0006430093    TARDBP          0              1          0      22.2</span></span>
<span><span class="co">##   PolyPhen SIFT</span></span>
<span><span class="co">## 1        P    D</span></span>
<span><span class="co">## 2        .    .</span></span>
<span><span class="co">## 3        B    T</span></span>
<span><span class="co">## 4        B    T</span></span>
<span><span class="co">## 5        .    .</span></span>
<span><span class="co">## 6        B    T</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="import-sample-info">Import sample info<a class="anchor" aria-label="anchor" href="#import-sample-info"></a>
</h3>
<p>add phenotype info to the gdb using the <code><a href="../reference/uploadCohort.html">uploadCohort()</a></code>
method:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># First we'll read in the phenotype data</span></span>
<span><span class="va">phenopath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rvatData/man/rvat_example.html" class="external-link">rvat_example</a></span><span class="op">(</span><span class="st">"rvatData.pheno"</span><span class="op">)</span></span>
<span><span class="va">pheno</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span> <span class="va">phenopath</span>, header <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Import into the gdb (the `IID` column is used)</span></span>
<span><span class="co"># the `name` parameter specifies the name of the table in the gdb, </span></span>
<span><span class="co"># the sample info that should be important is specified using the `value` parameter</span></span>
<span><span class="fu"><a href="../reference/uploadCohort.html">uploadCohort</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">gdb</span>, name <span class="op">=</span> <span class="st">"pheno"</span>, value <span class="op">=</span> <span class="va">pheno</span><span class="op">)</span></span></code></pre></div>
<p>The <code><a href="../reference/gdb.html">listCohort()</a></code> method lists all uploaded sample info
tables in the gdb:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/gdb.html">listCohort</a></span><span class="op">(</span><span class="va">gdb</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    name               value                     date</span></span>
<span><span class="co">## 1 pheno interactive_session Wed Feb 12 12:38:50 2025</span></span></code></pre>
<p>The <code><a href="../reference/getCohort.html">getCohort()</a></code> method retrieves a given sample info
table from the gdb:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pheno</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getCohort.html">getCohort</a></span><span class="op">(</span><span class="va">gdb</span>, cohort <span class="op">=</span> <span class="st">"pheno"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">pheno</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    IID sex pheno pop superPop       PC1       PC2        PC3         PC4</span></span>
<span><span class="co">## 1 ALS1   2     1 PJL      SAS  22.58749  4.738025 38.9157564 -10.1667342</span></span>
<span><span class="co">## 2 ALS2   2     1 BEB      SAS  26.37370 -1.445903 31.9712989  -6.0222555</span></span>
<span><span class="co">## 3 ALS3   1     1 TSI      EUR  30.75496 48.866843 -5.4092404  14.9474404</span></span>
<span><span class="co">## 4 ALS4   2     1 GWD      AFR -87.48035 -6.193628 -0.9504131  -1.2029932</span></span>
<span><span class="co">## 5 ALS5   1     1 MSL      AFR -87.79310 -5.323925 -1.0353300  -0.7403234</span></span>
<span><span class="co">## 6 ALS6   1     1 YRI      AFR -86.64071 -3.418417 -1.3964977   1.0561607</span></span>
<span><span class="co">##        age</span></span>
<span><span class="co">## 1 66.01743</span></span>
<span><span class="co">## 2 66.84306</span></span>
<span><span class="co">## 3 54.35749</span></span>
<span><span class="co">## 4 65.09777</span></span>
<span><span class="co">## 5 58.94664</span></span>
<span><span class="co">## 6 63.13635</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="genomatrix-objects">GenoMatrix objects<a class="anchor" aria-label="anchor" href="#genomatrix-objects"></a>
</h2>
<p>Central to <code>RVAT</code> is the <code><a href="../reference/genoMatrix.html">genoMatrix()</a></code> object.
The <code><a href="../reference/genoMatrix.html">genoMatrix()</a></code> object is an extension of the BioConductor
<a href="https://bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html" class="external-link">summarizedExperiment</a>
class. A <code><a href="../reference/genoMatrix.html">genoMatrix()</a></code> contains a matrix of genotypes,
variant info and sample info queried from a <code><a href="../reference/gdb.html">gdb()</a></code> object,
various methods are implemented for genoMatrix objects, which we’ll
explore in this section.</p>
<div class="section level3">
<h3 id="construct-a-genomatrix">Construct a genoMatrix<a class="anchor" aria-label="anchor" href="#construct-a-genomatrix"></a>
</h3>
<p>First, we’ll build a genoMatrix object using the <code><a href="../reference/getGT.html">getGT()</a></code>
method. Variants to include can be specified in two main ways: 1) by
directly passing the VAR_ids to the <code><a href="../reference/getGT.html">getGT()</a></code> method, 2) by
providing a set of genomic ranges to retrieve and 23 by supplying a
<code>varSet</code> (discussed later).</p>
<p>Here we’ll start with methods 1-2, varSets are discussed in
<code><a href="../articles/association_testing.html">vignette("association_testing")</a></code> tutorial.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get varinfo</span></span>
<span><span class="va">varinfo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getAnno.html">getAnno</a></span><span class="op">(</span><span class="va">gdb</span>, table <span class="op">=</span> <span class="st">"varinfo"</span>, where <span class="op">=</span> <span class="st">"gene_name = 'SOD1' and ModerateImpact = 1"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">GT</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getGT.html">getGT</a></span><span class="op">(</span><span class="va">gdb</span>, </span>
<span>            VAR_id <span class="op">=</span> <span class="va">varinfo</span><span class="op">$</span><span class="va">VAR_id</span>,</span>
<span>            cohort <span class="op">=</span> <span class="st">"pheno"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Retrieved genotypes for 38 variants</span></span></code></pre>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GT</span></span></code></pre></div>
<pre><code><span><span class="co">## rvat genoMatrix Object</span></span>
<span><span class="co">## class: genoMatrix </span></span>
<span><span class="co">## dim: 38 25000 </span></span>
<span><span class="co">## metadata(11): gdb gdbId ... geneticModel imputeMethod</span></span>
<span><span class="co">## assays(1): GT</span></span>
<span><span class="co">## rownames(38): 1268 1270 ... 1315 1316</span></span>
<span><span class="co">## rowData names(2): ploidy w</span></span>
<span><span class="co">## colnames(25000): ALS1 ALS2 ... CTRL19999 CTRL20000</span></span>
<span><span class="co">## colData names(10): IID sex ... PC4 age</span></span></code></pre>
<p>Similarly we can retrieve variants within a given range, like below.
Annotations can be included by specifying the <code>anno</code>
parameter, and will be available in the <code>rowData</code> of the
<code>genoMatrix</code>:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GT_fromranges</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getGT.html">getGT</a></span><span class="op">(</span></span>
<span>  <span class="va">gdb</span>,</span>
<span>  ranges <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>CHROM <span class="op">=</span> <span class="st">"chr21"</span>, start <span class="op">=</span> <span class="fl">31659666</span>, end <span class="op">=</span> <span class="fl">31668931</span><span class="op">)</span>,</span>
<span>  cohort <span class="op">=</span> <span class="st">"pheno"</span>,</span>
<span>  anno <span class="op">=</span> <span class="st">"varInfo"</span>,</span>
<span>  annoFields <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"VAR_id"</span>, <span class="st">"CHROM"</span>, <span class="st">"POS"</span>, <span class="st">"REF"</span>, <span class="st">"ALT"</span>, <span class="st">"HighImpact"</span>, <span class="st">"ModerateImpact"</span>, <span class="st">"Synonymous"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Retrieved genotypes for 49 variants</span></span></code></pre>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GT_fromranges</span> <span class="op">&lt;-</span> <span class="va">GT_fromranges</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">GT_fromranges</span><span class="op">)</span><span class="op">$</span><span class="va">ModerateImpact</span> <span class="op">==</span> <span class="fl">1</span>,<span class="op">]</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="basic-operations">Basic operations<a class="anchor" aria-label="anchor" href="#basic-operations"></a>
</h3>
<p>Also see <a href="https://bioconductor.org/packages/release/bioc/vignettes/SummarizedExperiment/inst/doc/SummarizedExperiment.html" class="external-link">this</a>
tutorial on SummarizedExperiments objects, the <code><a href="../reference/genoMatrix.html">genoMatrix()</a></code>
class inherits from <code>summarizedExperiment</code>, meaning that all
summarizedExperiment methods also apply to the <code>genoMatrix</code>
object. Here we first show some basic operations:</p>
<p>retrieve colData (i.e. sample info):</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">GT</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 6 rows and 10 columns</span></span>
<span><span class="co">##              IID       sex     pheno         pop    superPop       PC1</span></span>
<span><span class="co">##      &lt;character&gt; &lt;numeric&gt; &lt;integer&gt; &lt;character&gt; &lt;character&gt; &lt;numeric&gt;</span></span>
<span><span class="co">## ALS1        ALS1         2         1         PJL         SAS   22.5875</span></span>
<span><span class="co">## ALS2        ALS2         2         1         BEB         SAS   26.3737</span></span>
<span><span class="co">## ALS3        ALS3         1         1         TSI         EUR   30.7550</span></span>
<span><span class="co">## ALS4        ALS4         2         1         GWD         AFR  -87.4804</span></span>
<span><span class="co">## ALS5        ALS5         1         1         MSL         AFR  -87.7931</span></span>
<span><span class="co">## ALS6        ALS6         1         1         YRI         AFR  -86.6407</span></span>
<span><span class="co">##            PC2       PC3        PC4       age</span></span>
<span><span class="co">##      &lt;numeric&gt; &lt;numeric&gt;  &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">## ALS1   4.73803 38.915756 -10.166734   66.0174</span></span>
<span><span class="co">## ALS2  -1.44590 31.971299  -6.022256   66.8431</span></span>
<span><span class="co">## ALS3  48.86684 -5.409240  14.947440   54.3575</span></span>
<span><span class="co">## ALS4  -6.19363 -0.950413  -1.202993   65.0978</span></span>
<span><span class="co">## ALS5  -5.32393 -1.035330  -0.740323   58.9466</span></span>
<span><span class="co">## ALS6  -3.41842 -1.396498   1.056161   63.1363</span></span></code></pre>
<p>retrieve rowData (i.e. variant info):</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">GT</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 6 rows and 2 columns</span></span>
<span><span class="co">##           ploidy         w</span></span>
<span><span class="co">##      &lt;character&gt; &lt;numeric&gt;</span></span>
<span><span class="co">## 1268     diploid         1</span></span>
<span><span class="co">## 1270     diploid         1</span></span>
<span><span class="co">## 1271     diploid         1</span></span>
<span><span class="co">## 1273     diploid         1</span></span>
<span><span class="co">## 1275     diploid         1</span></span>
<span><span class="co">## 1276     diploid         1</span></span></code></pre>
<p>show metadata:</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Annotated-class.html" class="external-link">metadata</a></span><span class="op">(</span><span class="va">GT</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $gdb</span></span>
<span><span class="co">## [1] "/private/var/folders/cl/wvc0rvjx4vd5rzt2_fhpmfth0000gp/T/RtmpCTy08L/rvat_tutorials.gdb"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $gdbId</span></span>
<span><span class="co">## [1] "u7fd78yvg5fivj57qv4af7ayh65s"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $genomeBuild</span></span>
<span><span class="co">## [1] "GRCh38"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $ploidyLevels</span></span>
<span><span class="co">## [1] "diploid"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $m</span></span>
<span><span class="co">## [1] 25000</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $nvar</span></span>
<span><span class="co">## [1] 38</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $varSetName</span></span>
<span><span class="co">## [1] "unnamed"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $unit</span></span>
<span><span class="co">## [1] "unnamed"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $cohort</span></span>
<span><span class="co">## [1] "pheno"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $geneticModel</span></span>
<span><span class="co">## [1] "allelic"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $imputeMethod</span></span>
<span><span class="co">## [1] "none"</span></span></code></pre>
<p>A <code>genoMatrix</code> object can be subsetted similarly as a
data.frame:</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GT</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## rvat genoMatrix Object</span></span>
<span><span class="co">## class: genoMatrix </span></span>
<span><span class="co">## dim: 5 5 </span></span>
<span><span class="co">## metadata(11): gdb gdbId ... geneticModel imputeMethod</span></span>
<span><span class="co">## assays(1): GT</span></span>
<span><span class="co">## rownames(5): 1268 1270 1271 1273 1275</span></span>
<span><span class="co">## rowData names(2): ploidy w</span></span>
<span><span class="co">## colnames(5): ALS1 ALS2 ALS3 ALS4 ALS5</span></span>
<span><span class="co">## colData names(10): IID sex ... PC4 age</span></span></code></pre>
<p>subset subjects with pheno == 1</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GT</span><span class="op">[</span>,<span class="va">GT</span><span class="op">$</span><span class="va">pheno</span> <span class="op">==</span> <span class="fl">1</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## rvat genoMatrix Object</span></span>
<span><span class="co">## class: genoMatrix </span></span>
<span><span class="co">## dim: 38 5000 </span></span>
<span><span class="co">## metadata(11): gdb gdbId ... geneticModel imputeMethod</span></span>
<span><span class="co">## assays(1): GT</span></span>
<span><span class="co">## rownames(38): 1268 1270 ... 1315 1316</span></span>
<span><span class="co">## rowData names(2): ploidy w</span></span>
<span><span class="co">## colnames(5000): ALS1 ALS2 ... ALS4999 ALS5000</span></span>
<span><span class="co">## colData names(10): IID sex ... PC4 age</span></span></code></pre>
<p>subset first 2 variants for subjects older than 70:</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GT</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span>,<span class="va">GT</span><span class="op">$</span><span class="va">age</span> <span class="op">&gt;</span> <span class="fl">70</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## rvat genoMatrix Object</span></span>
<span><span class="co">## class: genoMatrix </span></span>
<span><span class="co">## dim: 2 3910 </span></span>
<span><span class="co">## metadata(11): gdb gdbId ... geneticModel imputeMethod</span></span>
<span><span class="co">## assays(1): GT</span></span>
<span><span class="co">## rownames(2): 1268 1270</span></span>
<span><span class="co">## rowData names(2): ploidy w</span></span>
<span><span class="co">## colnames(3910): ALS17 ALS31 ... CTRL19990 CTRL19995</span></span>
<span><span class="co">## colData names(10): IID sex ... PC4 age</span></span></code></pre>
<p>Although you’ll usually not work directly on the genotypes, they can
be retrieved as followingL</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">GT</span><span class="op">)</span><span class="op">$</span><span class="va">GT</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##      ALS1 ALS2 ALS3 ALS4 ALS5</span></span>
<span><span class="co">## 1268    0    0    0    0    0</span></span>
<span><span class="co">## 1270    0    0    0    0    0</span></span>
<span><span class="co">## 1271    0    0    0    0    0</span></span>
<span><span class="co">## 1273    0    0    0    0    0</span></span>
<span><span class="co">## 1275    0    0    0    0    0</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="rvat-methods">RVAT methods<a class="anchor" aria-label="anchor" href="#rvat-methods"></a>
</h3>
<p>return the variant allele frequencies using the <code><a href="../reference/genoMatrix.html">getAF()</a></code>
method:</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/genoMatrix.html">getAF</a></span><span class="op">(</span><span class="va">GT</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##         1268         1270         1271         1273         1275         1276 </span></span>
<span><span class="co">## 1.260187e-04 2.021591e-05 2.252049e-05 1.007293e-04 2.000000e-05 2.056428e-05 </span></span>
<span><span class="co">##         1277         1278         1279         1281         1282         1283 </span></span>
<span><span class="co">## 2.000000e-05 2.157404e-05 2.000000e-05 2.183025e-05 2.162256e-05 2.237237e-05 </span></span>
<span><span class="co">##         1284         1285         1287         1288         1289         1290 </span></span>
<span><span class="co">## 2.000000e-05 2.000000e-05 2.128656e-05 2.052798e-05 2.000000e-05 1.420000e-03 </span></span>
<span><span class="co">##         1291         1292         1293         1294         1296         1297 </span></span>
<span><span class="co">## 2.003928e-05 2.045157e-05 2.193463e-05 2.029386e-05 2.133834e-05 8.000000e-05 </span></span>
<span><span class="co">##         1298         1299         1300         1304         1305         1306 </span></span>
<span><span class="co">## 2.139129e-05 2.850859e-04 2.008355e-05 2.015560e-05 2.115059e-05 2.050945e-05 </span></span>
<span><span class="co">##         1307         1308         1309         1312         1313         1314 </span></span>
<span><span class="co">## 2.000000e-05 4.648568e-05 2.072024e-05 4.112688e-05 2.297689e-05 2.025686e-05 </span></span>
<span><span class="co">##         1315         1316 </span></span>
<span><span class="co">## 4.123031e-05 2.115507e-05</span></span></code></pre>
<p>reurn the alternate allele counts:</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/genoMatrix.html">getAC</a></span><span class="op">(</span><span class="va">GT</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 1268 1270 1271 1273 1275 1276 1277 1278 1279 1281 1282 1283 1284 1285 1287 1288 </span></span>
<span><span class="co">##    6    1    1    5    1    1    1    1    1    1    1    1    1    1    1    1 </span></span>
<span><span class="co">## 1289 1290 1291 1292 1293 1294 1296 1297 1298 1299 1300 1304 1305 1306 1307 1308 </span></span>
<span><span class="co">##    1   71    1    1    1    1    1    4    1   14    1    1    1    1    1    2 </span></span>
<span><span class="co">## 1309 1312 1313 1314 1315 1316 </span></span>
<span><span class="co">##    1    2    1    1    2    1</span></span></code></pre>
<p>Flip genotype dosages such that all GT values represent minor allele
counts (note that in this case that was already the case) using the
<code><a href="../reference/genoMatrix.html">flipToMinor()</a></code> method:</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GT</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/genoMatrix.html">flipToMinor</a></span><span class="op">(</span><span class="va">GT</span><span class="op">)</span></span></code></pre></div>
<p>Return a variant summary using the <code><a href="../reference/summariseGeno.html">summariseGeno()</a></code>
method:</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sumgeno</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/summariseGeno.html">summariseGeno</a></span><span class="op">(</span><span class="va">GT</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sumgeno</span><span class="op">)</span> </span></code></pre></div>
<pre><code><span><span class="co">##      VAR_id           AF callRate geno0 geno1 geno2 hweP</span></span>
<span><span class="co">## 1268   1268 1.260187e-04  0.95224 23800     6     0    1</span></span>
<span><span class="co">## 1270   1270 2.021591e-05  0.98932 24732     1     0    1</span></span>
<span><span class="co">## 1271   1271 2.252049e-05  0.88808 22201     1     0    1</span></span>
<span><span class="co">## 1273   1273 1.007293e-04  0.99276 24814     5     0    1</span></span>
<span><span class="co">## 1275   1275 2.000000e-05  1.00000 24999     1     0    1</span></span>
<span><span class="co">## 1276   1276 2.056428e-05  0.97256 24313     1     0    1</span></span></code></pre>
<p>Meanimpute genotype dosages using the <code><a href="../reference/recode.html">recode()</a></code>
method:</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/recode.html">recode</a></span><span class="op">(</span><span class="va">GT</span>, imputeMethod <span class="op">=</span> <span class="st">"meanImpute"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## rvat genoMatrix Object</span></span>
<span><span class="co">## class: genoMatrix </span></span>
<span><span class="co">## dim: 38 25000 </span></span>
<span><span class="co">## metadata(11): gdb gdbId ... geneticModel imputeMethod</span></span>
<span><span class="co">## assays(1): GT</span></span>
<span><span class="co">## rownames(38): 1268 1270 ... 1315 1316</span></span>
<span><span class="co">## rowData names(2): ploidy w</span></span>
<span><span class="co">## colnames(25000): ALS1 ALS2 ... CTRL19999 CTRL20000</span></span>
<span><span class="co">## colData names(10): IID sex ... PC4 age</span></span></code></pre>
<p>Apply dominant model using the <code><a href="../reference/recode.html">recode()</a></code> method:</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/recode.html">recode</a></span><span class="op">(</span><span class="va">GT</span>, geneticModel <span class="op">=</span> <span class="st">"dominant"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## rvat genoMatrix Object</span></span>
<span><span class="co">## class: genoMatrix </span></span>
<span><span class="co">## dim: 38 25000 </span></span>
<span><span class="co">## metadata(11): gdb gdbId ... geneticModel imputeMethod</span></span>
<span><span class="co">## assays(1): GT</span></span>
<span><span class="co">## rownames(38): 1268 1270 ... 1315 1316</span></span>
<span><span class="co">## rowData names(2): ploidy w</span></span>
<span><span class="co">## colnames(25000): ALS1 ALS2 ... CTRL19999 CTRL20000</span></span>
<span><span class="co">## colData names(10): IID sex ... PC4 age</span></span></code></pre>
<p>Aggregate genotype counts across variants, creating a single
aggregate score per individual (i.e. a burden score):</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GT_agg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/aggregate.html" class="external-link">aggregate</a></span><span class="op">(</span><span class="fu"><a href="../reference/recode.html">recode</a></span><span class="op">(</span><span class="va">GT</span>, imputeMethod <span class="op">=</span> <span class="st">"meanImpute"</span><span class="op">)</span>, returnGT <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">GT_agg</span><span class="op">$</span><span class="va">aggregate</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##      Min.   1st Qu.    Median      Mean   3rd Qu.      Max. </span></span>
<span><span class="co">## 0.0000000 0.0000410 0.0000460 0.0055313 0.0001275 2.0002520</span></span></code></pre>
<p>Retrieve IDs of carriers per variants:</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/genoMatrix.html">getCarriers</a></span><span class="op">(</span><span class="va">GT</span>, colDataFields <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pop"</span>,<span class="st">"age"</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,<span class="op">]</span> </span></code></pre></div>
<pre><code><span><span class="co">##    VAR_id       IID genotype pop      age</span></span>
<span><span class="co">## 1    1268   ALS1811        1 YRI 56.59491</span></span>
<span><span class="co">## 2    1268   ALS2109        1 LWK 53.64514</span></span>
<span><span class="co">## 3    1268   ALS2155        1 GIH 45.68899</span></span>
<span><span class="co">## 4    1268   ALS3076        1 YRI 60.34504</span></span>
<span><span class="co">## 5    1268   ALS4396        1 GWD 74.22662</span></span>
<span><span class="co">## 6    1268   ALS4476        1 CHB 62.46120</span></span>
<span><span class="co">## 7    1270 CTRL14060        1 KHV 67.61440</span></span>
<span><span class="co">## 8    1271 CTRL18522        1 PUR 47.41249</span></span>
<span><span class="co">## 9    1273   ALS1117        1 IBS 49.17095</span></span>
<span><span class="co">## 10   1273   ALS3310        1 FIN 50.09404</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="association-testing">Association testing<a class="anchor" aria-label="anchor" href="#association-testing"></a>
</h3>
<p>RVAT provides an unified interface to various rare variant collapsing
methods, while allowing for MAF weighting, filtering etc. There is a
dedicated tutorial for association testing
<code><a href="../articles/association_testing.html">vignette("association_testing")</a></code> , but we include a brief
primer here:</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">burden_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/assocTest.html">assocTest</a></span><span class="op">(</span><span class="va">GT</span>,</span>
<span>                            pheno <span class="op">=</span> <span class="st">"pheno"</span>,</span>
<span>                            covar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PC1"</span>, <span class="st">"PC2"</span>, <span class="st">"PC3"</span>, <span class="st">"PC4"</span><span class="op">)</span>,</span>
<span>                            test <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"firth"</span>, <span class="st">"skat"</span>, <span class="st">"acatv"</span><span class="op">)</span>,</span>
<span>                            name <span class="op">=</span> <span class="st">"example"</span><span class="op">)</span> <span class="co"># optional name of the analysis</span></span>
<span><span class="va">burden_results</span></span></code></pre></div>
<pre><code><span><span class="co">## rvbResult with 3 rows and 24 columns</span></span>
<span><span class="co">##          unit cohort varSetName    name pheno           covar geneticModel</span></span>
<span><span class="co">##   &lt;character&gt;  &lt;Rle&gt;      &lt;Rle&gt;   &lt;Rle&gt; &lt;Rle&gt;           &lt;Rle&gt;        &lt;Rle&gt;</span></span>
<span><span class="co">## 1     unnamed  pheno    unnamed example pheno PC1,PC2,PC3,PC4      allelic</span></span>
<span><span class="co">## 2     unnamed  pheno    unnamed example pheno PC1,PC2,PC3,PC4      allelic</span></span>
<span><span class="co">## 3     unnamed  pheno    unnamed example pheno PC1,PC2,PC3,PC4      allelic</span></span>
<span><span class="co">##   MAFweight  test      nvar caseCarriers ctrlCarriers meanCaseScore</span></span>
<span><span class="co">##       &lt;Rle&gt; &lt;Rle&gt; &lt;numeric&gt;    &lt;numeric&gt;    &lt;numeric&gt;     &lt;numeric&gt;</span></span>
<span><span class="co">## 1         1 firth        38           77           51     0.0170918</span></span>
<span><span class="co">## 2         1  skat        38           77           51     0.0170918</span></span>
<span><span class="co">## 3         1 acatv        38           77           51     0.0170918</span></span>
<span><span class="co">##   meanCtrlScore     caseN     ctrlN caseCallRate ctrlCallRate    effect</span></span>
<span><span class="co">##       &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;    &lt;numeric&gt;    &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">## 1    0.00264118      5000     20000     0.962705     0.962584   1.75552</span></span>
<span><span class="co">## 2    0.00264118      5000     20000     0.962705     0.962584        NA</span></span>
<span><span class="co">## 3    0.00264118      5000     20000     0.962705     0.962584        NA</span></span>
<span><span class="co">##    effectSE effectCIlower effectCIupper        OR          P</span></span>
<span><span class="co">##   &lt;numeric&gt;     &lt;numeric&gt;     &lt;numeric&gt; &lt;numeric&gt;  &lt;numeric&gt;</span></span>
<span><span class="co">## 1  0.176317       1.41495       2.10837   5.78647 0.0000e+00</span></span>
<span><span class="co">## 2        NA            NA            NA        NA 2.1049e-10</span></span>
<span><span class="co">## 3        NA            NA            NA        NA 0.0000e+00</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="more-on-the-gdb">More on the gdb<a class="anchor" aria-label="anchor" href="#more-on-the-gdb"></a>
</h2>
<div class="section level3">
<h3 id="subsetting-a-gdb">Subsetting a gdb<a class="anchor" aria-label="anchor" href="#subsetting-a-gdb"></a>
</h3>
<p>A gdb can be subsetted, retaining queried variants. This can be
useful to downsize a (big) gdb for targeted local analyses.</p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/subsetGdb.html">subsetGdb</a></span><span class="op">(</span><span class="va">gdb</span>, </span>
<span>          output <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"/rvat_tutorials_subset.gdb"</span><span class="op">)</span>,</span>
<span>          where <span class="op">=</span> <span class="st">"gene_name = 'ABCA4'"</span>,</span>
<span>          intersect <span class="op">=</span> <span class="st">"varinfo"</span>, </span>
<span>          overWrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">gdb_subset</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gdb.html">gdb</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"/rvat_tutorials_subset.gdb"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">anno</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getAnno.html">getAnno</a></span><span class="op">(</span><span class="va">gdb_subset</span>,table <span class="op">=</span> <span class="st">"varinfo"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">anno</span><span class="op">$</span><span class="va">gene_name</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## ABCA4 </span></span>
<span><span class="co">##   589</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="using-rsqlite">Using RSQLite<a class="anchor" aria-label="anchor" href="#using-rsqlite"></a>
</h3>
<p>Although several convenience methods exist to query and update a
<code>gdb</code> (such as <code><a href="../reference/getGT.html">getGT()</a></code>, <code><a href="../reference/getAnno.html">getAnno()</a></code>,
<code><a href="../reference/getCohort.html">getCohort()</a></code>, <code><a href="../reference/uploadAnno.html">uploadAnno()</a></code> and
<code><a href="../reference/uploadCohort.html">uploadCohort()</a></code>), users can also query the gdb as any other
SQLite database. Either from the command-line (as demonstrated in
<code><a href="../articles/cmdline.html">vignette("cmdline")</a></code>), or by connecting to the database
within R using <a href="https://cran.r-project.org/web/packages/RSQLite/index.html" class="external-link">RSQLite</a>
for example.</p>
<p>An example:</p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># connect to the gdb</span></span>
<span><span class="va">gdb</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gdb.html">gdb</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"/rvat_tutorials.gdb"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># select high impact SOD1 variants</span></span>
<span><span class="va">SOD1_highimpact</span> <span class="op">&lt;-</span> <span class="fu">RSQLite</span><span class="fu">::</span><span class="fu">dbGetQuery</span><span class="op">(</span><span class="va">gdb</span>,</span>
<span>                              <span class="st">"select distinct VAR_id from varinfo where gene_name = 'SOD1' and HighImpact = 1;"</span><span class="op">)</span></span></code></pre></div>
<p>More sophisticated examples will be included in our workflow
examples.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Kevin Kenna, Paul Hop.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
