<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Changelog • rvat</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Changelog">
<meta property="og:description" content="rvat">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rvat</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.3.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/basics.html">Getting started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/basics.html">Getting started</a>
    </li>
    <li>
      <a href="../articles/association_testing.html">Association testing</a>
    </li>
    <li>
      <a href="../articles/explore_visualize_results.html">Visualization</a>
    </li>
    <li>
      <a href="../articles/variant_annotation.html">Variant annotation</a>
    </li>
    <li>
      <a href="../articles/GSA.html">Gene set analysis</a>
    </li>
    <li>
      <a href="../articles/spatial_clustering.html">Spatial clustering</a>
    </li>
    <li>
      <a href="../articles/cmdline.html">Command-line tool</a>
    </li>
  </ul>
</li>
<li>
  <a href="../articles/changelog.html">Changelog</a>
</li>
<li>
  <a href="../articles/faq.html">F.A.Q.</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/KennaLab/rvat" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Changelog</h1>
            
      
      
      <div class="hidden name"><code>changelog.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="v-0-3">v.0.3<a class="anchor" aria-label="anchor" href="#v-0-3"></a>
</h2>
<div class="section level3">
<h3 id="v-0-3-1">v.0.3.1<a class="anchor" aria-label="anchor" href="#v-0-3-1"></a>
</h3>
<ul>
<li>Includes new <em>mutationPlot</em> function to create mutation plots
that visualize variant-level and gene-level association results along a
transcript structure. Optionally, custom tracks such as protein domains
or mutation clusters can be overlaid.<br>
</li>
<li>The <em>mutationPlot</em> function is demonstrated in the updated <a href="spatial_clustering.html">spatial clustering</a> tutorial.<br>
</li>
<li>Includes a new method <em>getRanges</em> for <code>varSetList</code>
and <code>varSetFile</code> that retrieves genomic ranges for variant
sets. Can be used to retrieve either genomic coordinates or other
user-defined coordinates such as CDS coordinates genearated with the
<code>mapCDS</code> method.</li>
</ul>
</div>
<div class="section level3">
<h3 id="v-0-3-0">v.0.3.0<a class="anchor" aria-label="anchor" href="#v-0-3-0"></a>
</h3>
<p>Large update! In particular, changes have been made to:</p>
<ul>
<li>
<strong>Documentation and tutorials</strong>
<ul>
<li>All methods/classes/functions now contain examples in their
documentation.</li>
<li>New tutorials are included on <a href="variant_annotation.html">variant annotation</a>, <a href="GSA.html">gene set analysis</a> and <a href="spatial_clustering.html">spatial clustering</a>.<br>
Existing workflows have been updated with new features/functions .</li>
</ul>
</li>
<li>
<strong>Association testing</strong>
<ul>
<li>RVAT results now include a <code>OR</code> field indicating the
odds-ratio for logistic tests (firth and glm).
<ins><strong>Importantly:</strong></ins> For firth and flm tests
<code>effect</code>/<code>effectSE</code>/<code>effectCIlower</code>/<code>effectCIupper</code>
are now on the log-scale!</li>
<li>Effect alleles (i.e. the allele to which the effect estimate refers)
and non-effect alleles are now included in the results of single variant
tests.</li>
</ul>
</li>
<li>
<strong>Metadata</strong>
<ul>
<li>GDB now includes additional metadata, including a unique identifier,
the genome build and creation date and can be retrieved using
corresponding getters (e.g. <code>getGenomeBuild</code> and
<code>getGdbId</code>). The genome build can be set when building the
gdb (<code>genomeBuild</code> parameter), and will then be used
correctly assign ploidies on the sex-chromosomes in downstream analyses
(so no need to set the <code>checkPloidy</code> parameter anymore,
although this is still an option). The gdb identifier can be useful to
track which gdb was used to generate results, as the id will be included
in the metadata of downstream files such as results files and
varSetFiles (see below). Also, it is used methods such as
<code>assocTest</code> and <code>summariseGeno</code> to check whether
supplied varSetFile/varSetList files were generated from the specified
gdb.</li>
<li>Files generated using RVAT, such as rvatResult and varSetFiles, now
include a header with metadata, including the RVAT version used to
generate the file, the gdb identifier, the genome build and the creation
date. This metadata will be included when loading or connecting to these
files with the corresponding RVAT methods
(e.g. <code>rvbResult(&lt;file&gt;)</code> and
<code>varSetFile(&lt;file&gt;)</code>). Note that if you read in RVAT
results using non-RVAT functions (e.g. <code>read.table</code>), then
you will have to skip lines that start with ‘#’.</li>
</ul>
</li>
<li>
<strong>Robustness, fixes &amp; efficiency</strong>
<ul>
<li>Unit tests are now included that cover the majority of code.</li>
<li>RVAT now passes R CMD CHECK without errors / warnings.</li>
<li>Complete rewrite of the RVAT command-line interface. This is mostly
an ‘under-the-hood’ update to increase robustness and maintainability.
For users it is relevant that 1) CLI + help pages are now in-sync with
interactive version 2) parameters can now be provided without an ‘=’,
e.g. <code>--gdb &lt;path to gdb&gt;</code> rather than
<code>--gdb=&lt;path to gdb&gt;</code>
</li>
<li>Increased efficiency: significant speed-up of extracting ranges
(<code>ranges</code> parameter in <code>getGT</code> and
<code>getAnno</code>).</li>
<li>Several bug fixes / improvements.</li>
</ul>
</li>
<li>
<strong>Miscellanous</strong>
<ul>
<li>Annotations can now be included directly when loading genotypes
using <code><a href="../reference/getGT.html">getGT()</a></code>by setting the <code>anno</code> parameter, or
setting <code>inludeVarInfo = TRUE</code> to include the ‘var’ table. If
‘REF’ and ‘ALT’ alleles are included in the annotations, ‘effectAllele’
and ‘otherAllele’ are assigned in the <code>genoMatrix</code> and
automatically updated when alleles are flipped. effectAllele and
otherAllele are passed on as a fields in single variant results.</li>
<li>
<code>subsetGdb</code>: Now has a <code>VAR_id</code> parameter to
directly subset VAR_ids and a <code>tables</code> parameter to specify
which tables to keep.</li>
<li>
<code>getCohort</code> now returns records only for the samples
included in the respective cohort, rather than running NA fields for all
excluded samples. The <code>keepAll</code> parameter can be set to
<code>TRUE</code> to return all records, like in previous versions.</li>
<li>
<code>uploadAnno</code>: now has a <code>keepUnmapped</code> flag,
if set to TRUE (default) variants that do not map to gdb are
discarded.</li>
<li>
<code>writeVcf</code> now includes a <code>includeVarId</code>
parameter that controls whether VAR_ids should be included in the ‘ID’
field. This parameter defaults to <code>FALSE</code>, in which case the
‘ID’ field from the ‘var’ table is included. <em>Note</em>: this is a
change relative to the previous version, in which VAR_ids were written
to the ‘ID’ field by default. Also, FORMAT field not included anymore
when generating a sites-only vcf (<code>includeGeno=TRUE</code>).</li>
<li>‘AF’ field in rowData slot of genoMatrix is removed.</li>
<li>
<code>mergeAggregateFiles</code> is now split into two methods:
<code>mergeAggregateFiles</code> and
<code>collapseAggregateFiles</code>. <code>mergeAggregateFiles</code>
behaves identically to former <code>mergeAggregateFiles</code> method
with <code>collapse = FALSE</code>, whereas
<code>collapseAggregateFiles</code> behaved identically to former
<code>mergeAggregateFiles</code> method with
<code>collapse = TRUE</code>.</li>
</ul>
</li>
</ul>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Kevin Kenna, Paul Hop.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
