<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>RVAT in the command-line â€¢ rvat</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="RVAT in the command-line">
<meta property="og:description" content="rvat">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rvat</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/basics.html">Basics</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/basics.html">Basics</a>
    </li>
    <li>
      <a href="../articles/association_testing.html">Association testing</a>
    </li>
    <li>
      <a href="../articles/explore_visualize_results.html">Visualization</a>
    </li>
    <li>
      <a href="../articles/cmdline.html">Command-line tool</a>
    </li>
  </ul>
</li>
<li>
  <a href="../articles/faq.html">F.A.Q.</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/kkenna/rvat" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>RVAT in the command-line</h1>
            
      
      
      <div class="hidden name"><code>cmdline.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="getting-started">Getting started<a class="anchor" aria-label="anchor" href="#getting-started"></a>
</h2>
<p>Some output will be written during the tutorials, a temporary directory will be used for that:</p>
<p>Identify RVAT executable path:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co">#rvat=$(Rscript -e "cat(system.file('exec/rvat.R', package = 'rvat'))")</span></span></code></pre></div>
<p>Main help page:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="ex">Rscript</span> <span class="va">$rvat</span></span></code></pre></div>
<pre><code><span class="co">## Start time: 2023-05-25 15:24:41</span>
<span class="co">## RVAT version: 0.2.0</span>
<span class="co">## RVAT arguments detected:</span>
<span class="co">## Error: Rare variant analysis toolkit (rvat)</span>
<span class="co">## </span>
<span class="co">## Provide a valid function call for detailed help.</span>
<span class="co">##   --buildGdb</span>
<span class="co">##   --concatGdb</span>
<span class="co">##   --uploadAnno</span>
<span class="co">##   --mapVariants</span>
<span class="co">##   --uploadCohort</span>
<span class="co">##   --listAnno</span>
<span class="co">##   --listCohort</span>
<span class="co">##   --dropTable</span>
<span class="co">##   --subsetGdb</span>
<span class="co">##   --buildVarSet</span>
<span class="co">##   --spatialClust</span>
<span class="co">##   --assocTest</span>
<span class="co">##   --summariseGeno</span>
<span class="co">##   --aggregate</span>
<span class="co">##   --mergeAggregateFiles</span>
<span class="co">##   --geneSetAssoc</span>
<span class="co">##   --buildResamplingFile</span>
<span class="co">##   --vcfInfo2Table</span>
<span class="co">## Execution halted</span></code></pre>
<p>Function specific help page:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="ex">Rscript</span> <span class="va">$rvat</span> <span class="at">--buildGdb</span></span></code></pre></div>
<pre><code><span class="co">## Start time: 2023-05-25 15:24:46</span>
<span class="co">## RVAT version: 0.2.0</span>
<span class="co">## RVAT arguments detected:</span>
<span class="co">## --buildGdb=--buildGdb</span>
<span class="co">## </span>
<span class="co">##   buildGdb</span>
<span class="co">##   </span>
<span class="co">##   Creates a new gdb file and returns a connection object of type gdb-class. The gdb can be structured and populated using a provided vcf, bgen or plink file. </span>
<span class="co">##   If no input variant file is provided then only an empty gdb is created.</span>
<span class="co">##   </span>
<span class="co">##   Usage:</span>
<span class="co">##     Rscript rvat.R --buildGdb --vcf={vcf} --output={output} [options]</span>
<span class="co">##  </span>
<span class="co">##   Arguments:</span>
<span class="co">##     output           Path for output gdb file</span>
<span class="co">##     vcf              Input vcf file used to structure and populate gdb. Warning this function makes the following of assumptions: 1) strict adherence to vcf format (GT subfield first element in genotype firelds), 2) multiallelic records have been split, 3) desired genotype QC has already been applied (DP,GQ filters), 4) GT values conform to the set {0/0,0/1,1/0,1/1,./.,0|0,0|1,1|0,1|1,.|.}. Multiallelic parsing and genotype QC can be performed using vcftools and/or accompanying parser scripts included on the rvat github.</span>
<span class="co">##     skipIndexes      Flag to skip generation of indexes for var and dosage table (VAR_id;CHROM, POS,REF,ALT). Typically only required if you plan to use gdbConcat to concatenate a series of separately generated gdb files before use</span>
<span class="co">##     overWrite        overwrite if `output` already exists? Defaults to `FALSE`, in which case an error is raised.</span>
<span class="co">##   </span>
<span class="co">## Error: </span>
                                                                                                                                                             
<span class="co">## Execution halted</span></code></pre>
</div>
<div class="section level2">
<h2 id="build-gdb">Build GDB<a class="anchor" aria-label="anchor" href="#build-gdb"></a>
</h2>
<div class="section level3">
<h3 id="import-vcf">Import VCF:<a class="anchor" aria-label="anchor" href="#import-vcf"></a>
</h3>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="ex">Rscript</span> <span class="va">$rvat</span> <span class="at">--buildGdb</span> <span class="at">--vcf</span><span class="op">=</span><span class="va">${vcfpath}</span> <span class="at">--output</span><span class="op">=</span><span class="va">$outdir</span>/rvat_tutorials_cmdline.gdb <span class="at">--overWrite</span><span class="op">=</span>TRUE</span></code></pre></div>
<pre><code><span class="co">## Start time: 2023-05-25 15:24:51</span>
<span class="co">## RVAT version: 0.2.0</span>
<span class="co">## RVAT arguments detected:</span>
<span class="co">## --buildGdb=--buildGdb</span>
<span class="co">## --vcf=/Users/phop2/Library/R/3.6/library/rvatData/extdata/rvatData.vcf.gz</span>
<span class="co">## --output=/var/folders/cl/wvc0rvjx4vd5rzt2_fhpmfth0000gp/T//RtmptNGlXA/rvat_tutorials_cmdline.gdb</span>
<span class="co">## --overWrite=TRUE</span>
<span class="co">## 2023-05-25 15:24:55  Creating gdb tables</span>
<span class="co">## 25000 sample IDs detected</span>
<span class="co">## 2023-05-25 15:24:56  Parsing vcf records</span>
<span class="co">## 2023-05-25 15:25:17  Processing completed for 1000 records. Committing to db.</span>
<span class="co">## 2023-05-25 15:25:32  Processing completed for 1763 records. Committing to db.</span>
<span class="co">## 2023-05-25 15:25:32  Creating var table indexes</span>
<span class="co">## 2023-05-25 15:25:32  Creating SM table index</span>
<span class="co">## 2023-05-25 15:25:32  Creating dosage table index</span>
<span class="co">## 2023-05-25 15:25:32  Creating ranged var table</span>
<span class="co">## 2023-05-25 15:25:33  Complete</span>
<span class="co">## rvat gdb object</span>
<span class="co">##  Path: /var/folders/cl/wvc0rvjx4vd5rzt2_fhpmfth0000gp/T//RtmptNGlXA/rvat_tutorials_cmdline.gdb </span>
<span class="co">## </span>
<span class="co">## Finished!</span>
<span class="co">## End time: 2023-05-25 15:25:33</span></code></pre>
</div>
<div class="section level3">
<h3 id="import-sample-metadata">Import sample metadata<a class="anchor" aria-label="anchor" href="#import-sample-metadata"></a>
</h3>
<div class="sourceCode" id="cb8"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="ex">Rscript</span> <span class="va">$rvat</span> <span class="at">--uploadCohort</span> <span class="at">--gdb</span><span class="op">=</span><span class="va">$outdir</span>/rvat_tutorials_cmdline.gdb <span class="at">--name</span><span class="op">=</span>pheno <span class="at">--value</span><span class="op">=</span><span class="va">${phenopath}</span></span></code></pre></div>
<pre><code><span class="co">## Start time: 2023-05-25 15:25:33</span>
<span class="co">## RVAT version: 0.2.0</span>
<span class="co">## RVAT arguments detected:</span>
<span class="co">## --uploadCohort=--uploadCohort</span>
<span class="co">## --gdb=/var/folders/cl/wvc0rvjx4vd5rzt2_fhpmfth0000gp/T//RtmptNGlXA/rvat_tutorials_cmdline.gdb</span>
<span class="co">## --name=pheno</span>
<span class="co">## --value=/Users/phop2/Library/R/3.6/library/rvatData/extdata/rvatData.pheno</span>
<span class="co">## Loading cohort 'pheno' from '/Users/phop2/Library/R/3.6/library/rvatData/extdata/rvatData.pheno'</span>
<span class="co">## </span>
<span class="co">## 10 fields detected (IID,sex,pheno,pop,superPop,PC1,PC2,PC3,PC4,age)</span>
<span class="co">## </span>
<span class="co">## 0 males, 0 females and 25000 unknown gender</span>
<span class="co">## Retained 25000 of 25000 uploaded samples that could be mapped to dosage matrix</span>
<span class="co">## [1] 1</span>
<span class="co">## </span>
<span class="co">## Finished!</span>
<span class="co">## End time: 2023-05-25 15:25:38</span></code></pre>
<p>list imported cohorts:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="ex">Rscript</span> <span class="va">$rvat</span> <span class="at">--listCohort</span> <span class="at">--gdb</span><span class="op">=</span><span class="va">$outdir</span>/rvat_tutorials_cmdline.gdb</span></code></pre></div>
<pre><code><span class="co">## Start time: 2023-05-25 15:25:38</span>
<span class="co">## RVAT version: 0.2.0</span>
<span class="co">## RVAT arguments detected:</span>
<span class="co">## --listCohort=--listCohort</span>
<span class="co">## --gdb=/var/folders/cl/wvc0rvjx4vd5rzt2_fhpmfth0000gp/T//RtmptNGlXA/rvat_tutorials_cmdline.gdb</span>
<span class="co">##    name                                                              value</span>
<span class="co">## 1 pheno /Users/phop2/Library/R/3.6/library/rvatData/extdata/rvatData.pheno</span>
<span class="co">##                       date</span>
<span class="co">## 1 Thu May 25 15:25:38 2023</span>
<span class="co">## </span>
<span class="co">## Finished!</span>
<span class="co">## End time: 2023-05-25 15:25:43</span></code></pre>
</div>
<div class="section level3">
<h3 id="import-variant-annotation-tables">Import variant annotation tables<a class="anchor" aria-label="anchor" href="#import-variant-annotation-tables"></a>
</h3>
<div class="sourceCode" id="cb12"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="ex">Rscript</span> <span class="va">$rvat</span> <span class="at">--uploadAnno</span> <span class="at">--gdb</span><span class="op">=</span><span class="va">$outdir</span>/rvat_tutorials_cmdline.gdb <span class="at">--name</span><span class="op">=</span>varInfo <span class="at">--value</span><span class="op">=</span><span class="va">${varinfopath}</span></span></code></pre></div>
<pre><code><span class="co">## Start time: 2023-05-25 15:25:43</span>
<span class="co">## RVAT version: 0.2.0</span>
<span class="co">## RVAT arguments detected:</span>
<span class="co">## --uploadAnno=--uploadAnno</span>
<span class="co">## --gdb=/var/folders/cl/wvc0rvjx4vd5rzt2_fhpmfth0000gp/T//RtmptNGlXA/rvat_tutorials_cmdline.gdb</span>
<span class="co">## --name=varInfo</span>
<span class="co">## --value=/Users/phop2/Library/R/3.6/library/rvatData/extdata/rvatData.varinfo</span>
<span class="co">## Loading table 'varInfo' from '/Users/phop2/Library/R/3.6/library/rvatData/extdata/rvatData.varinfo'</span>
<span class="co">## </span>
<span class="co">## 20 fields detected (CHROM,POS,ID,REF,ALT,QUAL,FILTER,AC,AN,AF,gene_name,HighImpact,ModerateImpact,Synonymous,aaref,aapos,aaalt,CADDphred,PolyPhen,SIFT)</span>
<span class="co">## </span>
<span class="co">## </span>
<span class="co">## Finished!</span>
<span class="co">## End time: 2023-05-25 15:25:48</span></code></pre>
<p>list anno tables:</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="ex">Rscript</span> <span class="va">$rvat</span> <span class="at">--listAnno</span> <span class="at">--gdb</span><span class="op">=</span><span class="va">$outdir</span>/rvat_tutorials_cmdline.gdb</span></code></pre></div>
<pre><code><span class="co">## Start time: 2023-05-25 15:25:48</span>
<span class="co">## RVAT version: 0.2.0</span>
<span class="co">## RVAT arguments detected:</span>
<span class="co">## --listAnno=--listAnno</span>
<span class="co">## --gdb=/var/folders/cl/wvc0rvjx4vd5rzt2_fhpmfth0000gp/T//RtmptNGlXA/rvat_tutorials_cmdline.gdb</span>
<span class="co">##      name                                                                value</span>
<span class="co">## 1 varInfo /Users/phop2/Library/R/3.6/library/rvatData/extdata/rvatData.varinfo</span>
<span class="co">##                       date</span>
<span class="co">## 1 Thu May 25 15:25:48 2023</span>
<span class="co">## </span>
<span class="co">## Finished!</span>
<span class="co">## End time: 2023-05-25 15:25:53</span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="association-tests">Association tests<a class="anchor" aria-label="anchor" href="#association-tests"></a>
</h2>
<div class="section level3">
<h3 id="build-varsets">Build varSets<a class="anchor" aria-label="anchor" href="#build-varsets"></a>
</h3>
<p>build varSet containing moderate variants:</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="ex">Rscript</span> <span class="va">$rvat</span> <span class="at">--buildVarSet</span> <span class="at">--gdb</span><span class="op">=</span><span class="va">$outdir</span>/rvat_tutorials_cmdline.gdb <span class="at">--varSetName</span><span class="op">=</span>Moderate <span class="at">--unitTable</span><span class="op">=</span>varInfo <span class="at">--unitName</span><span class="op">=</span>gene_name <span class="at">--where</span><span class="op">=</span><span class="st">"ModerateImpact = 1"</span> <span class="at">--output</span><span class="op">=</span><span class="va">${outdir}</span>/moderate_cmdline.txt.gz</span></code></pre></div>
<pre><code><span class="co">## Start time: 2023-05-25 15:25:53</span>
<span class="co">## RVAT version: 0.2.0</span>
<span class="co">## RVAT arguments detected:</span>
<span class="co">## --buildVarSet=--buildVarSet</span>
<span class="co">## --gdb=/var/folders/cl/wvc0rvjx4vd5rzt2_fhpmfth0000gp/T//RtmptNGlXA/rvat_tutorials_cmdline.gdb</span>
<span class="co">## --varSetName=Moderate</span>
<span class="co">## --unitTable=varInfo</span>
<span class="co">## --unitName=gene_name</span>
<span class="co">## --where=ModerateImpact = 1</span>
<span class="co">## --output=/var/folders/cl/wvc0rvjx4vd5rzt2_fhpmfth0000gp/T//RtmptNGlXA/moderate_cmdline.txt.gz</span>
<span class="co">## Generated varSetFile: /var/folders/cl/wvc0rvjx4vd5rzt2_fhpmfth0000gp/T//RtmptNGlXA/moderate_cmdline.txt.gz</span>
<span class="co">## varSetFile object</span>
<span class="co">## Path:/var/folders/cl/wvc0rvjx4vd5rzt2_fhpmfth0000gp/T//RtmptNGlXA/moderate_cmdline.txt.gz</span>
<span class="co">## Units:10</span>
<span class="co">## </span>
<span class="co">## Finished!</span>
<span class="co">## End time: 2023-05-25 15:25:57</span></code></pre>
<p>build multiple varSets</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="va">tables=(</span>HighImpact ModerateImpact CADDphred PolyPhen<span class="va">)</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="va">weights=(</span>1 1 CADDphred 1<span class="va">)</span></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a><span class="va">wheres=(</span><span class="st">"HighImpact = 1"</span> <span class="st">"ModerateImpact = 1"</span> <span class="st">"CADDphred is not 'NA'"</span> <span class="st">"PolyPhen is 'D'"</span><span class="va">)</span></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a><span class="va">n=${#tables</span><span class="op">[@]</span><span class="va">}</span></span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> <span class="kw">((</span><span class="va">i</span><span class="op">=</span><span class="dv">0</span><span class="kw">;</span><span class="va">i</span><span class="op">&lt;</span><span class="va">${n}</span><span class="kw">;</span><span class="va">i</span><span class="op">++</span><span class="kw">))</span></span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>   <span class="cf">do</span></span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a>      <span class="ex">Rscript</span> <span class="va">$rvat</span> <span class="at">--buildVarSet</span> <span class="at">--gdb</span><span class="op">=</span><span class="va">$outdir</span>/rvat_tutorials_cmdline.gdb <span class="at">--unitTable</span><span class="op">=</span>varInfo <span class="at">--unitName</span><span class="op">=</span>gene_name <span class="at">--varSetName</span><span class="op">=</span><span class="va">${tables</span><span class="op">[</span><span class="va">$i</span><span class="op">]</span><span class="va">}</span> <span class="at">--where</span><span class="op">=</span><span class="st">"</span><span class="va">${wheres</span><span class="op">[</span><span class="va">$i</span><span class="op">]</span><span class="va">}</span><span class="st">"</span> <span class="at">--weightName</span><span class="op">=</span><span class="va">${weights</span><span class="op">[</span><span class="va">$i</span><span class="op">]</span><span class="va">}</span> <span class="at">--output</span><span class="op">=</span><span class="va">$outdir</span>/<span class="va">${tables</span><span class="op">[</span><span class="va">$i</span><span class="op">]</span><span class="va">}</span>_cmdline.txt.gz</span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a>   <span class="cf">done</span> <span class="op">&amp;&gt;</span> <span class="va">$outdir</span>/varsets.log</span></code></pre></div>
<p>Merge variant sets for coordinated set based association testing of multiple filters</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="va">varSets=(</span>HighImpact ModerateImpact CADDphred PolyPhen<span class="va">)</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> varSet <span class="kw">in</span> <span class="va">${varSets</span><span class="op">[@]</span><span class="va">}</span></span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>  <span class="cf">do</span></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>     <span class="fu">gunzip</span> <span class="at">-c</span> <span class="va">$outdir</span>/<span class="va">${varSet}</span>_cmdline.txt.gz</span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>  <span class="cf">done</span> <span class="kw">|</span> <span class="fu">sort</span> <span class="at">-t</span> <span class="st">'|'</span> <span class="at">-s</span> <span class="at">-k1,1</span> <span class="kw">|</span> <span class="fu">gzip</span> <span class="op">&gt;</span> <span class="va">$outdir</span>/mergedVarsets_cmdline.txt.gz</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="run-burden-tests">Run burden tests<a class="anchor" aria-label="anchor" href="#run-burden-tests"></a>
</h3>
<div class="sourceCode" id="cb20"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="ex">Rscript</span> <span class="va">$rvat</span> <span class="at">--assocTest</span> <span class="at">--gdb</span><span class="op">=</span><span class="va">$outdir</span>/rvat_tutorials_cmdline.gdb <span class="at">--varSet</span><span class="op">=</span><span class="va">$outdir</span>/mergedVarsets_cmdline.txt.gz <span class="at">--name</span><span class="op">=</span>maxMAF001 <span class="at">--cohort</span><span class="op">=</span>pheno <span class="at">--pheno</span><span class="op">=</span>pheno <span class="at">--test</span><span class="op">=</span>firth,skat_robust,acatv <span class="at">--output</span><span class="op">=</span><span class="va">$outdir</span>/burden_tests_merged_cmdline.txt.gz <span class="op">&amp;&gt;</span> <span class="va">$outdir</span>/burden_tests_merged_cmdline.log</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="run-singlevar-tests">Run singlevar tests<a class="anchor" aria-label="anchor" href="#run-singlevar-tests"></a>
</h3>
<div class="sourceCode" id="cb21"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="ex">Rscript</span> <span class="va">$rvat</span> <span class="at">--assocTest</span> <span class="at">--singlevar</span> <span class="at">--gdb</span><span class="op">=</span><span class="va">$outdir</span>/rvat_tutorials_cmdline.gdb <span class="at">--varSet</span><span class="op">=</span><span class="va">${outdir}</span>/moderate_cmdline.txt.gz <span class="at">--name</span><span class="op">=</span>maxMAF001 <span class="at">--cohort</span><span class="op">=</span>pheno <span class="at">--pheno</span><span class="op">=</span>pheno <span class="at">--test</span><span class="op">=</span>firth <span class="at">--output</span><span class="op">=</span><span class="va">$outdir</span>/singlevar_tests_merged_cmdline.txt.gz <span class="op">&amp;&gt;</span> <span class="va">$outdir</span>/singlevar_tests_merged_cmdline.log</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="todo">todo<a class="anchor" aria-label="anchor" href="#todo"></a>
</h3>
<p>todo: gdbConcat, summariseGeno, etc.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Kevin Kenna, Paul Hop.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
