<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Gene set analysis • rvat</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Gene set analysis">
<meta property="og:description" content="rvat">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rvat</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.3.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/basics.html">Getting started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/basics.html">Getting started</a>
    </li>
    <li>
      <a href="../articles/association_testing.html">Association testing</a>
    </li>
    <li>
      <a href="../articles/explore_visualize_results.html">Visualization</a>
    </li>
    <li>
      <a href="../articles/variant_annotation.html">Variant annotation</a>
    </li>
    <li>
      <a href="../articles/GSA.html">Gene set analysis</a>
    </li>
    <li>
      <a href="../articles/spatial_clustering.html">Spatial clustering</a>
    </li>
    <li>
      <a href="../articles/cmdline.html">Command-line tool</a>
    </li>
  </ul>
</li>
<li>
  <a href="../articles/changelog.html">Changelog</a>
</li>
<li>
  <a href="../articles/faq.html">F.A.Q.</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/KennaLab/rvat" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Gene set analysis</h1>
            
      
      
      <div class="hidden name"><code>GSA.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Libraries</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">rvat</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">rvatData</span><span class="op">)</span></span></code></pre></div>
<p>Some output will be written during the tutorials, please change
<code>outdir</code>, defined below to your preferred directory:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">outdir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="getting-started">Getting started<a class="anchor" aria-label="anchor" href="#getting-started"></a>
</h2>
<p>Load set of example rare variant results</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">rvbresults</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="data-formats">Data formats<a class="anchor" aria-label="anchor" href="#data-formats"></a>
</h2>
<div class="section level3">
<h3 id="genesetlist">geneSetList<a class="anchor" aria-label="anchor" href="#genesetlist"></a>
</h3>
<ul>
<li><p>a <code><a href="../reference/geneSetList.html">geneSetList()</a></code> is a class for storing gene sets (+
metadata) and several methods for working with them.<br></p></li>
<li>
<p>Gene set lists can be generated from multiple sources, currently
the following formats are supported:</p>
<ul>
<li>GMT-format (see [MSigDB] (<a href="https://www.gsea-msigdb.org/gsea/msigdb/" class="external-link uri">https://www.gsea-msigdb.org/gsea/msigdb/</a>).<br>
</li>
<li>data.frame<br>
</li>
<li>list</li>
</ul>
</li>
</ul>
<div class="section level4">
<h4 id="generating-a-genesetlist-object">Generating a geneSetList object<a class="anchor" aria-label="anchor" href="#generating-a-genesetlist-object"></a>
</h4>
<p>The <code><a href="../reference/buildGeneSet.html">buildGeneSet()</a></code> function can be used to load in data
from one of the formats described above. For example, below we build a
genesetlist based on a <a href="https://www.gsea-msigdb.org/gsea/msigdb" class="external-link">MSigDB</a> GMT file
including Gene Ontology (molecular function) terms:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Build a geneSetList from a gmt-file</span></span>
<span><span class="va">genesetlist</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/buildGeneSet.html">buildGeneSet</a></span><span class="op">(</span>gmtpath <span class="op">=</span> <span class="st">"data/c5.go.mf.v2023.2.Hs.symbols.gmt"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Check genesets:</span></span>
<span><span class="va">genesetlist</span></span></code></pre></div>
<pre><code><span><span class="co">## geneSetList</span></span>
<span><span class="co">## Contains 1799 sets</span></span></code></pre>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">genesetlist</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "GOMF_1_4_ALPHA_OLIGOGLUCAN_PHOSPHORYLASE_ACTIVITY"             </span></span>
<span><span class="co">## [2] "GOMF_1_ACYLGLYCEROPHOSPHOCHOLINE_O_ACYLTRANSFERASE_ACTIVITY"   </span></span>
<span><span class="co">## [3] "GOMF_1_ALKYL_2_ACETYLGLYCEROPHOSPHOCHOLINE_ESTERASE_ACTIVITY"  </span></span>
<span><span class="co">## [4] "GOMF_1_PHOSPHATIDYLINOSITOL_3_KINASE_ACTIVITY"                 </span></span>
<span><span class="co">## [5] "GOMF_1_PHOSPHATIDYLINOSITOL_3_KINASE_REGULATOR_ACTIVITY"       </span></span>
<span><span class="co">## [6] "GOMF_1_PHOSPHATIDYLINOSITOL_4_5_BISPHOSPHATE_3_KINASE_ACTIVITY"</span></span></code></pre>
<p>Alternatively, custom genesets can be provided like so:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genesets_custom</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  <span class="st">"ALS_genes"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SOD1"</span>, <span class="st">"NEK1"</span>, <span class="st">"FUS"</span><span class="op">)</span>,</span>
<span>  <span class="st">"PD_genes"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"LRRK2"</span>, <span class="st">"GBA"</span>, <span class="st">"SNCA"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">genesetlist_custom</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/buildGeneSet.html">buildGeneSet</a></span><span class="op">(</span></span>
<span>    <span class="va">genesets_custom</span></span>
<span><span class="op">)</span></span>
<span><span class="va">genesetlist_custom</span></span></code></pre></div>
<pre><code><span><span class="co">## geneSetList</span></span>
<span><span class="co">## Contains 2 sets</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="basic-usage">Basic usage<a class="anchor" aria-label="anchor" href="#basic-usage"></a>
</h4>
<p>Some basic geneSetList methods:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">genesetlist</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 1799</span></span></code></pre>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/lengths.html" class="external-link">lengths</a></span><span class="op">(</span><span class="va">genesetlist</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##              GOMF_1_4_ALPHA_OLIGOGLUCAN_PHOSPHORYLASE_ACTIVITY </span></span>
<span><span class="co">##                                                              5 </span></span>
<span><span class="co">##    GOMF_1_ACYLGLYCEROPHOSPHOCHOLINE_O_ACYLTRANSFERASE_ACTIVITY </span></span>
<span><span class="co">##                                                              8 </span></span>
<span><span class="co">##   GOMF_1_ALKYL_2_ACETYLGLYCEROPHOSPHOCHOLINE_ESTERASE_ACTIVITY </span></span>
<span><span class="co">##                                                              8 </span></span>
<span><span class="co">##                  GOMF_1_PHOSPHATIDYLINOSITOL_3_KINASE_ACTIVITY </span></span>
<span><span class="co">##                                                             10 </span></span>
<span><span class="co">##        GOMF_1_PHOSPHATIDYLINOSITOL_3_KINASE_REGULATOR_ACTIVITY </span></span>
<span><span class="co">##                                                             16 </span></span>
<span><span class="co">## GOMF_1_PHOSPHATIDYLINOSITOL_4_5_BISPHOSPHATE_3_KINASE_ACTIVITY </span></span>
<span><span class="co">##                                                              7</span></span></code></pre>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Annotated-class.html" class="external-link">metadata</a></span><span class="op">(</span><span class="va">genesetlist</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $rvatVersion</span></span>
<span><span class="co">## [1] "0.3.1"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $source</span></span>
<span><span class="co">## [1] "data/c5.go.mf.v2023.2.Hs.symbols.gmt"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $creationDate</span></span>
<span><span class="co">## [1] "2024-11-20 15:05:29"</span></span></code></pre>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genesetlist</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class "geneSet"</span></span>
<span><span class="co">## Slot "geneSetName":</span></span>
<span><span class="co">## [1] "GOMF_1_ACYLGLYCEROPHOSPHOCHOLINE_O_ACYLTRANSFERASE_ACTIVITY"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Slot "units":</span></span>
<span><span class="co">## [1] "LPCAT1,LPCAT2,LPCAT3,LPCAT4,MBOAT1,MBOAT2,PRDX6,TAFAZZIN"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Slot "w":</span></span>
<span><span class="co">## [1] "1,1,1,1,1,1,1,1"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Slot "metadata":</span></span>
<span><span class="co">## [1] "https://www.gsea-msigdb.org/gsea/msigdb/human/geneset/GOMF_1_ACYLGLYCEROPHOSPHOCHOLINE_O_ACYLTRANSFERASE_ACTIVITY"</span></span></code></pre>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genesetlist</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## geneSetList</span></span>
<span><span class="co">## Contains 3 sets</span></span></code></pre>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.list.html" class="external-link">as.list</a></span><span class="op">(</span><span class="va">genesetlist</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $GOMF_1_4_ALPHA_OLIGOGLUCAN_PHOSPHORYLASE_ACTIVITY</span></span>
<span><span class="co">## [1] "MTAP" "PYGB" "PYGL" "PYGM" "TYMP"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $GOMF_1_ACYLGLYCEROPHOSPHOCHOLINE_O_ACYLTRANSFERASE_ACTIVITY</span></span>
<span><span class="co">## [1] "LPCAT1"   "LPCAT2"   "LPCAT3"   "LPCAT4"   "MBOAT1"   "MBOAT2"   "PRDX6"   </span></span>
<span><span class="co">## [8] "TAFAZZIN"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $GOMF_1_ALKYL_2_ACETYLGLYCEROPHOSPHOCHOLINE_ESTERASE_ACTIVITY</span></span>
<span><span class="co">## [1] "ASPG"     "LCAT"     "PAFAH1B2" "PAFAH1B3" "PAFAH2"   "PLA2G10"  "PLA2G6"  </span></span>
<span><span class="co">## [8] "PLA2G7"  </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $GOMF_1_PHOSPHATIDYLINOSITOL_3_KINASE_ACTIVITY</span></span>
<span><span class="co">##  [1] "ATM"     "PIK3C2A" "PIK3C2B" "PIK3C2G" "PIK3C3"  "PIK3CA"  "PIK3CB" </span></span>
<span><span class="co">##  [8] "PIK3CD"  "PIK3CG"  "PIK3R3" </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $GOMF_1_PHOSPHATIDYLINOSITOL_3_KINASE_REGULATOR_ACTIVITY</span></span>
<span><span class="co">##  [1] "CCKBR"   "CISH"    "P3R3URF" "PIK3R1"  "PIK3R2"  "PIK3R3"  "PIK3R5" </span></span>
<span><span class="co">##  [8] "PIK3R6"  "SLA2"    "SOCS1"   "SOCS2"   "SOCS3"   "SOCS4"   "SOCS5"  </span></span>
<span><span class="co">## [15] "SOCS6"   "SOCS7"  </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $GOMF_1_PHOSPHATIDYLINOSITOL_4_5_BISPHOSPHATE_3_KINASE_ACTIVITY</span></span>
<span><span class="co">## [1] "IPMK"    "PIK3C2A" "PIK3CA"  "PIK3CB"  "PIK3CD"  "PIK3CG"  "PIK3R6"</span></span></code></pre>
<p>The <code><a href="../reference/getGeneSet.html">getGeneSet()</a></code> method can be used to extract one or
more genesets from a <code>geneSetList:</code></p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Extract one specific geneset:</span></span>
<span><span class="va">geneset</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getGeneSet.html">getGeneSet</a></span><span class="op">(</span><span class="va">genesetlist</span>, </span>
<span>                      geneSet <span class="op">=</span> <span class="st">"GOMF_E_BOX_BINDING"</span><span class="op">)</span></span>
<span><span class="va">geneset</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class "geneSet"</span></span>
<span><span class="co">## Slot "geneSetName":</span></span>
<span><span class="co">## [1] "GOMF_E_BOX_BINDING"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Slot "units":</span></span>
<span><span class="co">## [1] "AHR,ASCL1,ASCL2,ATOH1,ATOH7,ATOH8,BHLHA15,BHLHE22,BHLHE23,BHLHE40,BHLHE41,BMAL1,BMAL2,CIART,CLOCK,CRY1,FIGLA,GATA3,HAND2,HDAC1,HES1,HIF1A,MAX,MITF,MYBBP1A,MYC,MYF5,MYF6,MYOD1,MYOG,NEUROD1,NEUROD2,NEUROD4,NEUROD6,NEUROG1,NEUROG2,NEUROG3,NR1D1,OLIG1,OLIG2,OLIG3,PER1,PPARG,PRMT5,PTF1A,SCRT2,SCX,SNAI1,SNAI2,SREBF2,TAL1,TCF12,TCF15,TCF21,TCF3,TCF4,TFAP4,TWIST1,ZEB1"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Slot "w":</span></span>
<span><span class="co">## [1] "1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Slot "metadata":</span></span>
<span><span class="co">## [1] "https://www.gsea-msigdb.org/gsea/msigdb/human/geneset/GOMF_E_BOX_BINDING"</span></span></code></pre>
<p>It can also be used to retrieve gene sets that contain one or more
genes of interest:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Extract all genesets that contain SOD1:</span></span>
<span><span class="va">geneset</span> <span class="op">=</span> <span class="fu"><a href="../reference/getGeneSet.html">getGeneSet</a></span><span class="op">(</span><span class="va">genesetlist</span>, unit <span class="op">=</span> <span class="st">"SOD1"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">geneset</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "GOMF_ANTIOXIDANT_ACTIVITY"             </span></span>
<span><span class="co">## [2] "GOMF_COPPER_ION_BINDING"               </span></span>
<span><span class="co">## [3] "GOMF_GTPASE_BINDING"                   </span></span>
<span><span class="co">## [4] "GOMF_OXIDOREDUCTASE_ACTIVITY"          </span></span>
<span><span class="co">## [5] "GOMF_PHOSPHATASE_BINDING"              </span></span>
<span><span class="co">## [6] "GOMF_PROTEIN_FOLDING_CHAPERONE_BINDING"</span></span></code></pre>
<p>Specific units/genes can also be dropped from the
<code><a href="../reference/geneSetList.html">geneSetList()</a></code>, which is useful if you want to exclude top
genes for gene set analysis for example:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genesetlist_noSOD1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/geneSetList.html">dropUnits</a></span><span class="op">(</span><span class="va">genesetlist</span>, unit <span class="op">=</span> <span class="st">"SOD1"</span><span class="op">)</span></span></code></pre></div>
<p>Convert <code><a href="../reference/geneSetList.html">geneSetList()</a></code> to a data.frame or a list:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.list.html" class="external-link">as.list</a></span><span class="op">(</span><span class="va">genesetlist</span><span class="op">)</span>, <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $GOMF_1_4_ALPHA_OLIGOGLUCAN_PHOSPHORYLASE_ACTIVITY</span></span>
<span><span class="co">## [1] "MTAP" "PYGB" "PYGL" "PYGM" "TYMP"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $GOMF_1_ACYLGLYCEROPHOSPHOCHOLINE_O_ACYLTRANSFERASE_ACTIVITY</span></span>
<span><span class="co">## [1] "LPCAT1"   "LPCAT2"   "LPCAT3"   "LPCAT4"   "MBOAT1"   "MBOAT2"   "PRDX6"   </span></span>
<span><span class="co">## [8] "TAFAZZIN"</span></span></code></pre>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">genesetlist</span><span class="op">)</span>, <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                                                    geneSetName</span></span>
<span><span class="co">## 1            GOMF_1_4_ALPHA_OLIGOGLUCAN_PHOSPHORYLASE_ACTIVITY</span></span>
<span><span class="co">## 2  GOMF_1_ACYLGLYCEROPHOSPHOCHOLINE_O_ACYLTRANSFERASE_ACTIVITY</span></span>
<span><span class="co">## 3 GOMF_1_ALKYL_2_ACETYLGLYCEROPHOSPHOCHOLINE_ESTERASE_ACTIVITY</span></span>
<span><span class="co">##                                                      units               w</span></span>
<span><span class="co">## 1                                 MTAP,PYGB,PYGL,PYGM,TYMP       1,1,1,1,1</span></span>
<span><span class="co">## 2 LPCAT1,LPCAT2,LPCAT3,LPCAT4,MBOAT1,MBOAT2,PRDX6,TAFAZZIN 1,1,1,1,1,1,1,1</span></span>
<span><span class="co">## 3 ASPG,LCAT,PAFAH1B2,PAFAH1B3,PAFAH2,PLA2G10,PLA2G6,PLA2G7 1,1,1,1,1,1,1,1</span></span>
<span><span class="co">##                                                                                                             metadata</span></span>
<span><span class="co">## 1            https://www.gsea-msigdb.org/gsea/msigdb/human/geneset/GOMF_1_4_ALPHA_OLIGOGLUCAN_PHOSPHORYLASE_ACTIVITY</span></span>
<span><span class="co">## 2  https://www.gsea-msigdb.org/gsea/msigdb/human/geneset/GOMF_1_ACYLGLYCEROPHOSPHOCHOLINE_O_ACYLTRANSFERASE_ACTIVITY</span></span>
<span><span class="co">## 3 https://www.gsea-msigdb.org/gsea/msigdb/human/geneset/GOMF_1_ALKYL_2_ACETYLGLYCEROPHOSPHOCHOLINE_ESTERASE_ACTIVITY</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="remapping-ids">Remapping IDs<a class="anchor" aria-label="anchor" href="#remapping-ids"></a>
</h4>
<p>The IDs in a <code><a href="../reference/geneSetList.html">geneSetList()</a></code> can be remapped using the
<code><a href="../reference/geneSetList.html">remapIDs()</a></code> method to match those used in the analyses.</p>
<p>Below we remap the gene symbols to Ensembl IDs, based on a
linker-file.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Linker file mapping Entrez gene IDs to Ensembl IDs:</span></span>
<span><span class="va">linker</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_tsv</a></span><span class="op">(</span><span class="st">"data/Homo_sapiens.GRCh38.105.gene.txt.gz"</span>, show_col_types <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">linker</span> <span class="op">&lt;-</span> <span class="va">linker</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gene_id"</span>, <span class="st">"gene_name"</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># note: the `duplicated_ids` parameter indicates what to do with IDs that map to multiple IDs </span></span>
<span><span class="va">genesetlist_remapped</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/geneSetList.html">remapIDs</a></span><span class="op">(</span><span class="va">genesetlist</span>, </span>
<span>                                 dict <span class="op">=</span> <span class="va">linker</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">1</span><span class="op">)</span><span class="op">]</span>, </span>
<span>                                 targets <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">rvbresults</span><span class="op">$</span><span class="va">unit</span><span class="op">)</span>, </span>
<span>                                 duplicate_ids <span class="op">=</span> <span class="st">"keep_first"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 15574/15682 IDs in the geneSetList are present in the linker file.</span></span></code></pre>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genesetlist_remapped</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class "geneSet"</span></span>
<span><span class="co">## Slot "geneSetName":</span></span>
<span><span class="co">## [1] "GOMF_1_4_ALPHA_OLIGOGLUCAN_PHOSPHORYLASE_ACTIVITY"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Slot "units":</span></span>
<span><span class="co">## [1] "ENSG00000099810,ENSG00000100994,ENSG00000100504,ENSG00000068976,ENSG00000025708"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Slot "w":</span></span>
<span><span class="co">## [1] "1,1,1,1,1"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Slot "metadata":</span></span>
<span><span class="co">## [1] "https://www.gsea-msigdb.org/gsea/msigdb/human/geneset/GOMF_1_4_ALPHA_OLIGOGLUCAN_PHOSPHORYLASE_ACTIVITY"</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="genesetfile">geneSetFile<a class="anchor" aria-label="anchor" href="#genesetfile"></a>
</h3>
<p>We have also implemented a gene set file structure
(<code><a href="../reference/geneSetFile.html">geneSetFile()</a></code>) akin to a <code><a href="../reference/varSetFile.html">varSetFile()</a></code>, from
which gene sets can be loaded without loading the entire file to memory.
The <code><a href="../reference/buildGeneSet.html">buildGeneSet()</a></code> function can convert a file in one of
the formats mentioned in <a href="#genesetlist">geneSetList</a>. Below
we build a GO and KEGG <code><a href="../reference/geneSetFile.html">geneSetFile()</a></code>:</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># build a genesetfile</span></span>
<span><span class="fu"><a href="../reference/buildGeneSet.html">buildGeneSet</a></span><span class="op">(</span>gmtpath <span class="op">=</span> <span class="st">"data/c5.go.mf.v2023.2.Hs.symbols.gmt"</span>,</span>
<span>            output <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"/genesetfile.txt.gz"</span><span class="op">)</span></span>
<span>            <span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># connect </span></span>
<span><span class="va">genesetfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/geneSetFile.html">geneSetFile</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"/genesetfile.txt.gz"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">genesetfile</span></span></code></pre></div>
<pre><code><span><span class="co">## geneSetFile</span></span>
<span><span class="co">## Path: /var/folders/cl/wvc0rvjx4vd5rzt2_fhpmfth0000gp/T//RtmpMoc1RL/genesetfile.txt.gz</span></span>
<span><span class="co">## Sets: 1799</span></span></code></pre>
<p>Extract a gene set:</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># extract a geneset from a genesetfile</span></span>
<span><span class="va">genesetlist</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getGeneSet.html">getGeneSet</a></span><span class="op">(</span><span class="va">genesetfile</span>, </span>
<span>                          geneSet <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GOMF_UBIQUITIN_LIKE_MODIFIER_ACTIVATING_ENZYME_ACTIVITY"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/listUnits.html">listUnits</a></span><span class="op">(</span><span class="va">genesetlist</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "ATG7"  "MOCS3" "NAE1"  "SAE1"  "UBA1"  "UBA2"  "UBA3"  "UBA5"  "UBA6" </span></span>
<span><span class="co">## [10] "UBA7"</span></span></code></pre>
<p>Convert to a <code><a href="../reference/geneSetList.html">geneSetList()</a></code> (i.e. load all sets):</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genesetlist</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/geneSetFile.html">as.geneSetList</a></span><span class="op">(</span><span class="va">genesetfile</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="gene-set-analyses">Gene-set analyses<a class="anchor" aria-label="anchor" href="#gene-set-analyses"></a>
</h2>
<p>Broadly, the GSA methods implemented in RVAT can be divided into
competitive and self-contained tests, where the former tests whether
genes in the gene set are more associated with the phenotype than genes
outside the gene set, while the latter jointly tests whether genes in
the gene set are associated with the phenotype without considering genes
outside the set <span class="citation">(de Leeuw et al. 2016)</span>.
Because competitive GSA tests for an enrichment relative to the genes
outside the gene set, it controls for polygenicity as well as biases
such as confounding and technical variability. Self-contained tests, on
the other hand, are generally more powerful but may result in inflated
test-statistics in case of polygenicity or residual biases in the data.
See <span class="citation">(de Leeuw et al. 2016)</span> for a brilliant
review of GSA methods.</p>
<div class="section level3">
<h3 id="competitive-gsa">Competitive GSA<a class="anchor" aria-label="anchor" href="#competitive-gsa"></a>
</h3>
<ul>
<li>
<p>Currently the following competitive methods are implemented:</p>
<ul>
<li><p><strong>linear model (test = “lm”)</strong>: tests whether gene
set membership predicts higher gene-association scores. Allows for
adjustment for covariates such as gene size and number of variants in a
gene.</p></li>
<li><p><strong>hypergeometric test (test = “fisher”)</strong>: tests
whether the proportion of <em>P</em>-values below the specified
threshold is greater than the proportion outside of it.</p></li>
</ul>
</li>
</ul>
<p>In below example, we perform competitive gene set analysis using a
linear model on gene burden results:</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Adjust for the number of variants in a gene</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="va">rvbresults</span><span class="op">[</span><span class="va">rvbresults</span><span class="op">$</span><span class="va">test</span><span class="op">==</span><span class="st">"firth"</span> <span class="op">&amp;</span> </span>
<span>                    <span class="va">rvbresults</span><span class="op">$</span><span class="va">varSetName</span> <span class="op">==</span> <span class="st">"ModerateImpact"</span>, <span class="op">]</span></span>
<span><span class="va">GSAresults</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/geneSetAssoc.html">geneSetAssoc</a></span><span class="op">(</span></span>
<span>  <span class="va">res</span>,</span>
<span>  <span class="va">genesetlist</span>,</span>
<span>  covar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nvar"</span><span class="op">)</span>,</span>
<span>  test <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"lm"</span><span class="op">)</span>, </span>
<span>  minSetSize <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  maxSetSize <span class="op">=</span> <span class="fl">500</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/rvatResult.html">topResult</a></span><span class="op">(</span><span class="va">GSAresults</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## gsaResult with 10 rows and 11 columns</span></span>
<span><span class="co">##               geneSetName  test covar threshold geneSetSize  genesObs    effect</span></span>
<span><span class="co">##                     &lt;Rle&gt; &lt;Rle&gt; &lt;Rle&gt; &lt;numeric&gt;   &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">## 1  GOMF_MRNA_METHYLTRAN..    lm  nvar        NA          13        13  0.878627</span></span>
<span><span class="co">## 2        GOMF_BMP_BINDING    lm  nvar        NA          20        19  0.537316</span></span>
<span><span class="co">## 3  GOMF_DYNEIN_LIGHT_IN..    lm  nvar        NA          28        28  0.419752</span></span>
<span><span class="co">## 4  GOMF_DYNEIN_HEAVY_CH..    lm  nvar        NA          16        16  0.553052</span></span>
<span><span class="co">## 5  GOMF_PROTEIN_KINASE_..    lm  nvar        NA          25        24  0.436416</span></span>
<span><span class="co">## 6  GOMF_TRANSFERRIN_REC..    lm  nvar        NA          11        10  0.664977</span></span>
<span><span class="co">## 7  GOMF_PHOSPHATIDYLCHO..    lm  nvar        NA          31        28  0.388610</span></span>
<span><span class="co">## 8  GOMF_MONOATOMIC_ANIO..    lm  nvar        NA          95        87  0.213296</span></span>
<span><span class="co">## 9  GOMF_LIGASE_ACTIVITY..    lm  nvar        NA          41        41  0.305862</span></span>
<span><span class="co">## 10 GOMF_COA_HYDROLASE_A..    lm  nvar        NA          24        21  0.424778</span></span>
<span><span class="co">##     effectSE effectCIlower effectCIupper          P</span></span>
<span><span class="co">##    &lt;numeric&gt;     &lt;numeric&gt;     &lt;numeric&gt;  &lt;numeric&gt;</span></span>
<span><span class="co">## 1   0.276810     0.4232922           Inf 0.00075271</span></span>
<span><span class="co">## 2   0.229032     0.1605722           Inf 0.00949267</span></span>
<span><span class="co">## 3   0.188903     0.1090181           Inf 0.01314612</span></span>
<span><span class="co">## 4   0.249569     0.1425262           Inf 0.01335099</span></span>
<span><span class="co">## 5   0.203871     0.1010619           Inf 0.01615762</span></span>
<span><span class="co">## 6   0.315631     0.1457836           Inf 0.01757354</span></span>
<span><span class="co">## 7   0.188720     0.0781772           Inf 0.01974526</span></span>
<span><span class="co">## 8   0.107252     0.0368734           Inf 0.02337241</span></span>
<span><span class="co">## 9   0.156014     0.0492282           Inf 0.02497780</span></span>
<span><span class="co">## 10  0.217878     0.0663827           Inf 0.02561858</span></span></code></pre>
<p>The gene association scores can be visualized using the
<code>densityPlot</code> method, which compares the Z-scores for a given
gene set with the background, this can be useful to inspect whether the
association is driven by few genes or not:</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/densityPlot.html">densityPlot</a></span><span class="op">(</span><span class="va">res</span>, </span>
<span>            geneSet <span class="op">=</span> <span class="st">"GOMF_MRNA_METHYLTRANSFERASE_ACTIVITY"</span>, </span>
<span>            geneSetList <span class="op">=</span> <span class="va">genesetlist</span><span class="op">)</span></span></code></pre></div>
<p><img src="GSA_files/figure-html/unnamed-chunk-17-1.png" width="700"></p>
<p>Outlying gene association scores can be remedied by either setting
Z-score cutoffs (i.e. all Z-scores exceeding these values will be set to
the respective cutoff), or inverse normal transforming the Z-scores:</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GSAresults_Zcutoffs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/geneSetAssoc.html">geneSetAssoc</a></span><span class="op">(</span></span>
<span>  <span class="va">res</span>,</span>
<span>  <span class="va">genesetlist</span>,</span>
<span>  covar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nvar"</span><span class="op">)</span>,</span>
<span>  test <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"lm"</span><span class="op">)</span>, </span>
<span>  minSetSize <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  maxSetSize <span class="op">=</span> <span class="fl">500</span>,</span>
<span>  Zcutoffs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">4</span>, <span class="fl">4</span><span class="op">)</span> <span class="co"># lower and upper bounds</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">GSAresults_INT</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/geneSetAssoc.html">geneSetAssoc</a></span><span class="op">(</span></span>
<span>  <span class="va">res</span>,</span>
<span>  <span class="va">genesetlist</span>,</span>
<span>  covar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nvar"</span><span class="op">)</span>,</span>
<span>  test <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"lm"</span><span class="op">)</span>, </span>
<span>  minSetSize <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  maxSetSize <span class="op">=</span> <span class="fl">500</span>,</span>
<span>  INT <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Conditional gene set analyses can be performed to testher whether
gene sets are associated independently with the phenotype of interest.
In the example below we condition the most significant geneset on the
top 10 most significant genesets. The result includes a
<code>condition</code> field, indicating for which gene set the given
was adjusted:</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">topresults</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rvatResult.html">topResult</a></span><span class="op">(</span><span class="va">GSAresults</span><span class="op">)</span></span>
<span><span class="va">GSAresults_conditional</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/geneSetAssoc.html">geneSetAssoc</a></span><span class="op">(</span></span>
<span>  <span class="va">res</span>,</span>
<span>  <span class="fu"><a href="../reference/getGeneSet.html">getGeneSet</a></span><span class="op">(</span><span class="va">genesetlist</span>, <span class="va">topresults</span><span class="op">$</span><span class="va">geneSetName</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span>,</span>
<span>  condition <span class="op">=</span> <span class="fu"><a href="../reference/getGeneSet.html">getGeneSet</a></span><span class="op">(</span><span class="va">genesetlist</span>, <span class="va">topresults</span><span class="op">$</span><span class="va">geneSetName</span><span class="op">[</span><span class="fl">2</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span><span class="op">)</span>,</span>
<span>  covar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nvar"</span><span class="op">)</span>,</span>
<span>  test <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"lm"</span><span class="op">)</span>, </span>
<span>  minSetSize <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  maxSetSize <span class="op">=</span> <span class="fl">500</span></span>
<span><span class="op">)</span></span>
<span><span class="va">GSAresults_conditional</span></span></code></pre></div>
<pre><code><span><span class="co">## gsaResult with 9 rows and 12 columns</span></span>
<span><span class="co">##              geneSetName  test covar threshold geneSetSize  genesObs    effect</span></span>
<span><span class="co">##                    &lt;Rle&gt; &lt;Rle&gt; &lt;Rle&gt; &lt;numeric&gt;   &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">## 1 GOMF_MRNA_METHYLTRAN..    lm  nvar        NA          13        13  0.879184</span></span>
<span><span class="co">## 2 GOMF_MRNA_METHYLTRAN..    lm  nvar        NA          13        13  0.879127</span></span>
<span><span class="co">## 3 GOMF_MRNA_METHYLTRAN..    lm  nvar        NA          13        13  0.879093</span></span>
<span><span class="co">## 4 GOMF_MRNA_METHYLTRAN..    lm  nvar        NA          13        13  0.879132</span></span>
<span><span class="co">## 5 GOMF_MRNA_METHYLTRAN..    lm  nvar        NA          13        13  0.879321</span></span>
<span><span class="co">## 6 GOMF_MRNA_METHYLTRAN..    lm  nvar        NA          13        13  0.879593</span></span>
<span><span class="co">## 7 GOMF_MRNA_METHYLTRAN..    lm  nvar        NA          13        13  0.879231</span></span>
<span><span class="co">## 8 GOMF_MRNA_METHYLTRAN..    lm  nvar        NA          13        13  0.879130</span></span>
<span><span class="co">## 9 GOMF_MRNA_METHYLTRAN..    lm  nvar        NA          13        13  0.878994</span></span>
<span><span class="co">##    effectSE effectCIlower effectCIupper           P              condition</span></span>
<span><span class="co">##   &lt;numeric&gt;     &lt;numeric&gt;     &lt;numeric&gt;   &lt;numeric&gt;            &lt;character&gt;</span></span>
<span><span class="co">## 1  0.276776      0.423905           Inf 0.000746510       GOMF_BMP_BINDING</span></span>
<span><span class="co">## 2  0.276789      0.423826           Inf 0.000747425 GOMF_COA_HYDROLASE_A..</span></span>
<span><span class="co">## 3  0.276781      0.423806           Inf 0.000747489 GOMF_DYNEIN_HEAVY_CH..</span></span>
<span><span class="co">## 4  0.276781      0.423846           Inf 0.000747119 GOMF_DYNEIN_LIGHT_IN..</span></span>
<span><span class="co">## 5  0.276789      0.424021           Inf 0.000745611 GOMF_LIGASE_ACTIVITY..</span></span>
<span><span class="co">## 6  0.276788      0.424294           Inf 0.000743074 GOMF_MONOATOMIC_ANIO..</span></span>
<span><span class="co">## 7  0.276786      0.423936           Inf 0.000746358 GOMF_PHOSPHATIDYLCHO..</span></span>
<span><span class="co">## 8  0.276783      0.423839           Inf 0.000747218 GOMF_PROTEIN_KINASE_..</span></span>
<span><span class="co">## 9  0.276784      0.423702           Inf 0.000748516 GOMF_TRANSFERRIN_REC..</span></span></code></pre>
<p>Finally, instead of the linear model, a hypergeometric test (test =
“fisher”), can be performed if you’d like to tests whether the
proportion of <em>P</em>-values below a specified threshold is greater
than the proportion outside of it. The <code>threshold</code> parameter
specifies the <em>P</em>-value cutoff to define significant genes:</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GSAresults_fisher</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/geneSetAssoc.html">geneSetAssoc</a></span><span class="op">(</span></span>
<span>  <span class="va">res</span>,</span>
<span>  <span class="va">genesetlist</span>,</span>
<span>  test <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"fisher"</span><span class="op">)</span>, </span>
<span>  threshold <span class="op">=</span> <span class="fl">1e-4</span>,</span>
<span>  minSetSize <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  maxSetSize <span class="op">=</span> <span class="fl">500</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/rvatResult.html">topResult</a></span><span class="op">(</span><span class="va">GSAresults_fisher</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## gsaResult with 10 rows and 11 columns</span></span>
<span><span class="co">##               geneSetName   test covar threshold geneSetSize  genesObs</span></span>
<span><span class="co">##                     &lt;Rle&gt;  &lt;Rle&gt; &lt;Rle&gt; &lt;numeric&gt;   &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">## 1  GOMF_EXTRACELLULAR_M.. fisher    NA     1e-04          10        10</span></span>
<span><span class="co">## 2  GOMF_STRUCTURAL_MOLE.. fisher    NA     1e-04          12        12</span></span>
<span><span class="co">## 3  GOMF_K63_LINKED_POLY.. fisher    NA     1e-04          26        24</span></span>
<span><span class="co">## 4  GOMF_POLYUBIQUITIN_M.. fisher    NA     1e-04          57        52</span></span>
<span><span class="co">## 5  GOMF_COPPER_ION_BIND.. fisher    NA     1e-04          63        56</span></span>
<span><span class="co">## 6  GOMF_CELL_ADHESION_M.. fisher    NA     1e-04          65        63</span></span>
<span><span class="co">## 7  GOMF_ANTIOXIDANT_ACT.. fisher    NA     1e-04          83        80</span></span>
<span><span class="co">## 8  GOMF_ATP_DEPENDENT_A.. fisher    NA     1e-04         122       112</span></span>
<span><span class="co">## 9  GOMF_PROTEIN_FOLDING.. fisher    NA     1e-04         133       126</span></span>
<span><span class="co">## 10 GOMF_PROTEIN_PHOSPHA.. fisher    NA     1e-04         142       138</span></span>
<span><span class="co">##       effect  effectSE effectCIlower effectCIupper          P</span></span>
<span><span class="co">##    &lt;numeric&gt; &lt;numeric&gt;     &lt;numeric&gt;     &lt;numeric&gt;  &lt;numeric&gt;</span></span>
<span><span class="co">## 1   507.6025        NA      18.98856           Inf 0.00271695</span></span>
<span><span class="co">## 2   411.5113        NA      15.81130           Inf 0.00325963</span></span>
<span><span class="co">## 3   198.1263        NA       7.89375           Inf 0.00651075</span></span>
<span><span class="co">## 4    89.4421        NA       3.63563           Inf 0.01406371</span></span>
<span><span class="co">## 5    82.9262        NA       3.37545           Inf 0.01513895</span></span>
<span><span class="co">## 6    73.7325        NA       2.99909           Inf 0.01701835</span></span>
<span><span class="co">## 7    57.7373        NA       2.35945           Inf 0.02157067</span></span>
<span><span class="co">## 8    41.0502        NA       1.68226           Inf 0.03009399</span></span>
<span><span class="co">## 9    36.4539        NA       1.49414           Inf 0.03380421</span></span>
<span><span class="co">## 10   33.2467        NA       1.36333           Inf 0.03697535</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="self-contained-gsa">Self-contained GSA<a class="anchor" aria-label="anchor" href="#self-contained-gsa"></a>
</h3>
<p>Self-contained gene set analyses have been implemented in the form of
‘gene set burden’ analysis.Gene set burden analyses extend the rationale
behind gene burden tests to sets of genes, aggregating variants across
gene sets rather than single genes. Since this follows the same
principle as gene burden testing, the same approach can be used as
described in <code><a href="../articles/association_testing.html">vignette("association_testing")</a></code>. However,
these types of analyses can be computationally demanding since the
number of variants to aggregate, especially for large gene sets, is
often many times larger than in single gene analyses. Therefore, RVAT
also includes an alternative two-step approach. In the first step,
burden scores are generated for each gene and stored in a compressed
format. In the second step, gene set analyses are performed by
aggregating the gene burden scores for each gene set, followed by
testing for an association between the gene set burden score and the
phenotype of interest. An example of this approach is shown below.</p>
<div class="section level4">
<h4 id="step-1-generate-burden-scores-per-gene">Step 1: generate burden scores per gene<a class="anchor" aria-label="anchor" href="#step-1-generate-burden-scores-per-gene"></a>
</h4>
<p>First, we’ll generate a varsetfile, including variants that we want
to include in the gene burden scores:</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gdb</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gdb.html">gdb</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/rvatData/man/rvat_example.html" class="external-link">rvat_example</a></span><span class="op">(</span><span class="st">"rvatData.gdb"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/buildVarSet.html">buildVarSet</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">gdb</span>, </span>
<span>            output <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"/moderate.txt.gz"</span><span class="op">)</span>,</span>
<span>            varSetName <span class="op">=</span> <span class="st">"Moderate"</span>, </span>
<span>            unitTable <span class="op">=</span> <span class="st">"varInfo"</span>, </span>
<span>            unitName <span class="op">=</span> <span class="st">"gene_name"</span>,</span>
<span>            where <span class="op">=</span> <span class="st">"ModerateImpact = 1"</span><span class="op">)</span></span></code></pre></div>
<p>Second, burden scores are generated per gene using the
<code>aggregate</code> method:</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/aggregate.html" class="external-link">aggregate</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">gdb</span>, </span>
<span>          varSet<span class="op">=</span><span class="fu"><a href="../reference/varSetFile.html">varSetFile</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"/moderate.txt.gz"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>          maxMAF<span class="op">=</span><span class="fl">0.001</span>,</span>
<span>          output<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"/moderate.aggregate.txt.gz"</span><span class="op">)</span>,</span>
<span>          verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>          <span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="step-2-run-gene-set-burden-tests">Step 2: run gene set burden tests<a class="anchor" aria-label="anchor" href="#step-2-run-gene-set-burden-tests"></a>
</h4>
<p>Note that the example gdb we’re using contains only a small number of
genes, therefore most of the genes included in GO, KEGG etc. are missing
from this gdb. Therefore, for this example, I’ll just create some dummy
genesets that contain genes which are present in the gdb:</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">varsetfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/varSetFile.html">varSetFile</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"/moderate.txt.gz"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">genesetlist</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/buildGeneSet.html">buildGeneSet</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"geneset1"</span> <span class="op">=</span> <span class="fu"><a href="../reference/listUnits.html">listUnits</a></span><span class="op">(</span><span class="va">varsetfile</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">3</span>,<span class="fl">4</span>,<span class="fl">9</span><span class="op">)</span><span class="op">]</span>,</span>
<span>       <span class="st">"geneset2"</span> <span class="op">=</span> <span class="fu"><a href="../reference/listUnits.html">listUnits</a></span><span class="op">(</span><span class="va">varsetfile</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">4</span>,<span class="fl">5</span>,<span class="fl">6</span>,<span class="fl">10</span><span class="op">)</span><span class="op">]</span>,</span>
<span>       <span class="st">"geneset3"</span> <span class="op">=</span> <span class="fu"><a href="../reference/listUnits.html">listUnits</a></span><span class="op">(</span><span class="va">varsetfile</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span><span class="op">]</span>,</span>
<span>       <span class="st">"geneset4"</span> <span class="op">=</span> <span class="fu"><a href="../reference/listUnits.html">listUnits</a></span><span class="op">(</span><span class="va">varsetfile</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">7</span>,<span class="fl">9</span>,<span class="fl">10</span><span class="op">)</span><span class="op">]</span>,</span>
<span>       <span class="st">"geneset5"</span> <span class="op">=</span> <span class="fu"><a href="../reference/listUnits.html">listUnits</a></span><span class="op">(</span><span class="va">varsetfile</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/write.html" class="external-link">write</a></span><span class="op">(</span><span class="va">genesetlist</span>,</span>
<span>      file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"/genesetfile.txt.gz"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Perform the gene set burden tests, for this we can use the
<code>assocTest</code> method, with the first argument being a
<code>aggregateFile</code> connection. Similar to a
<code>varSetFile</code> or <code>geneSetFile</code> we can connect to
the file containing aggregates (burden scores) using the
<code>aggregateFile</code> method:</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">aggfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/aggregateFile.html">aggregateFile</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"/moderate.aggregate.txt.gz"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>We can then run gene set burden tests using the
<code>aggregateFile</code>, by passing it as the first argument to
<code>assocTest</code> as shown below. Note that, in contrast to running
<code>assocTest</code> directly on a genoMatrix or gdb, we cannot
perform any variant filters when running <code>assocTest</code> on an
<code>aggregateFile</code>. Any variant filters should be applied when
generating the aggregates, as shown in step 1.</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">aggAssoc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/assocTest.html">assocTest</a></span><span class="op">(</span></span>
<span>  <span class="va">aggfile</span>,</span>
<span>  gdb <span class="op">=</span> <span class="va">gdb</span>,</span>
<span>  test <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"glm"</span>, <span class="st">"firth"</span><span class="op">)</span>,</span>
<span>  cohort <span class="op">=</span> <span class="st">"pheno"</span>,</span>
<span>  pheno<span class="op">=</span><span class="st">"pheno"</span>,</span>
<span>  geneSet <span class="op">=</span> <span class="va">genesetlist</span>,</span>
<span>  covar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"PC"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">)</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>As noted above, the same results can be obtained by running
<code>assocTest</code> directing on a <code>gdb</code> or
<code>genoMatrix</code> object, as shown below. However, the two-step
approach using is way more more efficient when running large-scale
analyses (i.e. ~20,000 genes).</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">varsetfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/varSetFile.html">varSetFile</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"/moderate.txt.gz"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">units</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/listUnits.html">listUnits</a></span><span class="op">(</span><span class="fu"><a href="../reference/getGeneSet.html">getGeneSet</a></span><span class="op">(</span><span class="va">genesetlist</span>, <span class="st">"geneset1"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">test</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/assocTest.html">assocTest</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">gdb</span>,</span>
<span>  test <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"glm"</span>, <span class="st">"firth"</span><span class="op">)</span>,</span>
<span>  varSet <span class="op">=</span> <span class="fu"><a href="../reference/varSetList.html">collapseVarSetList</a></span><span class="op">(</span><span class="fu"><a href="../reference/getVarSet.html">getVarSet</a></span><span class="op">(</span><span class="va">varsetfile</span>, unit <span class="op">=</span> <span class="va">units</span><span class="op">)</span>, drop <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>, <span class="co"># collapse variants across genes into one varSet</span></span>
<span>  cohort <span class="op">=</span> <span class="st">"pheno"</span>,</span>
<span>  pheno<span class="op">=</span><span class="st">"pheno"</span>,</span>
<span>  maxMAF<span class="op">=</span><span class="fl">0.001</span>,</span>
<span>  covar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"PC"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">)</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">test</span></span></code></pre></div>
<pre><code><span><span class="co">## rvbResult with 2 rows and 24 columns</span></span>
<span><span class="co">##          unit cohort varSetName  name pheno           covar geneticModel</span></span>
<span><span class="co">##   &lt;character&gt;  &lt;Rle&gt;      &lt;Rle&gt; &lt;Rle&gt; &lt;Rle&gt;           &lt;Rle&gt;        &lt;Rle&gt;</span></span>
<span><span class="co">## 1     unnamed  pheno    unnamed  none pheno PC1,PC2,PC3,PC4      allelic</span></span>
<span><span class="co">## 2     unnamed  pheno    unnamed  none pheno PC1,PC2,PC3,PC4      allelic</span></span>
<span><span class="co">##   MAFweight  test      nvar caseCarriers ctrlCarriers meanCaseScore</span></span>
<span><span class="co">##       &lt;Rle&gt; &lt;Rle&gt; &lt;numeric&gt;    &lt;numeric&gt;    &lt;numeric&gt;     &lt;numeric&gt;</span></span>
<span><span class="co">## 1         1   glm       543          359         1182     0.0761981</span></span>
<span><span class="co">## 2         1 firth       543          359         1182     0.0761981</span></span>
<span><span class="co">##   meanCtrlScore     caseN     ctrlN caseCallRate ctrlCallRate    effect</span></span>
<span><span class="co">##       &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;    &lt;numeric&gt;    &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">## 1     0.0634105      5000     20000      0.96219     0.962038  0.191465</span></span>
<span><span class="co">## 2     0.0634105      5000     20000      0.96219     0.962038  0.192340</span></span>
<span><span class="co">##    effectSE effectCIlower effectCIupper        OR          P</span></span>
<span><span class="co">##   &lt;numeric&gt;     &lt;numeric&gt;     &lt;numeric&gt; &lt;numeric&gt;  &lt;numeric&gt;</span></span>
<span><span class="co">## 1 0.0597474     0.0743618      0.308567   1.21102 0.00135265</span></span>
<span><span class="co">## 2 0.0596968     0.0741091      0.308223   1.21208 0.00153871</span></span></code></pre>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">aggAssoc</span><span class="op">[</span><span class="va">aggAssoc</span><span class="op">$</span><span class="va">geneSetName</span> <span class="op">==</span> <span class="st">"geneset1"</span>,<span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##   geneSetName cohort name pheno           covar  test geneSetSize genesObs</span></span>
<span><span class="co">## 1    geneset1  pheno none pheno PC1,PC2,PC3,PC4   glm           4        4</span></span>
<span><span class="co">## 2    geneset1  pheno none pheno PC1,PC2,PC3,PC4 firth           4        4</span></span>
<span><span class="co">##   caseN ctrlN meanCaseScore meanCtrlScore    effect   effectSE effectCIlower</span></span>
<span><span class="co">## 1  5000 20000    0.07619806    0.06341054 0.1914644 0.05974738    0.07436172</span></span>
<span><span class="co">## 2  5000 20000    0.07619806    0.06341054 0.1923395 0.05969681    0.07410904</span></span>
<span><span class="co">##   effectCIupper       OR           P</span></span>
<span><span class="co">## 1     0.3085671 1.211022 0.001352662</span></span>
<span><span class="co">## 2     0.3082229 1.212082 0.001538717</span></span></code></pre>
<p>Finally, gene set burden analyses can be run from the command-line,
just like other types of assocation tests:</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Sys.setenv.html" class="external-link">Sys.setenv</a></span><span class="op">(</span>rvat <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'exec/rvat.R'</span>, package <span class="op">=</span> <span class="st">'rvat'</span><span class="op">)</span>,</span>
<span>           vcfpath <span class="op">=</span> <span class="fu">rvatData</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/rvatData/man/rvat_example.html" class="external-link">rvat_example</a></span><span class="op">(</span><span class="st">'rvatData.vcf.gz'</span><span class="op">)</span>,</span>
<span>           phenopath <span class="op">=</span> <span class="fu">rvatData</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/rvatData/man/rvat_example.html" class="external-link">rvat_example</a></span><span class="op">(</span><span class="st">'rvatData.pheno'</span><span class="op">)</span>,</span>
<span>           varinfopath <span class="op">=</span> <span class="fu">rvatData</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/rvatData/man/rvat_example.html" class="external-link">rvat_example</a></span><span class="op">(</span><span class="st">'rvatData.varinfo'</span><span class="op">)</span>,</span>
<span>           gdbpath <span class="op">=</span> <span class="va">gdb</span><span class="op">@</span><span class="va">dbname</span>,</span>
<span>           outdir <span class="op">=</span> <span class="va">outdir</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb59"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb59-1"><a href="#cb59-1" tabindex="-1"></a><span class="ex">Rscript</span> <span class="va">$rvat</span> <span class="at">--assocTest</span> <span class="dt">\</span></span>
<span id="cb59-2"><a href="#cb59-2" tabindex="-1"></a>--gdb <span class="va">$gdbpath</span> <span class="dt">\</span></span>
<span id="cb59-3"><a href="#cb59-3" tabindex="-1"></a>--geneSet <span class="va">$outdir</span>/genesetfile.txt.gz <span class="dt">\</span></span>
<span id="cb59-4"><a href="#cb59-4" tabindex="-1"></a>--aggregateFile <span class="va">$outdir</span>/moderate.aggregate.txt.gz <span class="dt">\</span></span>
<span id="cb59-5"><a href="#cb59-5" tabindex="-1"></a>--name maxMAF001 <span class="dt">\</span></span>
<span id="cb59-6"><a href="#cb59-6" tabindex="-1"></a>--covar PC1,PC2,PC3,PC4 <span class="dt">\</span></span>
<span id="cb59-7"><a href="#cb59-7" tabindex="-1"></a>--cohort pheno <span class="dt">\</span></span>
<span id="cb59-8"><a href="#cb59-8" tabindex="-1"></a>--pheno pheno <span class="dt">\</span></span>
<span id="cb59-9"><a href="#cb59-9" tabindex="-1"></a>--test glm,firth <span class="dt">\</span></span>
<span id="cb59-10"><a href="#cb59-10" tabindex="-1"></a>--output <span class="va">$outdir</span>/agg_burden_tests_merged_cmdline.txt.gz</span></code></pre></div>
</div>
</div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-deleeuwStatisticalPropertiesGeneset2016" class="csl-entry">
Leeuw, Christiaan A. de, Benjamin M. Neale, Tom Heskes, and Danielle
Posthuma. 2016. <span>“The Statistical Properties of Gene-Set
Analysis.”</span> <em>Nature Reviews Genetics</em> 17 (6): 353–64. <a href="https://doi.org/10.1038/nrg.2016.29" class="external-link">https://doi.org/10.1038/nrg.2016.29</a>.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Kevin Kenna, Paul Hop.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
