<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Spatial clustering • rvat</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Spatial clustering">
<meta property="og:description" content="rvat">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rvat</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.3.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/basics.html">Getting started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/basics.html">Getting started</a>
    </li>
    <li>
      <a href="../articles/association_testing.html">Association testing</a>
    </li>
    <li>
      <a href="../articles/explore_visualize_results.html">Visualization</a>
    </li>
    <li>
      <a href="../articles/variant_annotation.html">Variant annotation</a>
    </li>
    <li>
      <a href="../articles/GSA.html">Gene set analysis</a>
    </li>
    <li>
      <a href="../articles/spatial_clustering.html">Spatial clustering</a>
    </li>
    <li>
      <a href="../articles/cmdline.html">Command-line tool</a>
    </li>
  </ul>
</li>
<li>
  <a href="../articles/changelog.html">Changelog</a>
</li>
<li>
  <a href="../articles/faq.html">F.A.Q.</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/kkenna/rvat" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Spatial clustering</h1>
            
      
      
      <div class="hidden name"><code>spatial_clustering.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">rvat</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">rvatData</span><span class="op">)</span></span></code></pre></div>
<p>This tutorial shows how regional analyses can be performed in RVAT
using <code><a href="../reference/spatialClust.html">spatialClust()</a></code> that is based on an unsupervised
approach based on the spatial clustering method by<span class="citation">(Loehlein Fier et al. 2017)</span>. It partitions
variants within genes into consecutive non-overlapping windows. In
brief, this algorithm applies a sliding window approach and models the
mutation intensity within windows using a non-homogeneous Poisson point
process. Within each window variants are then divided at points where
the distance is larger than expected under the exponential distribution,
thus defining spatially clustered groups of variants.</p>
<div class="section level2">
<h2 id="setup">setup<a class="anchor" aria-label="anchor" href="#setup"></a>
</h2>
<p>Initialize gdb and output directory use in this tutorial:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gdb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rvatData/man/create_example_gdb.html" class="external-link">create_example_gdb</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">outdir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="map-cds-positions">Map CDS positions<a class="anchor" aria-label="anchor" href="#map-cds-positions"></a>
</h2>
<p>To perform spatial clustering of coding variants, we’ll first map
genomic coordinates to CDS coordinates using the <code>mapToCDS</code>
method.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># The mapToCds method relies on an ensembl gtf/gff file for remapping </span></span>
<span><span class="va">gtf</span> <span class="op">&lt;-</span> <span class="fu">rtracklayer</span><span class="fu">::</span><span class="fu">import</span><span class="op">(</span><span class="st">"https://ftp.ensembl.org/pub/release-105/gtf/homo_sapiens/Homo_sapiens.GRCh38.105.gtf.gz"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## subset relevant genes</span></span>
<span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu"><a href="../reference/getAnno.html">getAnno</a></span><span class="op">(</span><span class="va">gdb</span>, <span class="st">"varInfo"</span>, fields <span class="op">=</span> <span class="st">"gene_name"</span><span class="op">)</span><span class="op">$</span><span class="va">gene_name</span><span class="op">)</span></span>
<span><span class="va">gtf</span> <span class="op">&lt;-</span> <span class="va">gtf</span><span class="op">[</span><span class="va">gtf</span><span class="op">$</span><span class="va">gene_biotype</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="st">"protein_coding"</span> <span class="op">&amp;</span> </span>
<span>           <span class="va">gtf</span><span class="op">$</span><span class="va">transcript_biotype</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="st">"protein_coding"</span> <span class="op">&amp;</span> </span>
<span>           <span class="va">gtf</span><span class="op">$</span><span class="va">gene_name</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="va">genes</span><span class="op">]</span></span>
<span></span>
<span><span class="co">## transcript ids</span></span>
<span><span class="va">transcripts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">gtf</span><span class="op">$</span><span class="va">transcript_id</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># perform remapping</span></span>
<span><span class="co">## note that the `exonPadding` parameter will cause remapping of variants that are 12 base pairs from the exon border to the border.</span></span>
<span><span class="fu"><a href="../reference/mapToCDS.html">mapToCDS</a></span><span class="op">(</span></span>
<span>  <span class="va">gdb</span>,</span>
<span>  gff <span class="op">=</span> <span class="va">gtf</span>,</span>
<span>  transcript_id <span class="op">=</span> <span class="va">transcripts</span>,</span>
<span>  exonPadding <span class="op">=</span> <span class="fl">12</span>,</span>
<span>  output <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"/cdsPOS.txt.gz"</span><span class="op">)</span>,</span>
<span>  biotype <span class="op">=</span> <span class="st">"protein_coding"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># upload to gdb</span></span>
<span><span class="fu">R.utils</span><span class="fu">::</span><span class="fu"><a href="https://henrikbengtsson.github.io/R.utils/reference/compressFile.html" class="external-link">gunzip</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"/cdsPOS.txt.gz"</span><span class="op">)</span>, remove <span class="op">=</span> <span class="cn">FALSE</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/uploadAnno.html">uploadAnno</a></span><span class="op">(</span><span class="va">gdb</span>, value <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"/cdsPOS.txt"</span><span class="op">)</span>, name <span class="op">=</span> <span class="st">"cdsPOS"</span>, skipRemap <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="perform-spatial-clustering">Perform spatial clustering<a class="anchor" aria-label="anchor" href="#perform-spatial-clustering"></a>
</h2>
<p>Here, we perform spatial clustering for each transcript as described
in <span class="citation">(Loehlein Fier et al. 2017)</span> and
implemented in the <code><a href="../reference/spatialClust.html">spatialClust()</a></code> method. It is similar to
the <code><a href="../reference/buildVarSet.html">buildVarSet()</a></code> method, and will likewise generate a
<code>varSetFile</code>. The ‘windowSize’ parameter controls the window
size, and the ‘posField` parameter can be used to specify the position
field to use, in this case ’cdsPOS’ as generated above.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/spatialClust.html">spatialClust</a></span><span class="op">(</span></span>
<span>  <span class="va">gdb</span>,</span>
<span>  unitTable <span class="op">=</span> <span class="st">"varInfo"</span>, </span>
<span>  varSetName <span class="op">=</span> <span class="st">"ModerateHigh"</span>,</span>
<span>  unitName <span class="op">=</span> <span class="st">"transcript_id"</span>,</span>
<span>  where <span class="op">=</span> <span class="st">"ModerateImpact = 1 or HighImpact = 1"</span>,</span>
<span>  windowSize <span class="op">=</span> <span class="fl">30</span>,</span>
<span>  overlap <span class="op">=</span> <span class="fl">15</span>,</span>
<span>  intersection <span class="op">=</span> <span class="st">"cdsPOS"</span>,</span>
<span>  posField <span class="op">=</span> <span class="st">"cdsPOS"</span>,</span>
<span>  output <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"/moderatehigh_clusters.txt.gz"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="association-testing">Association testing<a class="anchor" aria-label="anchor" href="#association-testing"></a>
</h2>
<p>Given that <code><a href="../reference/spatialClust.html">spatialClust()</a></code> generates a
<code><a href="../reference/varSetFile.html">varSetFile()</a></code>, we can as described in other tutorials like
the <code><a href="../articles/association_testing.html">vignette("association_testing")</a></code> tutorial.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># connect ot the varSetFile</span></span>
<span><span class="va">varsetfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/varSetFile.html">varSetFile</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"/moderatehigh_clusters.txt.gz"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># for this example we'll focus on the ALS gene FUS</span></span>
<span><span class="va">units</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/listUnits.html">listUnits</a></span><span class="op">(</span><span class="va">varsetfile</span><span class="op">)</span><span class="op">[</span><span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_detect.html" class="external-link">str_detect</a></span><span class="op">(</span><span class="fu"><a href="../reference/listUnits.html">listUnits</a></span><span class="op">(</span><span class="va">varsetfile</span><span class="op">)</span>, <span class="st">"ENST00000254108"</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># perform burden tests</span></span>
<span><span class="va">burden_spatialclusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/assocTest.html">assocTest</a></span><span class="op">(</span><span class="va">gdb</span>,</span>
<span>                               varSet <span class="op">=</span> <span class="fu"><a href="../reference/getVarSet.html">getVarSet</a></span><span class="op">(</span><span class="va">varsetfile</span>, unit <span class="op">=</span> <span class="va">units</span><span class="op">)</span>,</span>
<span>                               cohort<span class="op">=</span><span class="st">"pheno"</span>,</span>
<span>                               pheno <span class="op">=</span> <span class="st">"pheno"</span>,</span>
<span>                               covar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sex"</span>, <span class="st">"PC1"</span>, <span class="st">"PC2"</span>, <span class="st">"PC3"</span>, <span class="st">"PC4"</span><span class="op">)</span>,</span>
<span>                               test <span class="op">=</span> <span class="st">"firth"</span>,</span>
<span>                               name <span class="op">=</span> <span class="st">"spatial_clustering"</span>,</span>
<span>                               verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">burden_spatialclusters</span></span></code></pre></div>
<pre><code><span><span class="co">## rvbResult with 5 rows and 24 columns</span></span>
<span><span class="co">##                unit cohort         varSetName               name pheno</span></span>
<span><span class="co">##         &lt;character&gt;  &lt;Rle&gt;              &lt;Rle&gt;              &lt;Rle&gt; &lt;Rle&gt;</span></span>
<span><span class="co">## 1 ENST00000254108_0  pheno ModerateHigh_30_15 spatial_clustering pheno</span></span>
<span><span class="co">## 2 ENST00000254108_1  pheno ModerateHigh_30_15 spatial_clustering pheno</span></span>
<span><span class="co">## 3 ENST00000254108_2  pheno ModerateHigh_30_15 spatial_clustering pheno</span></span>
<span><span class="co">## 4 ENST00000254108_3  pheno ModerateHigh_30_15 spatial_clustering pheno</span></span>
<span><span class="co">## 5 ENST00000254108_4  pheno ModerateHigh_30_15 spatial_clustering pheno</span></span>
<span><span class="co">##                 covar geneticModel MAFweight  test      nvar caseCarriers</span></span>
<span><span class="co">##                 &lt;Rle&gt;        &lt;Rle&gt;     &lt;Rle&gt; &lt;Rle&gt; &lt;numeric&gt;    &lt;numeric&gt;</span></span>
<span><span class="co">## 1 sex,PC1,PC2,PC3,PC4      allelic         1 firth        19           15</span></span>
<span><span class="co">## 2 sex,PC1,PC2,PC3,PC4      allelic         1 firth         5            1</span></span>
<span><span class="co">## 3 sex,PC1,PC2,PC3,PC4      allelic         1 firth         9            2</span></span>
<span><span class="co">## 4 sex,PC1,PC2,PC3,PC4      allelic         1 firth         3            1</span></span>
<span><span class="co">## 5 sex,PC1,PC2,PC3,PC4      allelic         1 firth        20           18</span></span>
<span><span class="co">##   ctrlCarriers meanCaseScore meanCtrlScore     caseN     ctrlN caseCallRate</span></span>
<span><span class="co">##      &lt;numeric&gt;     &lt;numeric&gt;     &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;    &lt;numeric&gt;</span></span>
<span><span class="co">## 1           44   0.003064461   0.002265313      5000     20000     0.958979</span></span>
<span><span class="co">## 2            6   0.000205399   0.000305409      5000     20000     0.978760</span></span>
<span><span class="co">## 3           13   0.000416272   0.000666871      5000     20000     0.978400</span></span>
<span><span class="co">## 4            2   0.000213562   0.000113819      5000     20000     0.900733</span></span>
<span><span class="co">## 5           18   0.003639603   0.000939982      5000     20000     0.971170</span></span>
<span><span class="co">##   ctrlCallRate    effect  effectSE effectCIlower effectCIupper        OR</span></span>
<span><span class="co">##      &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;     &lt;numeric&gt;     &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">## 1     0.958750  0.312226  0.296377     -0.299949      0.870650  1.366463</span></span>
<span><span class="co">## 2     0.979250 -0.100711  0.908186     -2.365322      1.464124  0.904195</span></span>
<span><span class="co">## 3     0.978533 -0.284614  0.690642     -1.915206      0.912965  0.752304</span></span>
<span><span class="co">## 4     0.898733  1.111187  1.033718     -1.281935      3.129786  3.037963</span></span>
<span><span class="co">## 5     0.970683  1.444483  0.330752      0.790567      2.098382  4.239660</span></span>
<span><span class="co">##             P</span></span>
<span><span class="co">##     &lt;numeric&gt;</span></span>
<span><span class="co">## 1 3.04661e-01</span></span>
<span><span class="co">## 2 9.10852e-01</span></span>
<span><span class="co">## 3 6.70675e-01</span></span>
<span><span class="co">## 4 3.13982e-01</span></span>
<span><span class="co">## 5 2.65621e-05</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="generate-mutation-plot">generate mutation plot<a class="anchor" aria-label="anchor" href="#generate-mutation-plot"></a>
</h2>
<p>coming soon!</p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-loehleinfierAssociationAnalysisGenomesequencing2017a" class="csl-entry">
Loehlein Fier, Heide, Dmitry Prokopenko, Julian Hecker, Michael H. Cho,
Edwin K. Silverman, Scott T. Weiss, Rudolph E. Tanzi, and Christoph
Lange. 2017. <span>“On the Association Analysis of Genome-Sequencing
Data: <span>A</span> Spatial Clustering Approach for Partitioning the
Entire Genome into Nonoverlapping Windows.”</span> <em>Genetic
Epidemiology</em> 41 (4): 332–40. <a href="https://doi.org/10.1002/gepi.22040" class="external-link">https://doi.org/10.1002/gepi.22040</a>.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Kevin Kenna, Paul Hop.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
