<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Variant annotation • rvat</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Variant annotation">
<meta property="og:description" content="rvat">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rvat</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.3.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/basics.html">Getting started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/basics.html">Getting started</a>
    </li>
    <li>
      <a href="../articles/association_testing.html">Association testing</a>
    </li>
    <li>
      <a href="../articles/explore_visualize_results.html">Visualization</a>
    </li>
    <li>
      <a href="../articles/variant_annotation.html">Variant annotation</a>
    </li>
    <li>
      <a href="../articles/GSA.html">Gene set analysis</a>
    </li>
    <li>
      <a href="../articles/spatial_clustering.html">Spatial clustering</a>
    </li>
    <li>
      <a href="../articles/cmdline.html">Command-line tool</a>
    </li>
  </ul>
</li>
<li>
  <a href="../articles/changelog.html">Changelog</a>
</li>
<li>
  <a href="../articles/faq.html">F.A.Q.</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/kkenna/rvat" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Variant annotation</h1>
            
      
      
      <div class="hidden name"><code>variant_annotation.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">rvat</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">rvatData</span><span class="op">)</span></span></code></pre></div>
<p>Initialize environment:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">outdir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">gdbpath</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outdir</span>,<span class="st">"/rvat_tutorials.gdb"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Sys.setenv.html" class="external-link">Sys.setenv</a></span><span class="op">(</span>outdir <span class="op">=</span> <span class="va">outdir</span>,</span>
<span>           rvat <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'exec/rvat.R'</span>, package <span class="op">=</span> <span class="st">'rvat'</span><span class="op">)</span>,</span>
<span>           gdbrvatdata <span class="op">=</span> <span class="fu">rvatData</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/rvatData/man/rvat_example.html" class="external-link">rvat_example</a></span><span class="op">(</span><span class="st">'rvatData.gdb'</span><span class="op">)</span>,</span>
<span>           gdbpath <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outdir</span>,<span class="st">"/rvat_tutorials.gdb"</span><span class="op">)</span></span>
<span>           <span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="fu">cp</span> <span class="va">$gdbrvatdata</span> <span class="va">$gdbpath</span></span></code></pre></div>
<p>Connect to the example gdb:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gdb</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gdb.html">gdb</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outdir</span>,<span class="st">"/rvat_tutorials.gdb"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="annotate-variants-with-genomic-features">annotate variants with genomic features<a class="anchor" aria-label="anchor" href="#annotate-variants-with-genomic-features"></a>
</h2>
<p>In this example, we’ll annotate variants in the gdb with genomic
features from ensembl.</p>
<p>First, we import the ensembl features in gtf format using <a href="https://bioconductor.org/packages/release/bioc/html/rtracklayer.html" class="external-link">rtracklayer</a>,
resulting in a <a href="https://bioconductor.org/packages/release/bioc/html/GenomicRanges.html" class="external-link">GenomicRanges</a>
object:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gtf</span> <span class="op">&lt;-</span> <span class="fu">rtracklayer</span><span class="fu">::</span><span class="fu">import</span><span class="op">(</span><span class="st">"https://ftp.ensembl.org/pub/release-105/gtf/homo_sapiens/Homo_sapiens.GRCh38.105.gtf.gz"</span><span class="op">)</span></span></code></pre></div>
<p>Then, we’ll use the <code>mapVariants</code> method to map the
variants to protein-coding genes:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gtf_gene</span> <span class="op">&lt;-</span> <span class="va">gtf</span><span class="op">[</span><span class="va">gtf</span><span class="op">$</span><span class="va">gene_biotype</span> <span class="op">==</span> <span class="st">"protein_coding"</span> <span class="op">&amp;</span> <span class="va">gtf</span><span class="op">$</span><span class="va">type</span> <span class="op">==</span> <span class="st">"gene"</span><span class="op">]</span></span>
<span><span class="fu"><a href="../reference/mapVariants.html">mapVariants</a></span><span class="op">(</span><span class="va">gdb</span>,</span>
<span>            ranges <span class="op">=</span> <span class="va">gtf_gene</span>, </span>
<span>            uploadName <span class="op">=</span> <span class="st">"gene"</span>,</span>
<span>            overWrite <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>            <span class="op">)</span></span></code></pre></div>
<p>Similarly, we can map to other features included in ensembl:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gtf_coding</span> <span class="op">&lt;-</span> <span class="va">gtf</span><span class="op">[</span><span class="va">gtf</span><span class="op">$</span><span class="va">gene_biotype</span> <span class="op">==</span> <span class="st">"protein_coding"</span> <span class="op">&amp;</span> <span class="va">gtf</span><span class="op">$</span><span class="va">type</span> <span class="op">==</span> <span class="st">"CDS"</span><span class="op">]</span></span>
<span><span class="fu"><a href="../reference/mapVariants.html">mapVariants</a></span><span class="op">(</span><span class="va">gdb</span>,</span>
<span>            ranges <span class="op">=</span> <span class="va">gtf_gene</span>, </span>
<span>            uploadName <span class="op">=</span> <span class="st">"CDS"</span>,</span>
<span>            overWrite <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>            <span class="op">)</span></span></code></pre></div>
<p>Of course, sets of custom ranges can be used in the same manner:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">custom_ranges</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span></span>
<span>  seqnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chr15"</span>, <span class="st">"chr15"</span><span class="op">)</span>,</span>
<span>  ranges <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span></span>
<span>    start <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">51218500</span>, <span class="fl">51227800</span><span class="op">)</span>,</span>
<span>    end <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">51219500</span>, <span class="fl">51228000</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span> annotation <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/mapVariants.html">mapVariants</a></span><span class="op">(</span><span class="va">gdb</span>,</span>
<span>            ranges <span class="op">=</span> <span class="va">custom_ranges</span>, </span>
<span>            uploadName <span class="op">=</span> <span class="st">"custom_annotation"</span>,</span>
<span>            overWrite <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>            <span class="op">)</span></span></code></pre></div>
<p>The <code>mapVariants</code> method can also read directly from
various file formats including gtf,gff,bed and text files that contain
ranges (CHROM, start, end fields).</p>
<p>For example, below we map variants based on a gtf-file stored on
disk:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># export gene features as gtf</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">mcols</a></span><span class="op">(</span><span class="va">gtf_gene</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">mcols</a></span><span class="op">(</span><span class="va">gtf_gene</span><span class="op">)</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gene_id"</span>, <span class="st">"gene_name"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">gtf_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"/gtf_gene.gtf"</span><span class="op">)</span></span>
<span><span class="fu">rtracklayer</span><span class="fu">::</span><span class="fu">export</span><span class="op">(</span><span class="va">gtf_gene</span>, format <span class="op">=</span> <span class="st">"gtf"</span>, con <span class="op">=</span> <span class="va">gtf_path</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># mapVariants by specifying gff parameter:</span></span>
<span><span class="fu"><a href="../reference/mapVariants.html">mapVariants</a></span><span class="op">(</span><span class="va">gdb</span>,</span>
<span>            gff <span class="op">=</span> <span class="va">gtf_path</span>, </span>
<span>            uploadName <span class="op">=</span> <span class="st">"gene_from_gtf"</span>,</span>
<span>            overWrite <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>            <span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="annotate-variants-with-variant-effect-predictions">annotate variants with variant effect predictions<a class="anchor" aria-label="anchor" href="#annotate-variants-with-variant-effect-predictions"></a>
</h2>
<p>In this example we’ll annotate variants with AlphaMissense scores as
described in <span class="citation">(Cheng et al. 2023)</span>.</p>
<div class="section level3">
<h3 id="import-alphamissense">import AlphaMissense<a class="anchor" aria-label="anchor" href="#import-alphamissense"></a>
</h3>
<p>First, we’ll download the AlphaMissense scores. Note that this takes
up ~5Gb of storage.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="fu">wget</span> <span class="at">--no-verbose</span> <span class="at">-O</span> <span class="at">-</span> https://storage.googleapis.com/dm_alphamissense/AlphaMissense_hg38.tsv.gz <span class="kw">|</span> <span class="fu">gunzip</span> <span class="at">-c</span> <span class="kw">|</span> <span class="fu">sed</span> <span class="st">'1,3d; 4s/#//g'</span> <span class="op">&gt;</span> <span class="va">${outdir}</span>/AlphaMissense_hg38.tsv</span></code></pre></div>
<p>annotations which contain variant positions (CHROM, POS) and
optionally REF/ALT fields can be directly imported using
<code>uploadAnno</code>. By default <code>uploadAnno</code> will discard
all variants that do not map to the gdb. Note that this does take a
little while since we’re uploading a rather bulky table:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/uploadAnno.html">uploadAnno</a></span><span class="op">(</span></span>
<span>  <span class="va">gdb</span>,</span>
<span>  name <span class="op">=</span> <span class="st">"AlphaMissense"</span>,</span>
<span>  value <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"/AlphaMissense_hg38.tsv"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="run-burden-based-on-alphamissense-predictions">run burden based on AlphaMissense predictions<a class="anchor" aria-label="anchor" href="#run-burden-based-on-alphamissense-predictions"></a>
</h3>
<p>Build a varset including variants that are predicted to be ‘likely
pathogenic’ by AlphaMissense</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/buildVarSet.html">buildVarSet</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">gdb</span>, </span>
<span>            output <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"/AlphaMissense_likelypathogenic.txt.gz"</span><span class="op">)</span>,</span>
<span>            varSetName <span class="op">=</span> <span class="st">"AlphaMissense_likelypathogenic"</span>, </span>
<span>            unitTable <span class="op">=</span> <span class="st">"AlphaMissense"</span>, </span>
<span>            unitName <span class="op">=</span> <span class="st">"transcript_id"</span>,</span>
<span>            weightName <span class="op">=</span> <span class="st">"am_pathogenicity"</span>,</span>
<span>            where <span class="op">=</span> <span class="st">"am_class = 'likely_pathogenic'"</span></span>
<span>            <span class="op">)</span></span>
<span><span class="va">am_likely_pathogenic_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"/AlphaMissense_likelypathogenic.txt.gz"</span><span class="op">)</span></span></code></pre></div>
<p>Run a burden test for NEK1 including only variant predicted to be
likely pathogenic:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Initialize a GT object</span></span>
<span><span class="va">varsetfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/varSetFile.html">varSetFile</a></span><span class="op">(</span><span class="va">am_likely_pathogenic_path</span><span class="op">)</span></span>
<span><span class="va">varset</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getVarSet.html">getVarSet</a></span><span class="op">(</span><span class="va">varsetfile</span>, unit <span class="op">=</span> <span class="st">"ENST00000439128.6"</span><span class="op">)</span></span>
<span><span class="va">GT</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getGT.html">getGT</a></span><span class="op">(</span></span>
<span>    <span class="va">gdb</span>,</span>
<span>    varSet <span class="op">=</span> <span class="va">varset</span>,</span>
<span>    cohort <span class="op">=</span> <span class="st">"pheno"</span><span class="op">)</span></span>
<span><span class="va">burden</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/assocTest.html">assocTest</a></span><span class="op">(</span><span class="va">GT</span>,</span>
<span>                    pheno <span class="op">=</span> <span class="st">"pheno"</span>,</span>
<span>                    covar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PC1"</span>, <span class="st">"PC2"</span>, <span class="st">"PC3"</span>, <span class="st">"PC4"</span><span class="op">)</span>,</span>
<span>                    test <span class="op">=</span> <span class="st">"firth"</span>,</span>
<span>                    verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>                    <span class="op">)</span></span>
<span><span class="va">burden</span></span></code></pre></div>
<pre><code><span><span class="co">## rvbResult with 1 row and 24 columns</span></span>
<span><span class="co">##                unit cohort             varSetName  name pheno           covar</span></span>
<span><span class="co">##         &lt;character&gt;  &lt;Rle&gt;                  &lt;Rle&gt; &lt;Rle&gt; &lt;Rle&gt;           &lt;Rle&gt;</span></span>
<span><span class="co">## 1 ENST00000439128.6  pheno AlphaMissense_likely..  none pheno PC1,PC2,PC3,PC4</span></span>
<span><span class="co">##   geneticModel MAFweight  test      nvar caseCarriers ctrlCarriers</span></span>
<span><span class="co">##          &lt;Rle&gt;     &lt;Rle&gt; &lt;Rle&gt; &lt;numeric&gt;    &lt;numeric&gt;    &lt;numeric&gt;</span></span>
<span><span class="co">## 1      allelic         1 firth        17           81          140</span></span>
<span><span class="co">##   meanCaseScore meanCtrlScore     caseN     ctrlN caseCallRate ctrlCallRate</span></span>
<span><span class="co">##       &lt;numeric&gt;     &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;    &lt;numeric&gt;    &lt;numeric&gt;</span></span>
<span><span class="co">## 1     0.0105462    0.00446517      5000     20000     0.969329     0.969224</span></span>
<span><span class="co">##      effect  effectSE effectCIlower effectCIupper        OR           P</span></span>
<span><span class="co">##   &lt;numeric&gt; &lt;numeric&gt;     &lt;numeric&gt;     &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt;</span></span>
<span><span class="co">## 1   1.37704  0.219386      0.942105       1.80364   3.96314 2.03048e-09</span></span></code></pre>
<p>More details on creating variant sets and burden testing can be found
in <code><a href="../articles/association_testing.html">vignette("association_testing")</a></code></p>
</div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-chengAccurateProteomewideMissense2023" class="csl-entry">
Cheng, Jun, Guido Novati, Joshua Pan, Clare Bycroft, Akvilė Žemgulytė,
Taylor Applebaum, Alexander Pritzel, et al. 2023. <span>“Accurate
Proteome-Wide Missense Variant Effect Prediction with
<span>AlphaMissense</span>.”</span> <em>Science</em> 381 (6664):
eadg7492. <a href="https://doi.org/10.1126/science.adg7492" class="external-link">https://doi.org/10.1126/science.adg7492</a>.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Kevin Kenna, Paul Hop.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
