<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Aggregate genotypes into a single (burden) score for each individual — aggregate,gdb-method • rvat</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Aggregate genotypes into a single (burden) score for each individual — aggregate,gdb-method"><meta property="og:description" content="Returns an aggregate of genotypes for each individual.
Note, the gdb implementation is described here, aggregate can also be run directly on a
genoMatrix object as described in the genoMatrix documentation.
Specified genetic model, weights, MAF-weighting are taken into account when aggregating.
Aggregates are written to disk in the aggregateFile format, which can be used as input
for assocTest-aggregateFile to perform gene set burden analyses."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rvat</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/basics.html">Basics</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/basics.html">Basics</a>
    </li>
    <li>
      <a href="../articles/association_testing.html">Association testing</a>
    </li>
    <li>
      <a href="../articles/explore_visualize_results.html">Visualization</a>
    </li>
    <li>
      <a href="../articles/cmdline.html">Command-line tool</a>
    </li>
  </ul></li>
<li>
  <a href="../articles/faq.html">F.A.Q.</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/kkenna/rvat" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Aggregate genotypes into a single (burden) score for each individual</h1>
    
    <div class="hidden name"><code>aggregate-gdb-method.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Returns an aggregate of genotypes for each individual.
Note, the <code><a href="gdb.html">gdb</a></code> implementation is described here, <code>aggregate</code> can also be run directly on a
<code><a href="genoMatrix.html">genoMatrix</a></code> object as described in the <code><a href="genoMatrix.html">genoMatrix</a></code> documentation.
Specified genetic model, weights, MAF-weighting are taken into account when aggregating.
Aggregates are written to disk in the <code><a href="aggregateFile.html">aggregateFile</a></code> format, which can be used as input
for <code><a href="assocTest-aggregateFile.html">assocTest-aggregateFile</a></code> to perform gene set burden analyses.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="co"># S4 method for gdb</span>
<span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/aggregate-methods.html" class="external-link">aggregate</a></span><span class="op">(</span>
  <span class="va">x</span>,
  cohort <span class="op">=</span> <span class="st">"SM"</span>,
  varSet <span class="op">=</span> <span class="cn">NULL</span>,
  VAR_id <span class="op">=</span> <span class="cn">NULL</span>,
  pheno <span class="op">=</span> <span class="cn">NULL</span>,
  memlimit <span class="op">=</span> <span class="fl">1000</span>,
  geneticModel <span class="op">=</span> <span class="st">"allelic"</span>,
  imputeMethod <span class="op">=</span> <span class="st">"meanImpute"</span>,
  MAFweights <span class="op">=</span> <span class="st">"none"</span>,
  checkPloidy <span class="op">=</span> <span class="cn">NULL</span>,
  keep <span class="op">=</span> <span class="cn">NULL</span>,
  output <span class="op">=</span> <span class="cn">NULL</span>,
  signif <span class="op">=</span> <span class="fl">3</span>,
  minCallrateVar <span class="op">=</span> <span class="fl">0</span>,
  maxCallrateVar <span class="op">=</span> <span class="cn">Inf</span>,
  minCallrateSM <span class="op">=</span> <span class="fl">0</span>,
  maxCallrateSM <span class="op">=</span> <span class="cn">Inf</span>,
  minMAF <span class="op">=</span> <span class="fl">0</span>,
  maxMAF <span class="op">=</span> <span class="fl">1</span>,
  minMAC <span class="op">=</span> <span class="fl">0</span>,
  maxMAC <span class="op">=</span> <span class="cn">Inf</span>,
  minCarriers <span class="op">=</span> <span class="fl">0</span>,
  maxCarriers <span class="op">=</span> <span class="cn">Inf</span>,
  minCarrierFreq <span class="op">=</span> <span class="fl">0</span>,
  maxCarrierFreq <span class="op">=</span> <span class="cn">Inf</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>x</dt>
<dd><p>a <code><a href="gdb.html">gdb</a></code> object</p></dd>


<dt>cohort</dt>
<dd><p>If a valid cohort name is provided, then the uploaded data for this cohort is used to filter and annotate the genotypes
If not specified, all samples in the gdb will be loaded.</p></dd>


<dt>varSet</dt>
<dd><p>a <code><a href="varSetList.html">varSetList</a></code> or <code><a href="varSetFile.html">varSetFile</a></code> object.</p></dd>


<dt>VAR_id</dt>
<dd><p>A vector of VAR_ids, alternatively the varSet parameter can be specified.
The <code>memlimit</code> argument controls how many variants to aggregate at a time.</p></dd>


<dt>pheno</dt>
<dd><p>colData field to test as response variable, although not used within this method,
this can be useful to filter samples which have missing data for the response variable.</p></dd>


<dt>memlimit</dt>
<dd><p>Maximum number of variants to load at once (if <code>VAR_id</code> is specified).</p></dd>


<dt>geneticModel</dt>
<dd><p>Which genetic model to apply? ('allelic', 'recessive' or 'dominant').
Defaults to <code>allelic</code>.</p></dd>


<dt>imputeMethod</dt>
<dd><p>Which imputation method to apply? ('meanImpute' or 'missingToRef').
Defaults to <code>meanImpute</code>.</p></dd>


<dt>MAFweights</dt>
<dd><p>MAF weighting method. Currently Madsen-Browning ('mb') is implemented, by default no MAF weighting is applied.</p></dd>


<dt>checkPloidy</dt>
<dd><p>Version of the human genome to use when assigning variant ploidy (diploid,XnonPAR,YnonPAR).
Accepted inputs are GRCh37, hg19, GRCh38, hg38.
If no value is provided then all variants are assigned the default ploidy of "diploid".</p></dd>


<dt>keep</dt>
<dd><p>vector of sample IDs to keep, defaults to <code>NULL</code>, in which case all samples are kept.</p></dd>


<dt>output</dt>
<dd><p>Output file path for results.
Defaults to <code>NULL</code>, in which case results are not written.</p></dd>


<dt>minCallrateVar</dt>
<dd><p>Minimum genotype rate for variant retention.</p></dd>


<dt>maxCallrateVar</dt>
<dd><p>Maximum genotype rate for variant retention.</p></dd>


<dt>minCallrateSM</dt>
<dd><p>Minimum genotype rate for sample retention.</p></dd>


<dt>maxCallrateSM</dt>
<dd><p>Maximum genotype rate for sample retention.</p></dd>


<dt>minMAF</dt>
<dd><p>Minimum minor allele frequency for variant retention.</p></dd>


<dt>maxMAF</dt>
<dd><p>Maximum minor allele frequency for variant retention.</p></dd>


<dt>minMAC</dt>
<dd><p>Minimum minor allele count for variant retention.</p></dd>


<dt>maxMAC</dt>
<dd><p>Maximum minor allele count for variant retention.</p></dd>


<dt>minCarriers</dt>
<dd><p>Minimum carrier count for variant retention.</p></dd>


<dt>maxCarriers</dt>
<dd><p>Maximum carrier count for variant retention.</p></dd>


<dt>minCarrierFreq</dt>
<dd><p>Minimum carrier frequency for variant retention.</p></dd>


<dt>maxCarrierFreq</dt>
<dd><p>Maximum carrier frequency for variant retention.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    


    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Kevin Kenna, Paul Hop.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

