<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>assocTest-genoMatrix — assocTest-genoMatrix • rvat</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="assocTest-genoMatrix — assocTest-genoMatrix"><meta property="og:description" content="Run assocTest on a genoMatrix object. See the main assocTest page for details."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rvat</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/basics.html">Basics</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/basics.html">Basics</a>
    </li>
    <li>
      <a href="../articles/association_testing.html">Association testing</a>
    </li>
    <li>
      <a href="../articles/explore_visualize_results.html">Visualization</a>
    </li>
    <li>
      <a href="../articles/cmdline.html">Command-line tool</a>
    </li>
  </ul></li>
<li>
  <a href="../articles/faq.html">F.A.Q.</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/kkenna/rvat" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>assocTest-genoMatrix</h1>
    
    <div class="hidden name"><code>assocTest-genoMatrix.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Run <code><a href="assocTest.html">assocTest</a></code> on a <code><a href="genoMatrix.html">genoMatrix</a></code> object. See the main <code><a href="assocTest.html">assocTest</a></code> page for details.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="co"># S4 method for genoMatrix</span>
<span class="fu"><a href="assocTest.html">assocTest</a></span><span class="op">(</span>
  <span class="va">object</span>,
  <span class="va">pheno</span>,
  <span class="va">test</span>,
  name <span class="op">=</span> <span class="st">"none"</span>,
  continuous <span class="op">=</span> <span class="cn">FALSE</span>,
  singlevar <span class="op">=</span> <span class="cn">FALSE</span>,
  covar <span class="op">=</span> <span class="cn">NULL</span>,
  offset <span class="op">=</span> <span class="cn">NULL</span>,
  overwriteAggregate <span class="op">=</span> <span class="cn">TRUE</span>,
  geneticModel <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"allelic"</span>, <span class="st">"recessive"</span>, <span class="st">"dominant"</span><span class="op">)</span>,
  imputeMethod <span class="op">=</span> <span class="cn">NULL</span>,
  MAFweights <span class="op">=</span> <span class="st">"none"</span>,
  maxitFirth <span class="op">=</span> <span class="fl">1000</span>,
  keep <span class="op">=</span> <span class="cn">NULL</span>,
  output <span class="op">=</span> <span class="cn">NULL</span>,
  append <span class="op">=</span> <span class="cn">FALSE</span>,
  returnDF <span class="op">=</span> <span class="cn">FALSE</span>,
  methodResampling <span class="op">=</span> <span class="cn">NULL</span>,
  resamplingMatrix <span class="op">=</span> <span class="cn">NULL</span>,
  nResampling <span class="op">=</span> <span class="fl">1000</span>,
  outputResampling <span class="op">=</span> <span class="cn">FALSE</span>,
  memlimitResampling <span class="op">=</span> <span class="cn">NULL</span>,
  minCallrateVar <span class="op">=</span> <span class="fl">0</span>,
  maxCallrateVar <span class="op">=</span> <span class="cn">Inf</span>,
  minCallrateSM <span class="op">=</span> <span class="fl">0</span>,
  maxCallrateSM <span class="op">=</span> <span class="cn">Inf</span>,
  minMAF <span class="op">=</span> <span class="fl">0</span>,
  maxMAF <span class="op">=</span> <span class="fl">1</span>,
  minMAC <span class="op">=</span> <span class="fl">0</span>,
  maxMAC <span class="op">=</span> <span class="cn">Inf</span>,
  minCarriers <span class="op">=</span> <span class="fl">0</span>,
  maxCarriers <span class="op">=</span> <span class="cn">Inf</span>,
  minCarrierFreq <span class="op">=</span> <span class="fl">0</span>,
  maxCarrierFreq <span class="op">=</span> <span class="cn">Inf</span>,
  verbose <span class="op">=</span> <span class="cn">TRUE</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>object</dt>
<dd><p>a <code><a href="genoMatrix.html">genoMatrix</a></code> object</p></dd>


<dt>pheno</dt>
<dd><p>colData field to test as response variable, the response variable
can either be binary (0/1) or continuous. If the response variable is continuous set
<code>continuous</code> to <code>TRUE</code>.</p></dd>


<dt>test</dt>
<dd><p>Vector of statistical tests to run,
options include firth,glm,lm,scoreSPA,skat,skat_burden,skato,skat_fwe,skat_burden_fwe,
skato_fwe,skat_robust,skato_robust,skat_burden_robust, acatv, acatvSPA. See <code><a href="assocTest.html">assocTest</a></code> for details.</p></dd>


<dt>name</dt>
<dd><p>Optional name for the analysis, defaults to "none".</p></dd>


<dt>continuous</dt>
<dd><p>Is the response variable continuous? (TRUE/FALSE). Defaults to <code>FALSE</code>.</p></dd>


<dt>singlevar</dt>
<dd><p>Run single variant tests? (TRUE/FALSE).
Defaults to <code>FALSE</code>, in which case collapsing tests are ran.</p></dd>


<dt>covar</dt>
<dd><p>Character vector of covariates. These should be present in the colData slot of the genoMatrix.</p></dd>


<dt>overwriteAggregate</dt>
<dd><p>In case there is already an <code>aggregate</code> column in the <code>colData</code> of the genoMatrix
(i.e. <code>aggregate</code> has been run on the genoMatrix), should it be overwitten? Defaults to <code>TRUE</code>.</p></dd>


<dt>geneticModel</dt>
<dd><p>Which genetic model to apply? ('allelic', 'recessive' or 'dominant').
Defaults to <code>allelic</code>.</p></dd>


<dt>imputeMethod</dt>
<dd><p>Which imputation method to apply? ('meanImpute' or 'missingToRef').
Defaults to <code>meanImpute</code></p></dd>


<dt>MAFweights</dt>
<dd><p>Apply MAF weighting? Currently Madsen-Browning ('mb') is implemented.
Defaults to 'none'.</p></dd>


<dt>maxitFirth</dt>
<dd><p>Maximum number of iterations to use for estimating firth confidence intervals.</p></dd>


<dt>output</dt>
<dd><p>Output file path for results.
Defaults to <code>NULL</code>, in which case results are not written to disk, but returned as an <code><a href="rvatResult.html">rvatResult</a></code> object.</p></dd>


<dt>append</dt>
<dd><p>Relevant if the <code>output</code> parameter is not <code>NULL</code>. Should results be appended to <code>output</code>?
Defaults to <code>FALSE</code>.</p></dd>


<dt>methodResampling</dt>
<dd><p>Which method to use for resampling? ('permutation' currently implemented)
Defaults to <code>NULL</code>, in which case no resampling is performed.</p></dd>


<dt>resamplingMatrix</dt>
<dd><p>Pre-calculated resampling matrix (n x p), where n = number of samples, and p number of resamplings.
Can be generated using <code><a href="buildResamplingFile.html">buildResamplingFile</a></code>.</p></dd>


<dt>nResampling</dt>
<dd><p>Number of resamplings to perform if methodResampling is specified.</p></dd>


<dt>outputResampling</dt>
<dd><p>If <code>TRUE</code> or a filepath, results for each resampling are returned (or saved to the filepath).
This can be useful if permutations are used to calculated to estimate correlations among genes for example.
Defaults to <code>FALSE</code> in which case resampling is used to calculate resampled P-values,
results for individual resamplings are not returned.</p></dd>


<dt>memlimitResampling</dt>
<dd><p>Maximum number of resamplings to perform at a time.
Resampling generates a matrix of n x p, where n is the number of samples and p the number of resamplings
thus, for large number of resamplings it can be more efficient to split the permutations in chunks of size <code>memlimitResampling</code>.
Defaults to <code>NULL</code> in which case all permutations are performed.</p></dd>


<dt>minCallrateVar</dt>
<dd><p>Minimum genotype rate for variant retention.</p></dd>


<dt>maxCallrateVar</dt>
<dd><p>Maximum genotype rate for variant retention.</p></dd>


<dt>minCallrateSM</dt>
<dd><p>Minimum genotype rate for sample retention.</p></dd>


<dt>maxCallrateSM</dt>
<dd><p>Maximum genotype rate for sample retention.</p></dd>


<dt>minMAF</dt>
<dd><p>Minimum minor allele frequency for variant retention.</p></dd>


<dt>maxMAF</dt>
<dd><p>Maximum minor allele frequency for variant retention.</p></dd>


<dt>minMAC</dt>
<dd><p>Minimum minor allele count for variant retention.</p></dd>


<dt>maxMAC</dt>
<dd><p>Maximum minor allele count for variant retention.</p></dd>


<dt>minCarriers</dt>
<dd><p>Minimum carrier count for variant retention.</p></dd>


<dt>maxCarriers</dt>
<dd><p>Maximum carrier count for variant retention.</p></dd>


<dt>minCarrierFreq</dt>
<dd><p>Minimum carrier frequency for variant retention.</p></dd>


<dt>maxCarrierFreq</dt>
<dd><p>Maximum carrier frequency for variant retention.</p></dd>


<dt>verbose</dt>
<dd><p>Should the function be verbose? (TRUE/FALSE), defaults to <code>TRUE</code>.</p></dd>

</dl></div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Kevin Kenna, Paul Hop.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

