<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>assocTest-gdb — assocTest-gdb • rvat</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="assocTest-gdb — assocTest-gdb"><meta property="og:description" content="Run assocTest on a gdb object. See the main assocTest page for details."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rvat</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.3.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/basics.html">Getting started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/basics.html">Getting started</a>
    </li>
    <li>
      <a href="../articles/association_testing.html">Association testing</a>
    </li>
    <li>
      <a href="../articles/explore_visualize_results.html">Visualization</a>
    </li>
    <li>
      <a href="../articles/variant_annotation.html">Variant annotation</a>
    </li>
    <li>
      <a href="../articles/GSA.html">Gene set analysis</a>
    </li>
    <li>
      <a href="../articles/spatial_clustering.html">Spatial clustering</a>
    </li>
    <li>
      <a href="../articles/cmdline.html">Command-line tool</a>
    </li>
  </ul></li>
<li>
  <a href="../articles/changelog.html">Changelog</a>
</li>
<li>
  <a href="../articles/faq.html">F.A.Q.</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/KennaLab/rvat" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>assocTest-gdb</h1>
    
    <div class="hidden name"><code>assocTest-gdb.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Run <code><a href="assocTest.html">assocTest</a></code> on a <code><a href="gdb.html">gdb</a></code> object. See the main <code><a href="assocTest.html">assocTest</a></code> page for details.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="co"># S4 method for gdb</span></span>
<span><span class="fu"><a href="assocTest.html">assocTest</a></span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  <span class="va">pheno</span>,</span>
<span>  <span class="va">test</span>,</span>
<span>  cohort <span class="op">=</span> <span class="st">"SM"</span>,</span>
<span>  varSet <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  VAR_id <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  name <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>  continuous <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  singlevar <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  covar <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  offset <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  geneticModel <span class="op">=</span> <span class="st">"allelic"</span>,</span>
<span>  imputeMethod <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  MAFweights <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>  maxitFirth <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  checkPloidy <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  keep <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  output <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  methodResampling <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  resamplingFile <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  nResampling <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  outputResampling <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  memlimitResampling <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  minCallrateVar <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  maxCallrateVar <span class="op">=</span> <span class="cn">Inf</span>,</span>
<span>  minCallrateSM <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  maxCallrateSM <span class="op">=</span> <span class="cn">Inf</span>,</span>
<span>  minMAF <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  maxMAF <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  minMAC <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  maxMAC <span class="op">=</span> <span class="cn">Inf</span>,</span>
<span>  minCarriers <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  maxCarriers <span class="op">=</span> <span class="cn">Inf</span>,</span>
<span>  minCarrierFreq <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  maxCarrierFreq <span class="op">=</span> <span class="cn">Inf</span>,</span>
<span>  memlimit <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  strict <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>object</dt>
<dd><p>a <code><a href="gdb.html">gdb</a></code> object</p></dd>


<dt>pheno</dt>
<dd><p>colData field to test as response variable, the response variable
can either be binary (0/1) or continuous.
If the response variable is continuous set <code>continuous</code> to <code>TRUE</code>.
Multiple phenotypes can be specified, which will then be tested separately.</p></dd>


<dt>test</dt>
<dd><p>Vector of statistical tests to run,
options include firth,glm,lm,nbinom,skat,skat_burden,skato,skat_fwe,skat_burden_fwe,
skato_fwe,skat_robust,skato_robust,skat_burden_robust, acatv, acatvSPA.
See <code><a href="assocTest.html">assocTest</a></code> for details.</p></dd>


<dt>cohort</dt>
<dd><p>If a valid cohort name is provided, then the uploaded data for
this cohort is used to filter and annotate the genoMatrix object.
If not specified, all samples in the gdb will be loaded.</p></dd>


<dt>varSet</dt>
<dd><p>a <code><a href="varSetFile.html">varSetFile</a></code> or <code><a href="varSetList.html">varSetList</a></code> object, alternatively a vector of VAR_ids can be specified using the <code>VAR_id</code> parameter.</p></dd>


<dt>VAR_id</dt>
<dd><p>a vector of VAR_ids, alternatively the <code>varSet</code> parameter can be specified.
If single variant tests are ran, the <code>memlimit</code> argument controls how many variants to analyze at a time.</p></dd>


<dt>name</dt>
<dd><p>Optional name for the analysis, defaults to "none".</p></dd>


<dt>continuous</dt>
<dd><p>If the response variable continuous? (TRUE/FALSE). Defaults to <code>FALSE</code>.</p></dd>


<dt>singlevar</dt>
<dd><p>Run single variant tests? (TRUE/FALSE).
Defaults to <code>FALSE</code>, in which case aggregate tests are ran.</p></dd>


<dt>covar</dt>
<dd><p>Character vector of covariates, or a list of character vectors of covariates in which case each covariate set will be tested separately.</p></dd>


<dt>offset</dt>
<dd><p>Optional model offset, can be used to account for regenie LOCO predictions.</p></dd>


<dt>geneticModel</dt>
<dd><p>Which genetic model to apply? ('allelic', 'recessive' or 'dominant').
Defaults to <code>allelic</code>.
Multiple geneticModels can be specified, in which case each will be analyzed separately.</p></dd>


<dt>imputeMethod</dt>
<dd><p>Which imputation method to apply? ('meanImpute' or 'missingToRef').
Defaults to <code>meanImpute</code>.</p></dd>


<dt>MAFweights</dt>
<dd><p>MAF weighting method. Currently Madsen-Browning ('mb') is implemented, by default no MAF weighting is applied.
Multiple MAFweights can be specified, in which case each will be analyzed separately.</p></dd>


<dt>maxitFirth</dt>
<dd><p>Maximum number of iterations to use for estimating firth confidence intervals. Defaults to 1000.</p></dd>


<dt>checkPloidy</dt>
<dd><p>Version of the human genome to use when assigning variant ploidy (diploid, XnonPAR, YnonPAR).
Accepted inputs are GRCh37, hg19, GRCh38, hg38.
If not specified, the genome build in the <code><a href="gdb.html">gdb</a></code> will be used, if available (included if the <code>genomeBuild</code> parameter was set in <code><a href="buildGdb.html">buildGdb</a></code>).
Otherwise, if the genome build is not included in the gdb metadata, and no value is provided, then all variants are assigned the default ploidy of "diploid"</p></dd>


<dt>keep</dt>
<dd><p>Vector of sample IDs to keep, defaults to <code>NULL</code>, in which case all samples are kept.</p></dd>


<dt>output</dt>
<dd><p>Output file path for results.
Defaults to <code>NULL</code>, in which case results are not written.</p></dd>


<dt>methodResampling</dt>
<dd><p>Which method to use for resampling? ('permutation' currently implemented)
Defaults to <code>NULL</code>, in which case no resampling is performed.</p></dd>


<dt>resamplingFile</dt>
<dd><p>A <code><a href="resamplingFile.html">resamplingFile</a></code> object.</p></dd>


<dt>nResampling</dt>
<dd><p>Number of resamplings to perform if methodResampling is specified.</p></dd>


<dt>outputResampling</dt>
<dd><p>If <code>TRUE</code> or a filepath, results for each resampling are returned (or saved to the filepath).
This can be useful if permutations are used to calculated to estimate correlations among genes for example.
Defaults to <code>FALSE</code> in which case resampling is used to calculate resampled P-values,
results for individual resamplings are not returned.</p></dd>


<dt>memlimitResampling</dt>
<dd><p>Maximum number of resamplings to perform at a time.
Resampling generates a matrix of n x p, where n is the number of samples and p the number of resamplings
thus, for large number of resamplings it can be more efficient to split the permutations in chunks of size <code>memlimitResampling</code>.
Defaults to <code>NULL</code> in which case all permutations are performed.</p></dd>


<dt>minCallrateVar</dt>
<dd><p>Minimum genotype rate for variant retention.</p></dd>


<dt>maxCallrateVar</dt>
<dd><p>Maximum genotype rate for variant retention.</p></dd>


<dt>minCallrateSM</dt>
<dd><p>Minimum genotype rate for sample retention.</p></dd>


<dt>maxCallrateSM</dt>
<dd><p>Maximum genotype rate for sample retention.</p></dd>


<dt>minMAF</dt>
<dd><p>Minimum minor allele frequency for variant retention.</p></dd>


<dt>maxMAF</dt>
<dd><p>Maximum minor allele frequency for variant retention.</p></dd>


<dt>minMAC</dt>
<dd><p>Minimum minor allele count for variant retention.</p></dd>


<dt>maxMAC</dt>
<dd><p>Maximum minor allele count for variant retention.</p></dd>


<dt>minCarriers</dt>
<dd><p>Minimum carrier count for variant retention.</p></dd>


<dt>maxCarriers</dt>
<dd><p>Maximum carrier count for variant retention.</p></dd>


<dt>minCarrierFreq</dt>
<dd><p>Minimum carrier frequency for variant retention.</p></dd>


<dt>maxCarrierFreq</dt>
<dd><p>Maximum carrier frequency for variant retention.</p></dd>


<dt>memlimit</dt>
<dd><p>Maximum number of variants to load at once (if <code>VAR_id</code> is specified).</p></dd>


<dt>verbose</dt>
<dd><p>Should the function be verbose? (TRUE/FALSE), defaults to <code>TRUE</code>.</p></dd>


<dt>strict</dt>
<dd><p>Should strict checks be performed? Defaults to <code>TRUE</code>. Strict tests currently includes
checking whether supplied varSetFile/varSetList was generated from the same gdb as specified in <code>object</code>.</p></dd>

</dl></div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">rvatData</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">gdb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rvatData/man/create_example_gdb.html" class="external-link">create_example_gdb</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">varsetfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="varSetFile.html">varSetFile</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/rvatData/man/rvat_example.html" class="external-link">rvat_example</a></span><span class="op">(</span><span class="st">"rvatData_varsetfile.txt.gz"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># for example purposes, upload a small cohort</span></span></span>
<span class="r-in"><span><span class="va">cohort</span> <span class="op">&lt;-</span> <span class="fu"><a href="getCohort.html">getCohort</a></span><span class="op">(</span><span class="va">gdb</span>, <span class="st">"pheno"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">cohort</span> <span class="op">&lt;-</span> <span class="va">cohort</span><span class="op">[</span><span class="va">cohort</span><span class="op">$</span><span class="va">IID</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">GTsmall</span><span class="op">)</span>,<span class="op">]</span></span></span>
<span class="r-in"><span><span class="fu"><a href="uploadCohort.html">uploadCohort</a></span><span class="op">(</span><span class="va">gdb</span>, name <span class="op">=</span> <span class="st">"phenosmall"</span>, value <span class="op">=</span> <span class="va">cohort</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading cohort 'phenosmall' from interactive R session</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 10 fields detected (IID,sex,pheno,pop,superPop,PC1,PC2,PC3,PC4,age)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2602 males, 2398 females and 0 unknown gender</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Retained 5000 of 25000 uploaded samples that could be mapped to dosage matrix</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 1</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># run a firth burden test on a binary phenotype</span></span></span>
<span class="r-in"><span><span class="va">varsetlist</span> <span class="op">&lt;-</span> <span class="fu"><a href="getVarSet.html">getVarSet</a></span><span class="op">(</span><span class="va">varsetfile</span>, unit <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SOD1"</span>, <span class="st">"NEK1"</span><span class="op">)</span>, varSetName <span class="op">=</span> <span class="st">"High"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">rvb</span> <span class="op">&lt;-</span> <span class="fu"><a href="assocTest.html">assocTest</a></span><span class="op">(</span><span class="va">gdb</span>,</span></span>
<span class="r-in"><span>                 varSet <span class="op">=</span> <span class="va">varsetlist</span>,</span></span>
<span class="r-in"><span>                 cohort <span class="op">=</span> <span class="st">"phenosmall"</span>,</span></span>
<span class="r-in"><span>                 pheno <span class="op">=</span> <span class="st">"pheno"</span>,</span></span>
<span class="r-in"><span>                 covar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PC1"</span>, <span class="st">"PC2"</span>, <span class="st">"PC3"</span>, <span class="st">"PC4"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                 test <span class="op">=</span> <span class="st">"firth"</span>,</span></span>
<span class="r-in"><span>                 name <span class="op">=</span> <span class="st">"example"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Analysing NEK1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Retrieved genotypes for 33 variants</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 5000/25000 samples in the gdb are present in cohort 'phenosmall'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Analysing unit NEK1; varSet High</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Keeping 5000/5000 available samples for analysis.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 33/33 variants are retained for analysis</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 20/33 variants have zero carriers, these are dropped.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Analysing SOD1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Retrieved genotypes for 4 variants</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 5000/25000 samples in the gdb are present in cohort 'phenosmall'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Analysing unit SOD1; varSet High</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Keeping 5000/5000 available samples for analysis.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 4/4 variants are retained for analysis</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2/4 variants have zero carriers, these are dropped.</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># run ACAT-v and SKAT tests</span></span></span>
<span class="r-in"><span><span class="va">rvb</span> <span class="op">&lt;-</span> <span class="fu"><a href="assocTest.html">assocTest</a></span><span class="op">(</span><span class="va">gdb</span>,</span></span>
<span class="r-in"><span>                 varSet <span class="op">=</span> <span class="va">varsetlist</span>,</span></span>
<span class="r-in"><span>                 cohort <span class="op">=</span> <span class="st">"phenosmall"</span>,</span></span>
<span class="r-in"><span>                 pheno <span class="op">=</span> <span class="st">"pheno"</span>,</span></span>
<span class="r-in"><span>                 covar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PC1"</span>, <span class="st">"PC2"</span>, <span class="st">"PC3"</span>, <span class="st">"PC4"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                 test <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"acatvfirth"</span>, <span class="st">"skat_burden_robust"</span>, <span class="st">"skato_robust"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                 name <span class="op">=</span> <span class="st">"example"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Analysing NEK1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Retrieved genotypes for 33 variants</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 5000/25000 samples in the gdb are present in cohort 'phenosmall'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Analysing unit NEK1; varSet High</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Keeping 5000/5000 available samples for analysis.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 33/33 variants are retained for analysis</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 20/33 variants have zero carriers, these are dropped.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Analysing SOD1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Retrieved genotypes for 4 variants</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 5000/25000 samples in the gdb are present in cohort 'phenosmall'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Analysing unit SOD1; varSet High</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Keeping 5000/5000 available samples for analysis.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 4/4 variants are retained for analysis</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2/4 variants have zero carriers, these are dropped.</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># run a burden test on a continuous phenotype</span></span></span>
<span class="r-in"><span><span class="va">rvb</span> <span class="op">&lt;-</span> <span class="fu"><a href="assocTest.html">assocTest</a></span><span class="op">(</span><span class="va">gdb</span>,</span></span>
<span class="r-in"><span>                 varSet <span class="op">=</span> <span class="va">varsetlist</span>,</span></span>
<span class="r-in"><span>                 cohort <span class="op">=</span> <span class="st">"phenosmall"</span>,</span></span>
<span class="r-in"><span>                 pheno <span class="op">=</span> <span class="st">"age"</span>,</span></span>
<span class="r-in"><span>                 continuous <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>                 covar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PC1"</span>, <span class="st">"PC2"</span>, <span class="st">"PC3"</span>, <span class="st">"PC4"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                 test <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"lm"</span>, <span class="st">"skat"</span>, <span class="st">"acatv"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                 name <span class="op">=</span> <span class="st">"example"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Analysing NEK1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Retrieved genotypes for 33 variants</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 5000/25000 samples in the gdb are present in cohort 'phenosmall'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Analysing unit NEK1; varSet High</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Keeping 5000/5000 available samples for analysis.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 33/33 variants are retained for analysis</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 20/33 variants have zero carriers, these are dropped.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Analysing SOD1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Retrieved genotypes for 4 variants</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 5000/25000 samples in the gdb are present in cohort 'phenosmall'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Analysing unit SOD1; varSet High</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Keeping 5000/5000 available samples for analysis.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 4/4 variants are retained for analysis</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2/4 variants have zero carriers, these are dropped.</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># run single variant tests on a binary phenotype </span></span></span>
<span class="r-in"><span><span class="va">sv</span> <span class="op">&lt;-</span> <span class="fu"><a href="assocTest.html">assocTest</a></span><span class="op">(</span><span class="va">gdb</span>,</span></span>
<span class="r-in"><span>                varSet <span class="op">=</span> <span class="va">varsetlist</span>,</span></span>
<span class="r-in"><span>                cohort <span class="op">=</span> <span class="st">"phenosmall"</span>,</span></span>
<span class="r-in"><span>                pheno <span class="op">=</span> <span class="st">"pheno"</span>,</span></span>
<span class="r-in"><span>                singlevar <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>                covar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PC1"</span>, <span class="st">"PC2"</span>, <span class="st">"PC3"</span>, <span class="st">"PC4"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                test <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"firth"</span>, <span class="st">"glm"</span>, <span class="st">"scoreSPA"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                name <span class="op">=</span> <span class="st">"example"</span>,</span></span>
<span class="r-in"><span>                minCarriers <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Analysing NEK1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Retrieved genotypes for 33 variants</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 5000/25000 samples in the gdb are present in cohort 'phenosmall'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Analysing unit NEK1; varSet High</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Keeping 5000/5000 available samples for analysis.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 13/13 variants are retained for analysis</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>13/13 variants have less than 2 carriers, tests will be skipped for these variants.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Analysing SOD1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Retrieved genotypes for 4 variants</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 5000/25000 samples in the gdb are present in cohort 'phenosmall'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Analysing unit SOD1; varSet High</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Keeping 5000/5000 available samples for analysis.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2/2 variants are retained for analysis</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>2/2 variants have less than 2 carriers, tests will be skipped for these variants.</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># similarly a list of VAR_ids can be specified instead of a varSetList/varSetFile</span></span></span>
<span class="r-in"><span><span class="va">sv</span> <span class="op">&lt;-</span> <span class="fu"><a href="assocTest.html">assocTest</a></span><span class="op">(</span><span class="va">gdb</span>,</span></span>
<span class="r-in"><span>                VAR_id <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">20</span>,</span></span>
<span class="r-in"><span>                cohort <span class="op">=</span> <span class="st">"phenosmall"</span>,</span></span>
<span class="r-in"><span>                pheno <span class="op">=</span> <span class="st">"pheno"</span>,</span></span>
<span class="r-in"><span>                singlevar <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>                covar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PC1"</span>, <span class="st">"PC2"</span>, <span class="st">"PC3"</span>, <span class="st">"PC4"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                test <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"firth"</span>, <span class="st">"glm"</span>, <span class="st">"scoreSPA"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                name <span class="op">=</span> <span class="st">"example"</span>,</span></span>
<span class="r-in"><span>                minCarriers <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Analysing chunk1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Retrieved genotypes for 20 variants</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 5000/25000 samples in the gdb are present in cohort 'phenosmall'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Analysing unit chunk1; varSet none</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Keeping 5000/5000 available samples for analysis.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 8/8 variants are retained for analysis</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>4/8 variants have less than 2 carriers, tests will be skipped for these variants.</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># apply variant filters</span></span></span>
<span class="r-in"><span><span class="va">rvb</span> <span class="op">&lt;-</span> <span class="fu"><a href="assocTest.html">assocTest</a></span><span class="op">(</span><span class="va">gdb</span>,</span></span>
<span class="r-in"><span>                 varSet <span class="op">=</span> <span class="va">varsetlist</span>,</span></span>
<span class="r-in"><span>                 cohort <span class="op">=</span> <span class="st">"phenosmall"</span>,</span></span>
<span class="r-in"><span>                 pheno <span class="op">=</span> <span class="st">"pheno"</span>,</span></span>
<span class="r-in"><span>                 covar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PC1"</span>, <span class="st">"PC2"</span>, <span class="st">"PC3"</span>, <span class="st">"PC4"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                 test <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"firth"</span>, <span class="st">"skat_robust"</span>, <span class="st">"acatv"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                 name <span class="op">=</span> <span class="st">"example"</span>, </span></span>
<span class="r-in"><span>                 maxMAF <span class="op">=</span> <span class="fl">0.05</span>,</span></span>
<span class="r-in"><span>                 minCarriers <span class="op">=</span> <span class="fl">1</span>,</span></span>
<span class="r-in"><span>                 minCallrateVar <span class="op">=</span> <span class="fl">0.9</span>,</span></span>
<span class="r-in"><span>                 minCallrateSM <span class="op">=</span> <span class="fl">0.95</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Analysing NEK1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Retrieved genotypes for 33 variants</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 5000/25000 samples in the gdb are present in cohort 'phenosmall'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Analysing unit NEK1; varSet High</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 1609 samples don't pass callRate filters.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Keeping 3391/3391 available samples for analysis.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 10/10 variants are retained for analysis</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Analysing SOD1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Retrieved genotypes for 4 variants</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 5000/25000 samples in the gdb are present in cohort 'phenosmall'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Analysing unit SOD1; varSet High</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 255 samples don't pass callRate filters.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Keeping 4745/4745 available samples for analysis.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2/2 variants are retained for analysis</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Perform MAF-weighted burden tests (madsen-browning)</span></span></span>
<span class="r-in"><span><span class="va">rvb</span> <span class="op">&lt;-</span> <span class="fu"><a href="assocTest.html">assocTest</a></span><span class="op">(</span><span class="va">gdb</span>,</span></span>
<span class="r-in"><span>                 varSet <span class="op">=</span> <span class="va">varsetlist</span>,</span></span>
<span class="r-in"><span>                 cohort <span class="op">=</span> <span class="st">"phenosmall"</span>,</span></span>
<span class="r-in"><span>                 pheno <span class="op">=</span> <span class="st">"pheno"</span>,</span></span>
<span class="r-in"><span>                 covar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PC1"</span>, <span class="st">"PC2"</span>, <span class="st">"PC3"</span>, <span class="st">"PC4"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                 test <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"firth"</span>, <span class="st">"skat_robust"</span>, <span class="st">"acatv"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                 MAFweights <span class="op">=</span> <span class="st">"mb"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Analysing NEK1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Retrieved genotypes for 33 variants</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 5000/25000 samples in the gdb are present in cohort 'phenosmall'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Analysing unit NEK1; varSet High</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Keeping 5000/5000 available samples for analysis.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 13/13 variants are retained for analysis</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Analysing SOD1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Retrieved genotypes for 4 variants</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 5000/25000 samples in the gdb are present in cohort 'phenosmall'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Analysing unit SOD1; varSet High</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Keeping 5000/5000 available samples for analysis.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2/2 variants are retained for analysis</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># CADD-weighted burden test</span></span></span>
<span class="r-in"><span><span class="va">varsetlist_cadd</span> <span class="op">&lt;-</span> <span class="fu"><a href="getVarSet.html">getVarSet</a></span><span class="op">(</span><span class="va">varsetfile</span>, unit <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SOD1"</span>, <span class="st">"NEK1"</span><span class="op">)</span>, varSetName <span class="op">=</span> <span class="st">"CADD"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">rvb</span> <span class="op">&lt;-</span> <span class="fu"><a href="assocTest.html">assocTest</a></span><span class="op">(</span><span class="va">gdb</span>,</span></span>
<span class="r-in"><span>                 varSet <span class="op">=</span> <span class="va">varsetlist_cadd</span>,</span></span>
<span class="r-in"><span>                 cohort <span class="op">=</span> <span class="st">"phenosmall"</span>,</span></span>
<span class="r-in"><span>                 pheno <span class="op">=</span> <span class="st">"pheno"</span>,</span></span>
<span class="r-in"><span>                 covar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PC1"</span>, <span class="st">"PC2"</span>, <span class="st">"PC3"</span>, <span class="st">"PC4"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                 test <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"firth"</span>, <span class="st">"skat_robust"</span>, <span class="st">"acatv"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Analysing NEK1</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>NAs introduced by coercion</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Retrieved genotypes for 190 variants</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 5000/25000 samples in the gdb are present in cohort 'phenosmall'</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>NAs introduced by coercion</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Analysing unit NEK1; varSet CADD</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Keeping 5000/5000 available samples for analysis.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 119/119 variants are retained for analysis</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 75/119 variants have zero carriers, these are dropped.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Analysing SOD1</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>NAs introduced by coercion</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Retrieved genotypes for 49 variants</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 5000/25000 samples in the gdb are present in cohort 'phenosmall'</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>NAs introduced by coercion</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Analysing unit SOD1; varSet CADD</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Keeping 5000/5000 available samples for analysis.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 38/38 variants are retained for analysis</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 26/38 variants have zero carriers, these are dropped.</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Perform recessive burden test</span></span></span>
<span class="r-in"><span><span class="va">rvb</span> <span class="op">&lt;-</span> <span class="fu"><a href="assocTest.html">assocTest</a></span><span class="op">(</span><span class="va">gdb</span>,</span></span>
<span class="r-in"><span>                 varSet <span class="op">=</span> <span class="va">varsetlist</span>,</span></span>
<span class="r-in"><span>                 cohort <span class="op">=</span> <span class="st">"phenosmall"</span>,</span></span>
<span class="r-in"><span>                 pheno <span class="op">=</span> <span class="st">"pheno"</span>,</span></span>
<span class="r-in"><span>                 covar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PC1"</span>, <span class="st">"PC2"</span>, <span class="st">"PC3"</span>, <span class="st">"PC4"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                 test <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"firth"</span>, <span class="st">"skat_robust"</span>, <span class="st">"acatv"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                 geneticModel <span class="op">=</span> <span class="st">"recessive"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Analysing NEK1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Retrieved genotypes for 33 variants</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 5000/25000 samples in the gdb are present in cohort 'phenosmall'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Analysing unit NEK1; varSet High</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Keeping 5000/5000 available samples for analysis.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 33/33 variants are retained for analysis</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 33/33 variants have zero carriers, these are dropped.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Less than two samples have a non-zero burden score, skipping tests.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Analysing SOD1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Retrieved genotypes for 4 variants</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 5000/25000 samples in the gdb are present in cohort 'phenosmall'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Analysing unit SOD1; varSet High</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Keeping 5000/5000 available samples for analysis.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 4/4 variants are retained for analysis</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 4/4 variants have zero carriers, these are dropped.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Less than two samples have a non-zero burden score, skipping tests.</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Resampled burden test </span></span></span>
<span class="r-in"><span><span class="va">rvb</span> <span class="op">&lt;-</span> <span class="fu"><a href="assocTest.html">assocTest</a></span><span class="op">(</span><span class="va">gdb</span>,</span></span>
<span class="r-in"><span>                 varSet <span class="op">=</span> <span class="va">varsetlist</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,</span></span>
<span class="r-in"><span>                 cohort <span class="op">=</span> <span class="st">"phenosmall"</span>,</span></span>
<span class="r-in"><span>                 pheno <span class="op">=</span> <span class="st">"pheno"</span>,</span></span>
<span class="r-in"><span>                 covar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PC1"</span>, <span class="st">"PC2"</span>, <span class="st">"PC3"</span>, <span class="st">"PC4"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                 test <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"skat"</span>, <span class="st">"skat_burden"</span>, <span class="st">"acatv"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                 name <span class="op">=</span> <span class="st">"example"</span>,</span></span>
<span class="r-in"><span>                 methodResampling <span class="op">=</span> <span class="st">"permutation"</span>,</span></span>
<span class="r-in"><span>                 nResampling <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Analysing NEK1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Retrieved genotypes for 33 variants</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 5000/25000 samples in the gdb are present in cohort 'phenosmall'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Analysing unit NEK1; varSet High</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Keeping 5000/5000 available samples for analysis.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 33/33 variants are retained for analysis</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 20/33 variants have zero carriers, these are dropped.</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Kevin Kenna, Paul Hop.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer></div>

  


  

  </body></html>

