<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Perform association tests on binary or quantitative traits — assocTest • rvat</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Perform association tests on binary or quantitative traits — assocTest"><meta property="og:description" content="Main method for performing association tests on binary or quantitative traits.
Performs either aggregate tests or single variant tests, with a wide range of
available statistical methods (see 'Details' for more information).
Various parameters allow for implementation of for example recessive/dominant analyses,
MAF-weighted burden tests and permutation-based association tests.
Variant and sample filters (e.g. maximum MAF, minimum number of carriers) can be specified.
assocTest can be run either directly on a genoMatrix object or alternatively
on a gdb object, in which case assocTest can loop through multiple variant sets by providing
a varSetFile or varSetList as input.
Moreover, assocTest can be run on sets of genes ('gene set burden'),
for which gene burden scores have been generated using aggregate and stored in an aggregateFile.
For running assocTest on a genoMatrix object see: assocTest-genoMatrix for details
for running assocTest on a gdb  object see: assocTest-gdb for details
for running assocTest on a aggregateFile  object see: assocTest-aggregateFile for details"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rvat</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.3.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/basics.html">Getting started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/basics.html">Getting started</a>
    </li>
    <li>
      <a href="../articles/association_testing.html">Association testing</a>
    </li>
    <li>
      <a href="../articles/explore_visualize_results.html">Visualization</a>
    </li>
    <li>
      <a href="../articles/variant_annotation.html">Variant annotation</a>
    </li>
    <li>
      <a href="../articles/GSA.html">Gene set analysis</a>
    </li>
    <li>
      <a href="../articles/spatial_clustering.html">Spatial clustering</a>
    </li>
    <li>
      <a href="../articles/cmdline.html">Command-line tool</a>
    </li>
  </ul></li>
<li>
  <a href="../articles/changelog.html">Changelog</a>
</li>
<li>
  <a href="../articles/faq.html">F.A.Q.</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/KennaLab/rvat" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Perform association tests on binary or quantitative traits</h1>
    
    <div class="hidden name"><code>assocTest.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Main method for performing association tests on binary or quantitative traits.
Performs either aggregate tests or single variant tests, with a wide range of
available statistical methods (see 'Details' for more information).
Various parameters allow for implementation of for example recessive/dominant analyses,
MAF-weighted burden tests and permutation-based association tests.
Variant and sample filters (e.g. maximum MAF, minimum number of carriers) can be specified.
assocTest can be run either directly on a <code><a href="genoMatrix.html">genoMatrix</a></code> object or alternatively
on a <code><a href="gdb.html">gdb</a></code> object, in which case assocTest can loop through multiple variant sets by providing
a <code><a href="varSetFile.html">varSetFile</a></code> or <code><a href="varSetList.html">varSetList</a></code> as input.
Moreover, assocTest can be run on sets of genes ('gene set burden'),
for which gene burden scores have been generated using <code><a href="https://rdrr.io/r/stats/aggregate.html" class="external-link">aggregate</a></code> and stored in an <code><a href="aggregateFile.html">aggregateFile</a></code>.
For running assocTest on a <code><a href="genoMatrix.html">genoMatrix</a></code> object see: <code><a href="assocTest-genoMatrix.html">assocTest-genoMatrix</a></code> for details<br>
for running assocTest on a <code><a href="gdb.html">gdb</a></code>  object see: <code><a href="assocTest-gdb.html">assocTest-gdb</a></code> for details
for running assocTest on a <code><a href="aggregateFile.html">aggregateFile</a></code>  object see: <code><a href="assocTest-aggregateFile.html">assocTest-aggregateFile</a></code> for details</p>
    </div>


    <div id="value">
    <h2>Value</h2>
    

<p>The object returned is of class <code><a href="rvatResult.html">rvbResult</a></code> or <code><a href="rvatResult.html">singlevarResult</a></code>.
The columns are described below: (note that some columns are specific to either <code>rvbResult</code> or <code>singlevarResult</code></p>


<p>respectively)</p><ul><li><p><code>unit</code>:  (rvbResult) name of the unit tested (from <code>varSet</code> or from <code>genoMatrix</code> metadata)</p></li>
<li><p><code>VAR_id</code>:  (singlevarResult) VAR_id of the respective variant.</p></li>
<li><p><code>cohort</code>:  Name of the cohort</p></li>
<li><p><code>varSetName</code>:  varSetName included in the provided <code>varSet</code>, if provided, or from the <code>genoMatrix</code> metadata.</p></li>
<li><p><code>name</code>:  if specified, contains the <code>name</code> argument specified in assocTest.</p></li>
<li><p><code>pheno</code>:  Phenotype tested, as specified by the <code>pheno</code> argument.</p></li>
<li><p><code>covar</code>:  Covariates included, as specified by the <code>covar</code> argument.</p></li>
<li><p><code>geneticModel</code>:  genetic model used, as specified by the <code>geneticModel</code> argument.</p></li>
<li><p><code>MAFweight</code>: MAF-weighting applied, as specified by the <code>MAFweights</code> argument.</p></li>
<li><p><code>test</code>:  statistical test used, as specified by the <code>test</code> argument.</p></li>
<li><p><code>nvar</code>:  (rvbResult) number of variants included in the aggregate test.</p></li>
<li><p><code>caseMAC</code>:  (singlevarResult) Minor allele count among cases</p></li>
<li><p><code>ctrlMAC</code>:  (singlevarResult) Minor allele count among controls.</p></li>
<li><p><code>caseCarriers</code>:  (rvbResult) The number of cases that carry at least one variant among variants in the variant set.</p></li>
<li><p><code>ctrlCarriers</code>:  (rvbResult) The number of controls that carry at least one variant among variants in the variant set.</p></li>
<li><p><code>meanCaseScore</code>:  (rvbResult) Mean burden score among cases.</p></li>
<li><p><code>meanCtrlScore</code>:  (rvbResult) Mean burden score among controls.</p></li>
<li><p><code>caseN</code>:  Number of cases.</p></li>
<li><p><code>ctrlN</code>:  Number of controls.</p></li>
<li><p><code>caseCallRate</code>:  Variant call-rate in cases. For aggregate tests this is the average call-rate across variants.</p></li>
<li><p><code>ctrlCallRate</code>:  Variant call-rate in controls For aggregate tests this is the average call-rate across variants.</p></li>
<li><p><code>effectAllele</code>:  (singlevarResult) Allele to which the effect estimate refers.</p></li>
<li><p><code>otherAllele</code>:  (singlevarResult) Non-effect allele</p></li>
<li><p><code>effect</code>:  Effect estimate for specified statistical test.
Note that SKAT tests, ACAT-v tests, SPA tests and negative binomial tests don't yield effect estimates.</p></li>
<li><p><code>effectSE</code>:  Standard error of effect estimate of specified statistical test.
Note that SKAT tests, ACAT-v tests, SPA tests and negative binomial tests don't yield effect estimates.</p></li>
<li><p><code>effectCIlower</code>:  Lower confidence interval of effect estimate of specified statistical test.
Note that SKAT tests, ACAT-v tests, SPA tests and negative binomial tests don't yield effect estimates.</p></li>
<li><p><code>effectCIupper</code>:  Upper confidence interval of effect estimate of specified statistical test.
Note that SKAT tests, ACAT-v tests, SPA tests and negative binomial tests don't yield effect estimates.</p></li>
<li><p><code>OR</code>:  Odds-ratio for <code>glm</code> and <code>firth</code> tests.</p></li>
<li><p><code>P</code>:  P-value for specified statistical test.</p></li>
</ul></div>
    <div id="details">
    <h2>Details</h2>
    <p><strong>Aggregate tests</strong>
For aggregate tests, the following tests are implemented:</p>
<p><em>binary/quantitative traits</em>:</p><ul><li><p><code>skat</code>: SKAT test as implemented in the <code>SKAT</code> R package (Wu <em>et. al</em>, 2011).</p></li>
<li><p><code>skat_burden</code>: Burden test as implemented in the <code>SKAT</code> R package (Wu <em>et. al</em>, 2011).</p></li>
<li><p><code>skato</code>: SKAT-O (Optimal Unified Test) as implemented in the <code>SKAT</code> R package (Wu <em>et. al</em>, 2011).</p></li>
<li><p><code>acatv</code>: ACAT-V test as implemented in the <code>ACAT</code> R package (Liu <em>et al.</em>, 2019)</p></li>
</ul><p><em>quantitative traits</em>:</p><ul><li><p><code>lm</code>: linear model</p></li>
</ul><p><em>binary traits</em>:</p><ul><li><p><code>firth</code>: firth logistic regression (Firth, 1993). The maximum number of iterations
can be specified using the <code>maxitFirth</code> parameter</p></li>
<li><p><code>glm</code>: logistic regression</p></li>
<li><p><code>nbinom</code>: Negative binomial test</p></li>
<li><p><code>skat_robust</code>: robust SKAT test (robust in the presence of an unbalanced case/control ratio) as implemented in the <code>SKAT</code> R package (Zhao <em>et al.</em>, 2020).</p></li>
<li><p><code>skat_burden_robust</code>: robust burden test (robust in the presence of an unbalanced case/control ratio) as implemented in the <code>SKAT</code> R package (Zhao <em>et al.</em>, 2020).</p></li>
<li><p><code>skato_robust</code>: robust SKAT-O test (robust in the presence of an unbalanced case/control ratio) (Zhao <em>et al.</em>, 2020)</p></li>
<li><p><code>acatvSPA</code>: adjusted ACAT-v test, using a score test using saddlepoint approximation.</p></li>
<li><p><code>acatvfirth</code>: adjusted ACAT-v test, using Firth's logistic regression to perform single variant tests.</p></li>
</ul><p><strong>Single variant tests</strong></p>
<p>For single variants, the following tests are implemented:</p>
<p><em>quantitative traits</em>:</p><ul><li><p><code>lm</code>: linear model</p></li>
</ul><p><em>binary traits:</em></p><ul><li><p><code>scoreSPA</code>: score test with saddle point approximation.</p></li>
<li><p><code>firth</code>: firth logistic regression (Firth, 1993). The maximum number of iterations
can be specified using the <code>maxitFirth</code> parameter.</p></li>
<li><p><code>glm</code>: logistic regression</p></li>
</ul></div>
    <div id="references">
    <h2>References</h2>
    <p>Wu, M. C. et al. Rare-Variant Association Testing for Sequencing Data with the Sequence Kernel Association Test. The American Journal of Human Genetics 89, 82–93 (2011).</p>
<p>Liu, Y. et al. ACAT: A Fast and Powerful p Value Combination Method for Rare-Variant Analysis in Sequencing Studies. The American Journal of Human Genetics 104, 410–421 (2019).</p>
<p>Firth, D. Bias Reduction of Maximum Likelihood Estimates. Biometrika 80, 27–38 (1993).</p>
<p>Zhao, Z. et al. UK Biobank Whole-Exome Sequence Binary Phenome Analysis with Robust Region-Based Rare-Variant Test. The American Journal of Human Genetics 106, 3–12 (2020).</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Kevin Kenna, Paul Hop.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer></div>

  


  

  </body></html>

