<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Create a gdb file. — buildGdb • rvat</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Create a gdb file. — buildGdb"><meta property="og:description" content="Creates a new gdb file and returns a connection object of type gdb-class.
The gdb can be structured and populated using a provided vcf file.
If no input variant file is provided then only an empty gdb is created."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rvat</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.3.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/basics.html">Getting started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/basics.html">Getting started</a>
    </li>
    <li>
      <a href="../articles/association_testing.html">Association testing</a>
    </li>
    <li>
      <a href="../articles/explore_visualize_results.html">Visualization</a>
    </li>
    <li>
      <a href="../articles/variant_annotation.html">Variant annotation</a>
    </li>
    <li>
      <a href="../articles/GSA.html">Gene set analysis</a>
    </li>
    <li>
      <a href="../articles/spatial_clustering.html">Spatial clustering</a>
    </li>
    <li>
      <a href="../articles/cmdline.html">Command-line tool</a>
    </li>
  </ul></li>
<li>
  <a href="../articles/changelog.html">Changelog</a>
</li>
<li>
  <a href="../articles/faq.html">F.A.Q.</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/KennaLab/rvat" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Create a gdb file.</h1>
    
    <div class="hidden name"><code>buildGdb.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Creates a new <code><a href="gdb.html">gdb</a></code> file and returns a connection object of type gdb-class.
The gdb can be structured and populated using a provided vcf file.
If no input variant file is provided then only an empty gdb is created.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">buildGdb</span><span class="op">(</span></span>
<span>  <span class="va">vcf</span>,</span>
<span>  <span class="va">output</span>,</span>
<span>  skipIndexes <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  skipVarRanges <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  overWrite <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  genomeBuild <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  memlimit <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>vcf</dt>
<dd><p>Input vcf file used to structure and populate gdb. Warning this function makes the following of assumptions: 1) strict adherence to vcf format (GT subfield first element in genotype firelds), 2) multiallelic records have been split, 3) desired genotype QC has already been applied (DP,GQ filters), 4) GT values conform to the set 0/0,0/1,1/0,1/1,./.,0|0,0|1,1|0,1|1,.|.. Multiallelic parsing and genotype QC can be performed using vcftools and/or accompanying parser scripts included on the rvat github.</p></dd>


<dt>output</dt>
<dd><p>Path for output <code><a href="gdb.html">gdb</a></code> file</p></dd>


<dt>skipIndexes</dt>
<dd><p>Flag to skip generation of indexes for var and dosage table (VAR_id;CHROM, POS,REF,ALT). Typically only required if you plan to use gdbConcat to concatenate a series of separately generated gdb files before use</p></dd>


<dt>skipVarRanges</dt>
<dd><p>Flag to skip generation of ranged var table. Typically only required if you plan to use gdbConcat to concatenate a series of separately generated gdb files before use</p></dd>


<dt>overWrite</dt>
<dd><p>overwrite if <code>output</code> already exists? Defaults to <code>FALSE</code>, in which case an error is raised.</p></dd>


<dt>genomeBuild</dt>
<dd><p>Optional genome build to include in the gdb metadata. If specified, it will be used to set ploidies (diploid, XnonPAR, YnonPAR) if the genome build is implemented in RVAT (currently: GRCh37, hg19, GRCh38, hg38).</p></dd>


<dt>memlimit</dt>
<dd><p>Maximum number of vcf records to parse at a time, defaults to 1000.</p></dd>


<dt>verbose</dt>
<dd><p>Should the function be verbose? (TRUE/FALSE), defaults to <code>TRUE</code>.</p></dd>

</dl></div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">rvatData</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">vcfpath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rvatData/man/rvat_example.html" class="external-link">rvat_example</a></span><span class="op">(</span><span class="st">"rvatData.vcf.gz"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">gdbpath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># build a gdb from vcf. </span></span></span>
<span class="r-in"><span><span class="co"># the genomeBuild parameters stores the genome build in the gdb metadata</span></span></span>
<span class="r-in"><span><span class="co"># this will be used to assign ploidies on sex chromosomes (diploid, XnonPAR, YnonPAR)</span></span></span>
<span class="r-in"><span><span class="fu">buildGdb</span><span class="op">(</span></span></span>
<span class="r-in"><span>  vcf <span class="op">=</span> <span class="va">vcfpath</span>,</span></span>
<span class="r-in"><span>  output <span class="op">=</span> <span class="va">gdbpath</span>,</span></span>
<span class="r-in"><span>  genomeBuild <span class="op">=</span> <span class="st">"GRCh38"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2024-11-20 14:40:05	Creating gdb tables</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 25000 sample IDs detected</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2024-11-20 14:40:05	Parsing vcf records</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2024-11-20 14:40:32	Processing completed for 1000 records. Committing to db.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2024-11-20 14:40:52	Processing completed for 1802 records. Committing to db.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2024-11-20 14:40:52	Creating var table indexes</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2024-11-20 14:40:52	Creating SM table index</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2024-11-20 14:40:53	Creating dosage table index</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2024-11-20 14:40:53	Creating ranged var table</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2024-11-20 14:40:53	Complete</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rvat gdb object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  Path: /var/folders/cl/wvc0rvjx4vd5rzt2_fhpmfth0000gp/T//Rtmp6OlT01/file77054b68b4c </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># for large vcfs, the memlimit parameter can be lowered</span></span></span>
<span class="r-in"><span><span class="fu">buildGdb</span><span class="op">(</span></span></span>
<span class="r-in"><span>  vcf <span class="op">=</span> <span class="va">vcfpath</span>,</span></span>
<span class="r-in"><span>  output <span class="op">=</span> <span class="va">gdbpath</span>,</span></span>
<span class="r-in"><span>  genomeBuild <span class="op">=</span> <span class="st">"GRCh38"</span>,</span></span>
<span class="r-in"><span>  memlimit <span class="op">=</span> <span class="fl">100</span>,</span></span>
<span class="r-in"><span>  overWrite <span class="op">=</span> <span class="cn">TRUE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2024-11-20 14:40:53	Creating gdb tables</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 25000 sample IDs detected</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2024-11-20 14:40:53	Parsing vcf records</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2024-11-20 14:40:56	Processing completed for 100 records. Committing to db.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2024-11-20 14:40:58	Processing completed for 200 records. Committing to db.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2024-11-20 14:41:01	Processing completed for 300 records. Committing to db.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2024-11-20 14:41:03	Processing completed for 400 records. Committing to db.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2024-11-20 14:41:05	Processing completed for 500 records. Committing to db.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2024-11-20 14:41:08	Processing completed for 600 records. Committing to db.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2024-11-20 14:41:10	Processing completed for 700 records. Committing to db.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2024-11-20 14:41:12	Processing completed for 800 records. Committing to db.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2024-11-20 14:41:14	Processing completed for 900 records. Committing to db.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2024-11-20 14:41:17	Processing completed for 1000 records. Committing to db.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2024-11-20 14:41:19	Processing completed for 1100 records. Committing to db.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2024-11-20 14:41:21	Processing completed for 1200 records. Committing to db.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2024-11-20 14:41:23	Processing completed for 1300 records. Committing to db.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2024-11-20 14:41:26	Processing completed for 1400 records. Committing to db.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2024-11-20 14:41:28	Processing completed for 1500 records. Committing to db.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2024-11-20 14:41:30	Processing completed for 1600 records. Committing to db.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2024-11-20 14:41:32	Processing completed for 1700 records. Committing to db.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2024-11-20 14:41:35	Processing completed for 1800 records. Committing to db.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2024-11-20 14:41:35	Processing completed for 1802 records. Committing to db.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2024-11-20 14:41:35	Creating var table indexes</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2024-11-20 14:41:35	Creating SM table index</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2024-11-20 14:41:35	Creating dosage table index</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2024-11-20 14:41:35	Creating ranged var table</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2024-11-20 14:41:35	Complete</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rvat gdb object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  Path: /var/folders/cl/wvc0rvjx4vd5rzt2_fhpmfth0000gp/T//Rtmp6OlT01/file77054b68b4c </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># see ?gdb for more information on gdb-files, see ?concatGdb for concatenate gdb databases</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Kevin Kenna, Paul Hop.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer></div>

  


  

  </body></html>

