<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Connect and interact with a gdb — gdb • rvat</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Connect and interact with a gdb — gdb"><meta property="og:description" content='Compressed SQLite (".gdb") representation of genotype data and associated variant annotation and sample info tables
These allow for rapid and memory-efficient of loading sample genotype data en associated metadata within R.
The slots of the gdb class are inherited entirely from the RSQLite SQLiteConnection class.
A host of RVAT methods described here allow for convenient querying and manipulation of a gdb, for complex queries
users can also directly perform SQL queries on the gdb as exemplified in the examples and tutorials.'><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rvat</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.3.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/basics.html">Getting started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/basics.html">Getting started</a>
    </li>
    <li>
      <a href="../articles/association_testing.html">Association testing</a>
    </li>
    <li>
      <a href="../articles/explore_visualize_results.html">Visualization</a>
    </li>
    <li>
      <a href="../articles/variant_annotation.html">Variant annotation</a>
    </li>
    <li>
      <a href="../articles/GSA.html">Gene set analysis</a>
    </li>
    <li>
      <a href="../articles/spatial_clustering.html">Spatial clustering</a>
    </li>
    <li>
      <a href="../articles/cmdline.html">Command-line tool</a>
    </li>
  </ul></li>
<li>
  <a href="../articles/changelog.html">Changelog</a>
</li>
<li>
  <a href="../articles/faq.html">F.A.Q.</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/KennaLab/rvat" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Connect and interact with a gdb</h1>
    
    <div class="hidden name"><code>gdb.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Compressed SQLite (".gdb") representation of genotype data and associated variant annotation and sample info tables
These allow for rapid and memory-efficient of loading sample genotype data en associated metadata within R.
The slots of the gdb class are inherited entirely from the RSQLite SQLiteConnection class.
A host of RVAT methods described here allow for convenient querying and manipulation of a gdb, for complex queries
users can also directly perform SQL queries on the gdb as exemplified in the examples and tutorials.</p>
    </div>


    <div id="connect">
    <h2>Connect</h2>
    

<ul><li><p><code>gdb(x)</code>: Connect to a  gdb object  (created using <code><a href="buildGdb.html">buildGdb()</a></code>), where <code>x</code> is the filepath.</p></li>
</ul></div>
    <div id="build">
    <h2>Build</h2>
    

<p>In the following code snippets, x is a gdb object.</p><ul><li><p><code>buildGdb(x)</code>: build a gdb directly from a vcf, see <code><a href="buildGdb.html">buildGdb()</a></code> for details.</p></li>
</ul></div>
    <div id="getters">
    <h2>Getters</h2>
    

<p>In the following code snippets, x is a gdb object.</p><ul><li><p><code>getGT(x, ...)</code>: Retrieve genotype data from the gdb, returns a <code><a href="genoMatrix.html">genoMatrix()</a></code> object.
See <code><a href="getGT.html">getGT()</a></code> for details.</p></li>
<li><p><code>getAnno(x, ...)</code>: Get an annotation table from gdb. See the <code><a href="getAnno.html">getAnno()</a></code> documentation for details.</p></li>
<li><p><code>getCohort(x, ...)</code>: Get a cohort table from gdb. See the <code><a href="getCohort.html">getCohort()</a></code> documentation for details.</p></li>
<li><p><code>listCohort(x)</code>: List sample cohort tables that have been uploaded to the gdb.</p></li>
<li><p><code>listAnno(x)</code>: List variant info tables that have been uploaded to the gdb.</p></li>
<li><p><code>getGdbId(x)</code>, <code>getGdbPath(x)</code>, <code>getCreationDate(x)</code>, <code>getGenomeBuild(x)</code>: various methods to retrieve indicate metadata from gdb</p></li>
</ul></div>
    <div id="upload-and-delete-sample-and-variant-info-tables">
    <h2>Upload and delete sample and variant info tables</h2>
    

<p>In the following code snippets, x is a gdb object.</p><ul><li><p><code>uploadCohort(x, ...)</code>: Upload cohort data tables to gdb.
See the <code><a href="uploadCohort.html">uploadCohort()</a></code> documentation for details.</p></li>
<li><p><code>uploadAnno(x, ...)</code>: Upload variant annotation data into gdb.
See the <code><a href="uploadAnno.html">uploadAnno()</a></code> documentation for details</p></li>
<li><p><code>mapVariants(x, ...)</code>: Map variants in the gdb onto features provided in a bed,gff/gtf or ranges file.
See <code><a href="mapVariants.html">mapVariants()</a></code> for details.</p></li>
<li><p><code>dropTable(x, name)</code>: Drop table with the name specified in <code>name</code> from gdb and clear from annotation / cohort metadata tables.</p></li>
</ul></div>
    <div id="subsetting-amp-merging">
    <h2>Subsetting &amp; Merging</h2>
    

<ul><li><p><code>subSetGdb</code>: Create a new gdb that is a subset of the input gdb. See <code><a href="subsetGdb.html">subsetGdb()</a></code> for details.</p></li>
<li><p><code>concatGdb</code>: Function to concatenate gdb databases. See <code><a href="concatGdb.html">concatGdb()</a></code> for details.</p></li>
</ul></div>
    <div id="exports">
    <h2>Exports</h2>
    

<ul><li><p><code>writeVcf</code>: Convert a gdb to a vcf-file. See <code><a href="writeVcf.html">writeVcf()</a></code> for details.</p></li>
</ul></div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="getGT.html">getGT</a></code></p>
<p><code><a href="buildGdb.html">buildGdb</a></code></p>
<p><code><a href="getAnno.html">getAnno</a></code></p>
<p><code><a href="getCohort.html">getCohort</a></code></p>
<p><code><a href="uploadCohort.html">uploadCohort</a></code></p>
<p><code><a href="uploadAnno.html">uploadAnno</a></code></p>
<p><code><a href="mapVariants.html">mapVariants</a></code></p>
<p><code><a href="subsetGdb.html">subsetGdb</a></code></p>
<p><code><a href="concatGdb.html">concatGdb</a></code></p>
<p><code><a href="writeVcf.html">writeVcf</a></code></p></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">rvatData</span><span class="op">)</span></span></span>
<span class="r-in"><span> </span></span>
<span class="r-in"><span><span class="co"># build a gdb directly from a vcf file</span></span></span>
<span class="r-in"><span><span class="va">vcfpath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rvatData/man/rvat_example.html" class="external-link">rvat_example</a></span><span class="op">(</span><span class="st">"rvatData.vcf.gz"</span><span class="op">)</span> <span class="co"># example vcf from rvatData package</span></span></span>
<span class="r-in"><span><span class="va">gdbpath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># write gdb to temporary file</span></span></span>
<span class="r-in"><span><span class="va">gdb</span> <span class="op">&lt;-</span> <span class="fu"><a href="buildGdb.html">buildGdb</a></span><span class="op">(</span>vcf <span class="op">=</span> <span class="va">vcfpath</span>,</span></span>
<span class="r-in"><span>                output <span class="op">=</span> <span class="va">gdbpath</span>, </span></span>
<span class="r-in"><span>                genomeBuild <span class="op">=</span> <span class="st">"GRCh38"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2025-02-12 12:25:34	Creating gdb tables</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 25000 sample IDs detected</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2025-02-12 12:25:34	Parsing vcf records</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2025-02-12 12:26:04	Processing completed for 1000 records. Committing to db.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2025-02-12 12:26:26	Processing completed for 1802 records. Committing to db.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2025-02-12 12:26:26	Creating var table indexes</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2025-02-12 12:26:26	Creating SM table index</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2025-02-12 12:26:26	Creating dosage table index</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2025-02-12 12:26:26	Creating ranged var table</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2025-02-12 12:26:26	Complete</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># upload variant info</span></span></span>
<span class="r-in"><span><span class="va">varinfo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rvatData/man/rvat_example.html" class="external-link">rvat_example</a></span><span class="op">(</span><span class="st">"rvatData.varinfo"</span><span class="op">)</span> <span class="co"># example variant info from rvatData package</span></span></span>
<span class="r-in"><span><span class="fu"><a href="uploadAnno.html">uploadAnno</a></span><span class="op">(</span><span class="va">gdb</span>, name <span class="op">=</span> <span class="st">"varInfo"</span>, value <span class="op">=</span> <span class="va">varinfo</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading table 'varInfo' from '/Users/phop2/Library/R/x86_64/4.4/library/rvatData/extdata/rvatData.varinfo'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 17 fields detected (CHROM,POS,ID,REF,ALT,QUAL,FILTER,AC,AN,AF,gene_name,HighImpact,ModerateImpact,Synonymous,CADDphred,PolyPhen,SIFT)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 1</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># upload cohort info</span></span></span>
<span class="r-in"><span><span class="va">pheno</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rvatData/man/rvat_example.html" class="external-link">rvat_example</a></span><span class="op">(</span><span class="st">"rvatData.pheno"</span><span class="op">)</span> <span class="co"># example cohort info from rvatData package</span></span></span>
<span class="r-in"><span><span class="fu"><a href="uploadCohort.html">uploadCohort</a></span><span class="op">(</span><span class="va">gdb</span>, name <span class="op">=</span> <span class="st">"pheno"</span>, value <span class="op">=</span> <span class="va">pheno</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading cohort 'pheno' from '/Users/phop2/Library/R/x86_64/4.4/library/rvatData/extdata/rvatData.pheno'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 10 fields detected (IID,sex,pheno,pop,superPop,PC1,PC2,PC3,PC4,age)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 12865 males, 12135 females and 0 unknown gender</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Retained 25000 of 25000 uploaded samples that could be mapped to dosage matrix</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 1</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># list annotations and cohorts present in gdb</span></span></span>
<span class="r-in"><span><span class="fu">listAnno</span><span class="op">(</span><span class="va">gdb</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      name</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1 varInfo</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                                                         value</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1 /Users/phop2/Library/R/x86_64/4.4/library/rvatData/extdata/rvatData.varinfo</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                       date</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1 Wed Feb 12 12:26:26 2025</span>
<span class="r-in"><span><span class="fu">listCohort</span><span class="op">(</span><span class="va">gdb</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    name</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1 pheno</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                                                       value</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1 /Users/phop2/Library/R/x86_64/4.4/library/rvatData/extdata/rvatData.pheno</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                       date</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1 Wed Feb 12 12:26:26 2025</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># retrieve annotations for a genomic interval</span></span></span>
<span class="r-in"><span><span class="va">varinfo</span> <span class="op">&lt;-</span> <span class="fu"><a href="getAnno.html">getAnno</a></span><span class="op">(</span><span class="va">gdb</span>, </span></span>
<span class="r-in"><span>                   table <span class="op">=</span> <span class="st">"varinfo"</span>, </span></span>
<span class="r-in"><span>                   ranges <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>CHROM <span class="op">=</span> <span class="st">"chr1"</span>, start <span class="op">=</span> <span class="fl">11013847</span>, end <span class="op">=</span> <span class="fl">11016874</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">varinfo</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   VAR_id CHROM      POS          ID REF ALT     QUAL FILTER AC    AN</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1      1  chr1 11013912          NA   A   G   722.13     NA  1 45484</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2      2  chr1 11013928 rs755357622   C   T  2598.23     NA  1 49194</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3      3  chr1 11013936          NA   A   C  4135.36     NA  1 50000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4      4  chr1 11013936          NA   A   G  4135.36     NA  1 42636</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5      5  chr1 11013952          NA   C   G   517.13     NA  1 49574</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6      6  chr1 11016874  rs80356715   C   T 64910.10     NA 32 49766</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             AF gene_name HighImpact ModerateImpact Synonymous CADDphred</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1 0.0000219858    TARDBP          0              1          0      24.8</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2 0.0000203277    TARDBP          0              0          1         .</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 0.0000200000    TARDBP          0              1          0      24.1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4 0.0000234544    TARDBP          0              1          0      22.6</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5 0.0000201719    TARDBP          0              0          1         .</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6 0.0006430093    TARDBP          0              1          0      22.2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   PolyPhen SIFT</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1        P    D</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2        .    .</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3        B    T</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4        B    T</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5        .    .</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6        B    T</span>
<span class="r-in"><span><span class="co"># see ?getAnno for more details</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># retrieve cohort</span></span></span>
<span class="r-in"><span><span class="va">pheno</span> <span class="op">&lt;-</span> <span class="fu"><a href="getCohort.html">getCohort</a></span><span class="op">(</span><span class="va">gdb</span>, cohort <span class="op">=</span> <span class="st">"pheno"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">pheno</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    IID sex pheno pop superPop       PC1       PC2        PC3         PC4</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1 ALS1   2     1 PJL      SAS  22.58749  4.738025 38.9157564 -10.1667342</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2 ALS2   2     1 BEB      SAS  26.37370 -1.445903 31.9712989  -6.0222555</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 ALS3   1     1 TSI      EUR  30.75496 48.866843 -5.4092404  14.9474404</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4 ALS4   2     1 GWD      AFR -87.48035 -6.193628 -0.9504131  -1.2029932</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5 ALS5   1     1 MSL      AFR -87.79310 -5.323925 -1.0353300  -0.7403234</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6 ALS6   1     1 YRI      AFR -86.64071 -3.418417 -1.3964977   1.0561607</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        age</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1 66.01743</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2 66.84306</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 54.35749</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4 65.09777</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5 58.94664</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6 63.13635</span>
<span class="r-in"><span><span class="co"># see ?getCohort for more details</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># delete table</span></span></span>
<span class="r-in"><span><span class="fu"><a href="uploadAnno.html">uploadAnno</a></span><span class="op">(</span><span class="va">gdb</span>, name <span class="op">=</span> <span class="st">"varInfo2"</span>, value <span class="op">=</span> <span class="va">varinfo</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading table 'varInfo2' from interactive R session'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 18 fields detected (VAR_id,CHROM,POS,ID,REF,ALT,QUAL,FILTER,AC,AN,AF,gene_name,HighImpact,ModerateImpact,Synonymous,CADDphred,PolyPhen,SIFT)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 1</span>
<span class="r-in"><span><span class="fu"><a href="dropTable.html">dropTable</a></span><span class="op">(</span><span class="va">gdb</span>,  <span class="st">"varInfo2"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Table 'varInfo2' removed from gdb</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># retrieve gdb metadata</span></span></span>
<span class="r-in"><span><span class="fu">getGdbPath</span><span class="op">(</span><span class="va">gdb</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "/var/folders/cl/wvc0rvjx4vd5rzt2_fhpmfth0000gp/T//RtmpRC6myU/file1825cc33abe9"</span>
<span class="r-in"><span><span class="fu">getGenomeBuild</span><span class="op">(</span><span class="va">gdb</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "GRCh38"</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># retrieve genotypes</span></span></span>
<span class="r-in"><span><span class="co"># first extract variant ids that we want to retrieve the genotypes for</span></span></span>
<span class="r-in"><span><span class="co"># note that `where` here is an SQL compliant</span></span></span>
<span class="r-in"><span><span class="va">varinfo</span> <span class="op">&lt;-</span> <span class="fu"><a href="getAnno.html">getAnno</a></span><span class="op">(</span><span class="va">gdb</span>, </span></span>
<span class="r-in"><span>                   table <span class="op">=</span> <span class="st">"varinfo"</span>, </span></span>
<span class="r-in"><span>                   where <span class="op">=</span> <span class="st">"gene_name = 'SOD1' and ModerateImpact = 1"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># retrieve genotypes for variants extracted above </span></span></span>
<span class="r-in"><span><span class="va">GT</span> <span class="op">&lt;-</span> <span class="fu"><a href="getGT.html">getGT</a></span><span class="op">(</span><span class="va">gdb</span>, </span></span>
<span class="r-in"><span>            VAR_id <span class="op">=</span> <span class="va">varinfo</span><span class="op">$</span><span class="va">VAR_id</span>,</span></span>
<span class="r-in"><span>            cohort <span class="op">=</span> <span class="st">"pheno"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Retrieved genotypes for 38 variants</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># retrieve genotypes for a given genomic interval</span></span></span>
<span class="r-in"><span><span class="va">GT_fromranges</span> <span class="op">&lt;-</span> <span class="fu"><a href="getGT.html">getGT</a></span><span class="op">(</span></span></span>
<span class="r-in"><span> <span class="va">gdb</span>,</span></span>
<span class="r-in"><span> ranges <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>CHROM <span class="op">=</span> <span class="st">"chr21"</span>, start <span class="op">=</span> <span class="fl">31659666</span>, end <span class="op">=</span> <span class="fl">31668931</span><span class="op">)</span>,</span></span>
<span class="r-in"><span> cohort <span class="op">=</span> <span class="st">"pheno"</span>,</span></span>
<span class="r-in"><span> anno <span class="op">=</span> <span class="st">"varInfo"</span>,</span></span>
<span class="r-in"><span> annoFields <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"VAR_id"</span>, <span class="st">"CHROM"</span>, <span class="st">"POS"</span>, <span class="st">"REF"</span>, <span class="st">"ALT"</span>, <span class="st">"HighImpact"</span>, <span class="st">"ModerateImpact"</span>, <span class="st">"Synonymous"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Retrieved genotypes for 49 variants</span>
<span class="r-in"><span><span class="co"># Learn about the getGT method in ?getGT</span></span></span>
<span class="r-in"><span><span class="co"># Learn about the genoMatrix format in ?genoMatrix</span></span></span>
<span class="r-in"><span> </span></span>
<span class="r-in"><span><span class="co"># Learn more about the downstream methods available for the gdb class in the relevant help pages</span></span></span>
<span class="r-in"><span><span class="co"># e.g. ?mapVariants, ?subsetGdb, ?assocTest, ?writeVcf</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">close</a></span><span class="op">(</span><span class="va">gdb</span><span class="op">)</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Kevin Kenna, Paul Hop.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer></div>

  


  

  </body></html>

