<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>geneSetAssoc — geneSetAssoc • rvat</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="geneSetAssoc — geneSetAssoc"><meta property="og:description" content="Perform self-contained or competitive gene set analyses."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rvat</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.3.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/basics.html">Getting started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/basics.html">Getting started</a>
    </li>
    <li>
      <a href="../articles/association_testing.html">Association testing</a>
    </li>
    <li>
      <a href="../articles/explore_visualize_results.html">Visualization</a>
    </li>
    <li>
      <a href="../articles/variant_annotation.html">Variant annotation</a>
    </li>
    <li>
      <a href="../articles/GSA.html">Gene set analysis</a>
    </li>
    <li>
      <a href="../articles/spatial_clustering.html">Spatial clustering</a>
    </li>
    <li>
      <a href="../articles/cmdline.html">Command-line tool</a>
    </li>
  </ul></li>
<li>
  <a href="../articles/changelog.html">Changelog</a>
</li>
<li>
  <a href="../articles/faq.html">F.A.Q.</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/KennaLab/rvat" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>geneSetAssoc</h1>
    
    <div class="hidden name"><code>geneSetAssoc.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Perform self-contained or competitive gene set analyses.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">geneSetAssoc</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  geneSet <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  scoreMatrix <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  cormatrix <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  condition <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  covar <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  test <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"lm"</span>, <span class="st">"glm"</span>, <span class="st">"mlm"</span>, <span class="st">"fisher"</span>, <span class="st">"ttest"</span>, <span class="st">"ztest"</span>, <span class="st">"ACAT"</span><span class="op">)</span>,</span>
<span>  threshold <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  Zcutoffs <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  INT <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  scoreCutoffs <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  minSetSize <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  maxSetSize <span class="op">=</span> <span class="cn">Inf</span>,</span>
<span>  oneSided <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  memlimit <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  ID <span class="op">=</span> <span class="st">"unit"</span>,</span>
<span>  output <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>object</dt>
<dd><p>an <code><a href="rvatResult.html">rvbResult</a></code> object.</p></dd>


<dt>geneSet</dt>
<dd><p>a <code><a href="geneSetList.html">geneSetList</a></code> or <code><a href="geneSetFile.html">geneSetFile</a></code> object.</p></dd>


<dt>scoreMatrix</dt>
<dd><p>A matrix (rows = genes, columns = features) can be provided to perform enrichment
analyses on continuous values. These can be used to perform e.g. cell-type enrichment analyses.</p></dd>


<dt>cormatrix</dt>
<dd><p>a correlation matrix with row and column names corresponding to the units in the rvbResult.
Needs to be specified in order to run the 'mlm' (mixed linear model) test.
A burden score correlation matrix can be generated using the <code><a href="buildCorMatrix.html">buildCorMatrix</a></code> method.
The mixed mixed linear model is run using the GENESIS R package.</p></dd>


<dt>condition</dt>
<dd><p>Perform conditional analyses. Input can be a 1) a <code><a href="geneSetList.html">geneSetList</a></code> or <code><a href="geneSetFile.html">geneSetFile</a></code>, in which case
the genesets specified in the <code>geneSet</code> parameter will all be conditioned on the gene sets provided here.
2) a vector of gene set names present in <code>geneSet</code>, all genesets specified in the geneSetList/geneSetFile
will be conditioned on the genesets specified here.</p></dd>


<dt>covar</dt>
<dd><p>a vector of covariates</p></dd>


<dt>test</dt>
<dd><p>Vector of tests to perform.
Currently implemented tests are the competitive tests lm,mlm,fisher and self-contained tests ttest,ztest and ACAT.</p></dd>


<dt>threshold</dt>
<dd><p>A vector of thresholds for cutoff-based tests (fisher's exact test / glm).</p></dd>


<dt>Zcutoffs</dt>
<dd><p>A vector (length=2, minimum and maximum) of cutoffs to apply to the Z-scores.
Z scores below/above these cutoffs will be set equal to the cutoff.</p></dd>


<dt>INT</dt>
<dd><p>Apply inverse normal transformation to Z-scores? Defaults to <code>FALSE</code>.</p></dd>


<dt>scoreCutoffs</dt>
<dd><p>If <code>scoreMatrix</code> is specified, this parameter can be set to cap scores in the scorematrix.
It should be a vector of length 2 (minimum and maximum sd). Defaults to <code>NULL</code>, in which case no score cutoffs are applied.</p></dd>


<dt>minSetSize</dt>
<dd><p>Exclude genesets with size &lt; minSetSize</p></dd>


<dt>maxSetSize</dt>
<dd><p>Exclude genesets with size &gt; maxSetSize</p></dd>


<dt>oneSided</dt>
<dd><p>Calculate a one-sided P-value? Defaults to <code>TRUE</code>.</p></dd>


<dt>memlimit</dt>
<dd><p>Maximum number of genesets to process in one go.</p></dd>


<dt>ID</dt>
<dd><p>ID column in the rvbResult that corresponds with the IDs used in the geneSetList.
Defaults to 'unit'.</p></dd>


<dt>output</dt>
<dd><p>Optional: save results to specified path</p></dd>


<dt>verbose</dt>
<dd><p>Should the function be verbose? Defaults to <code>TRUE</code>.</p></dd>

</dl></div>
    <div id="references">
    <h2>References</h2>
    <p>Gogarten SM, Sofer T, Chen H, Yu C, Brody JA, Thornton TA, Rice KM, Conomos MP. Genetic association testing using the GENESIS R/Bioconductor package. Bioinformatics. 2019 Dec 15;35(24):5346-5348</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="buildGeneSet.html">buildGeneSet</a></code></p>
<p><code><a href="geneSetList.html">geneSetList</a></code></p>
<p><code><a href="geneSetFile.html">geneSetFile</a></code></p></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">rvatData</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">rvbresults</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="va">rvbresults</span><span class="op">[</span><span class="va">rvbresults</span><span class="op">$</span><span class="va">test</span> <span class="op">==</span> <span class="st">"firth"</span> <span class="op">&amp;</span> </span></span>
<span class="r-in"><span>                    <span class="va">rvbresults</span><span class="op">$</span><span class="va">varSetName</span> <span class="op">==</span> <span class="st">"ModerateImpact"</span>, <span class="op">]</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># example genesetlist used in examples below (see ?buildGeneSet on build geneSetLists/geneSetFiles)</span></span></span>
<span class="r-in"><span><span class="va">genesetlist</span> <span class="op">&lt;-</span> <span class="fu"><a href="buildGeneSet.html">buildGeneSet</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"geneset1"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SOD1"</span>, <span class="st">"NEK1"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>       <span class="st">"geneset2"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ABCA4"</span>, <span class="st">"SOD1"</span>, <span class="st">"NEK1"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>       <span class="st">"geneset3"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"FUS"</span>, <span class="st">"NEK1"</span><span class="op">)</span></span></span>
<span class="r-in"><span>       <span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Perform competitive gene set analysis using a linear model</span></span></span>
<span class="r-in"><span><span class="va">GSAresults</span> <span class="op">&lt;-</span> <span class="fu">geneSetAssoc</span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="va">res</span>,</span></span>
<span class="r-in"><span>  <span class="va">genesetlist</span>,</span></span>
<span class="r-in"><span>  covar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nvar"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  test <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"lm"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 1 Z-scores are +Inf, these are set to the maximum observed Z-score: 4.219.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 3 out of 3 sets are kept.</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Outlying gene association scores can be remedied by either setting Z-score cutoffs (i.e. all Z-scores exceeding these values will be set to the respective cutoff), </span></span></span>
<span class="r-in"><span><span class="co"># or inverse normal transforming the Z-scores:</span></span></span>
<span class="r-in"><span><span class="va">GSAresults</span> <span class="op">&lt;-</span> <span class="fu">geneSetAssoc</span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="va">res</span>,</span></span>
<span class="r-in"><span>  <span class="va">genesetlist</span>,</span></span>
<span class="r-in"><span>  covar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nvar"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  test <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"lm"</span><span class="op">)</span>, </span></span>
<span class="r-in"><span>  maxSetSize <span class="op">=</span> <span class="fl">500</span>,</span></span>
<span class="r-in"><span>  Zcutoffs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">4</span>, <span class="fl">4</span><span class="op">)</span> <span class="co"># lower and upper bounds</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 0 Z-scores &lt;-4 are set to -4</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 3 Z-scores &gt;4 are set to 4</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 3 out of 3 sets are kept.</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">GSAresults</span> <span class="op">&lt;-</span> <span class="fu">geneSetAssoc</span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="va">res</span>,</span></span>
<span class="r-in"><span>  <span class="va">genesetlist</span>,</span></span>
<span class="r-in"><span>  covar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nvar"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  test <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"lm"</span><span class="op">)</span>, </span></span>
<span class="r-in"><span>  maxSetSize <span class="op">=</span> <span class="fl">500</span>,</span></span>
<span class="r-in"><span>  INT <span class="op">=</span> <span class="cn">TRUE</span> <span class="co"># perform inverse normal transformation</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 3 out of 3 sets are kept.</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Conditional gene set analyses can be performed to test whether gene sets are associated independently with the phenotype of interest. </span></span></span>
<span class="r-in"><span><span class="co"># In the example below we test whether gene sets are independent of geneset1</span></span></span>
<span class="r-in"><span><span class="va">GSAresults</span> <span class="op">&lt;-</span> <span class="fu">geneSetAssoc</span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="va">res</span>,</span></span>
<span class="r-in"><span>  condition <span class="op">=</span> <span class="fu"><a href="getGeneSet.html">getGeneSet</a></span><span class="op">(</span><span class="va">genesetlist</span>, <span class="st">"geneset1"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  geneSet <span class="op">=</span> <span class="va">genesetlist</span>,</span></span>
<span class="r-in"><span>  covar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nvar"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  test <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"lm"</span><span class="op">)</span>, </span></span>
<span class="r-in"><span>  maxSetSize <span class="op">=</span> <span class="fl">500</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 1 Z-scores are +Inf, these are set to the maximum observed Z-score: 4.219.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 3 out of 3 sets are kept.</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># perform two-sided tests by setting `oneSided = FALSE`</span></span></span>
<span class="r-in"><span><span class="va">GSAresults</span> <span class="op">&lt;-</span> <span class="fu">geneSetAssoc</span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="va">res</span>,</span></span>
<span class="r-in"><span>  <span class="va">genesetlist</span>,</span></span>
<span class="r-in"><span>  covar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nvar"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  test <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"lm"</span><span class="op">)</span>, </span></span>
<span class="r-in"><span>  maxSetSize <span class="op">=</span> <span class="fl">500</span>,</span></span>
<span class="r-in"><span>  oneSided <span class="op">=</span> <span class="cn">TRUE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 1 Z-scores are +Inf, these are set to the maximum observed Z-score: 4.219.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 3 out of 3 sets are kept.</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Test whether the proportion of P-values below a specified threshold is greater than the proportion outside of it.</span></span></span>
<span class="r-in"><span><span class="co"># Using Fisher's exact test</span></span></span>
<span class="r-in"><span><span class="co"># The `threshold` parameter specifies the P-value cutoff to define significant genes:</span></span></span>
<span class="r-in"><span><span class="va">GSAresults</span> <span class="op">&lt;-</span> <span class="fu">geneSetAssoc</span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="va">res</span>,</span></span>
<span class="r-in"><span>  <span class="va">genesetlist</span>,</span></span>
<span class="r-in"><span>  test <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"fisher"</span><span class="op">)</span>, </span></span>
<span class="r-in"><span>  threshold <span class="op">=</span> <span class="fl">1e-4</span>,</span></span>
<span class="r-in"><span>  maxSetSize <span class="op">=</span> <span class="fl">500</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 1 Z-scores are +Inf, these are set to the maximum observed Z-score: 4.219.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 3 out of 3 sets are kept.</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Kevin Kenna, Paul Hop.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer></div>

  


  

  </body></html>

