<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>The genoMatrix class for storing genotype data and associated sample and variant info. — genoMatrix • rvat</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="The genoMatrix class for storing genotype data and associated sample and variant info. — genoMatrix"><meta property="og:description" content="The genoMatrix class is specifically designed to represent genotype data and associated sample and variant info.
It extends the Bioconductor SummarizedExperiment class to accommodate genotype data."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rvat</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.3.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/basics.html">Getting started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/basics.html">Getting started</a>
    </li>
    <li>
      <a href="../articles/association_testing.html">Association testing</a>
    </li>
    <li>
      <a href="../articles/explore_visualize_results.html">Visualization</a>
    </li>
    <li>
      <a href="../articles/variant_annotation.html">Variant annotation</a>
    </li>
    <li>
      <a href="../articles/GSA.html">Gene set analysis</a>
    </li>
    <li>
      <a href="../articles/spatial_clustering.html">Spatial clustering</a>
    </li>
    <li>
      <a href="../articles/cmdline.html">Command-line tool</a>
    </li>
  </ul></li>
<li>
  <a href="../articles/changelog.html">Changelog</a>
</li>
<li>
  <a href="../articles/faq.html">F.A.Q.</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/kkenna/rvat" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>The genoMatrix class for storing genotype data and associated sample and variant info.</h1>
    
    <div class="hidden name"><code>genoMatrix.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>The genoMatrix class is specifically designed to represent genotype data and associated sample and variant info.
It extends the Bioconductor <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html" class="external-link">SummarizedExperiment</a></code> class to accommodate genotype data.</p>
    </div>


    <div id="construct-a-genomatrix">
    <h2>Construct a genoMatrix</h2>
    

<p><code><a href="getGT.html">getGT()</a></code>: Construct a genoMatrix, see <code><a href="getGT.html">getGT()</a></code> for details.</p>
    </div>
    <div id="accessors">
    <h2>Accessors</h2>
    

<p>The accessors are fully described in <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">SummarizedExperiment::SummarizedExperiment</a></code>, and
include <code>assays(x)</code>, <code>rowData(x)</code>, <code>colData(x)</code>, <code>metadata(x)</code>, <code>dim(x)</code>, <code>nrow(x)</code>, <code>ncol(x)</code>, <code>colnames(x)</code> and <code>rownames(x)</code>.</p>
    </div>
    <div id="getters">
    <h2>Getters</h2>
    

<p>In the following code snippets, x is a genoMatrix object</p><ul><li><p><code>getAF(x)</code>: Returns alternate allele frequencies. These will be equal to the minor allele frequences after applying <code>flipToMinor()</code>.</p></li>
<li><p><code>getMAF(x)</code>: Returns variant minor allele frequencies.</p></li>
<li><p><code>getAC(x)</code>: Returns alternate allele counts.  These will be equal to the minor allele counts after applying <code>flipToMinor()</code>.</p></li>
<li><p><code>getMAC(x)</code>: Returns minor allele counts.</p></li>
<li><p><code>getCR(x, var=TRUE)</code>: Returns call-rates for variants or samples.
<code>var=TRUE</code> by default, set to <code>FALSE</code> to return sample call-rates.</p></li>
<li><p><code>getNCarriers(x)</code>: Returns number of carriers of the alternate allele for each variant.
Note that when geneticModel = 'recessive' it will return the number of homozygous carriers.</p></li>
<li><p><code>summariseGeno(x)</code>: Returns a per variant summary of genotype counts and hwe testing.</p></li>
<li><p><code>getCarriers(x, VAR_id=NULL,colDataFields=NULL,rowDataFields=NULL,groupBy=NULL)</code>:
Return sample IDs for carriers of each of the variants in the genoMatrix.
<code>VAR_id</code> can be specified to return for the specified subset of variants.
<code>colDataFields</code> and <code>rowDataFields</code> can be specified include additional fields from <code>colData(x)</code> or <code>rowData(x)</code> in the output.
The <code>groupBy</code> parameter can be set to calculate carrier frequency among groups (such as cohort or phenotype), multiple groups can be set.
The <code>aggregate</code> parameter can be set to TRUE to return mean burden scores among the groupings.</p></li>
</ul><p>note: Variant ploidy (diploid, XnonPAR, YnonPAR) are handled according to sample sex.
Samples for which sex are not provided are excluded during AF calculation at non-diploid variants.</p>
    </div>
    <div id="subsetting">
    <h2>Subsetting</h2>
    


<p>Subsetting a genoMatrix object is equivalent to subsetting a SummarizedExperiment object,
and is fully described in fully described in <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">SummarizedExperiment::SummarizedExperiment</a></code>.
Please see the example section for examples.</p>
    </div>
    <div id="recode">
    <h2>Recode</h2>
    

<p>In the following code snippets, x is a genoMatrix object</p><ul><li><p><code>flipToMinor(x)</code>: Function to flip genotype dosages such that all GT values represent minor allele counts.</p></li>
<li><p><code>recode(x, geneticModel,imputeMethod,weights,MAFweights)</code>: Returns a recoded genoMatrix object, genetic model, imputation, and weights can be recoded. See <code><a href="recode.html">recode()</a></code> for details.</p></li>
<li><p><code>updateGT(x, SM = NULL, anno = NULL)</code>: Safe replacement of the colData or rowData table within a genoMatrix with a new table.</p></li>
</ul></div>
    <div id="rare-variant-testing">
    <h2>Rare variant testing</h2>
    

<p>In the following code snippets, x is a genoMatrix object</p><ul><li><p><code><a href="https://rdrr.io/r/stats/aggregate.html" class="external-link">aggregate()</a></code>: Returns per sample aggregate dosage for a genoMatrix object.
The genoMatrix shouldn't contain missing values, use <code><a href="recode.html">recode()</a></code> to impute missing values.
Aggregation is dependent on <code>geneticModel</code>, <code>MAFweights</code> and <code>weights</code>, all can be set using the <code><a href="recode.html">recode()</a></code> method.
By default, an updated genoMatrix is returned with an <code>aggregate</code> field in <code>colData</code>. Set <code>returnGT</code> to <code>FALSE</code> to return a vector of aggregates.</p></li>
<li><p><code><a href="assocTest.html">assocTest()</a></code>: Perform aggregate (burden) and single variant assocation test. See <code><a href="assocTest.html">assocTest()</a></code> for details.</p></li>
</ul></div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="assocTest.html">assocTest</a></code></p>
<p><code><a href="recode.html">recode</a></code></p></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">rvatData</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">GT</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Basic operations -------------------------------------------------------------</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># retrieve rowData (i.e. variant info)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">GT</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DataFrame with 42 rows and 4 columns</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           ploidy         w         REF         ALT</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      &lt;character&gt; &lt;numeric&gt; &lt;character&gt; &lt;character&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1268     diploid         1           C           T</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1270     diploid         1           G           C</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1271     diploid         1        GCAT           G</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1273     diploid         1           A           G</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1275     diploid         1           A           T</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ...          ...       ...         ...         ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1312     diploid         1           G           C</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1313     diploid         1          TG           T</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1314     diploid         1           T           G</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1315     diploid         1           A           G</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1316     diploid         1           T           C</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># retrieve colData (i.e. sample info)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">GT</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DataFrame with 25000 rows and 10 columns</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                   IID       sex     pheno         pop    superPop       PC1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           &lt;character&gt; &lt;numeric&gt; &lt;integer&gt; &lt;character&gt; &lt;character&gt; &lt;numeric&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ALS1             ALS1         2         1         PJL         SAS   22.5875</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ALS2             ALS2         2         1         BEB         SAS   26.3737</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ALS3             ALS3         1         1         TSI         EUR   30.7550</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ALS4             ALS4         2         1         GWD         AFR  -87.4804</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ALS5             ALS5         1         1         MSL         AFR  -87.7931</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ...               ...       ...       ...         ...         ...       ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CTRL19996   CTRL19996         1         0         ACB         AFR  -82.9449</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CTRL19997   CTRL19997         1         0         FIN         EUR   28.4759</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CTRL19998   CTRL19998         1         0         ACB         AFR  -75.3378</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CTRL19999   CTRL19999         2         0         PEL         AMR   38.2611</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CTRL20000   CTRL20000         1         0         STU         SAS   25.1909</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                 PC2        PC3        PC4       age</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           &lt;numeric&gt;  &lt;numeric&gt;  &lt;numeric&gt; &lt;numeric&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ALS1        4.73803  38.915756 -10.166734   66.0174</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ALS2       -1.44590  31.971299  -6.022256   66.8431</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ALS3       48.86684  -5.409240  14.947440   54.3575</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ALS4       -6.19363  -0.950413  -1.202993   65.0978</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ALS5       -5.32393  -1.035330  -0.740323   58.9466</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ...             ...        ...        ...       ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CTRL19996  -6.99645  -0.145528   0.646073   52.7380</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CTRL19997  42.83731 -13.434231   8.139021   52.3501</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CTRL19998  -2.48278  -2.023737   0.389070   66.6893</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CTRL19999 -16.33764 -41.853517 -76.090642   60.5823</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CTRL20000   7.86327  38.511045 -13.762434   62.7233</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># retrieve sample IDs</span></span></span>
<span class="r-in"><span><span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">GT</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">samples</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "ALS1" "ALS2" "ALS3" "ALS4" "ALS5" "ALS6"</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># retrieve VAR ids</span></span></span>
<span class="r-in"><span><span class="va">vars</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">GT</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">vars</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "1268" "1270" "1271" "1273" "1275" "1276"</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># check dimensions</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">GT</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 42</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">GT</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 25000</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">GT</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1]    42 25000</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># A genoMatrix object can be subsetted similarly as a data.frame:</span></span></span>
<span class="r-in"><span><span class="va">GT</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rvat genoMatrix Object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> class: genoMatrix </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> dim: 5 5 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> metadata(11): gdb gdbId ... geneticModel imputeMethod</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> assays(1): GT</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rownames(5): 1268 1270 1271 1273 1275</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rowData names(4): ploidy w REF ALT</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colnames(5): ALS1 ALS2 ALS3 ALS4 ALS5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colData names(10): IID sex ... PC4 age</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Subset samples based on sample info</span></span></span>
<span class="r-in"><span><span class="va">GT</span><span class="op">[</span> ,<span class="va">GT</span><span class="op">$</span><span class="va">pheno</span> <span class="op">==</span> <span class="fl">1</span><span class="op">]</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rvat genoMatrix Object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> class: genoMatrix </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> dim: 42 5000 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> metadata(11): gdb gdbId ... geneticModel imputeMethod</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> assays(1): GT</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rownames(42): 1268 1270 ... 1315 1316</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rowData names(4): ploidy w REF ALT</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colnames(5000): ALS1 ALS2 ... ALS4999 ALS5000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colData names(10): IID sex ... PC4 age</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Subset first two variants</span></span></span>
<span class="r-in"><span><span class="va">GT</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span>,<span class="op">]</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rvat genoMatrix Object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> class: genoMatrix </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> dim: 2 25000 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> metadata(11): gdb gdbId ... geneticModel imputeMethod</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> assays(1): GT</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rownames(2): 1268 1270</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rowData names(4): ploidy w REF ALT</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colnames(25000): ALS1 ALS2 ... CTRL19999 CTRL20000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colData names(10): IID sex ... PC4 age</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Extract variant/sample summaries --------------------------------------------------</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># calculate allele frequencies, allele counts etc.</span></span></span>
<span class="r-in"><span><span class="va">af</span> <span class="op">&lt;-</span> <span class="fu">getAF</span><span class="op">(</span><span class="va">GT</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">maf</span> <span class="op">&lt;-</span> <span class="fu">getMAF</span><span class="op">(</span><span class="va">GT</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">ac</span> <span class="op">&lt;-</span> <span class="fu">getAC</span><span class="op">(</span><span class="va">GT</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">mac</span> <span class="op">&lt;-</span> <span class="fu">getMAC</span><span class="op">(</span><span class="va">GT</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">carriers</span> <span class="op">&lt;-</span> <span class="fu">getNCarriers</span><span class="op">(</span><span class="va">GT</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># generate call-rates</span></span></span>
<span class="r-in"><span><span class="va">var_cr</span> <span class="op">&lt;-</span> <span class="fu">getCR</span><span class="op">(</span><span class="va">GT</span><span class="op">)</span> <span class="co"># variant call-rates</span></span></span>
<span class="r-in"><span><span class="va">sample_cr</span> <span class="op">&lt;-</span> <span class="fu">getCR</span><span class="op">(</span><span class="va">GT</span>, var <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="co"># sample call-rates</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># generate variant summaries</span></span></span>
<span class="r-in"><span><span class="va">varsummary</span> <span class="op">&lt;-</span> <span class="fu"><a href="summariseGeno.html">summariseGeno</a></span><span class="op">(</span><span class="va">GT</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Recode genotypes  --------------------------------------------------</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># flip variants with AF &gt; 0.5 to the minor allele</span></span></span>
<span class="r-in"><span><span class="va">GT</span> <span class="op">&lt;-</span> <span class="fu">flipToMinor</span><span class="op">(</span><span class="va">GT</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># recode genotypes to domiant/recessive models</span></span></span>
<span class="r-in"><span><span class="fu"><a href="recode.html">recode</a></span><span class="op">(</span><span class="va">GT</span>, geneticModel <span class="op">=</span> <span class="st">"dominant"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rvat genoMatrix Object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> class: genoMatrix </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> dim: 42 25000 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> metadata(11): gdb gdbId ... geneticModel imputeMethod</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> assays(1): GT</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rownames(42): 1268 1270 ... 1315 1316</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rowData names(6): ploidy w ... effectAllele otherAllele</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colnames(25000): ALS1 ALS2 ... CTRL19999 CTRL20000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colData names(10): IID sex ... PC4 age</span>
<span class="r-in"><span><span class="fu"><a href="recode.html">recode</a></span><span class="op">(</span><span class="va">GT</span>, geneticModel <span class="op">=</span> <span class="st">"recessive"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rvat genoMatrix Object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> class: genoMatrix </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> dim: 42 25000 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> metadata(11): gdb gdbId ... geneticModel imputeMethod</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> assays(1): GT</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rownames(42): 1268 1270 ... 1315 1316</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rowData names(6): ploidy w ... effectAllele otherAllele</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colnames(25000): ALS1 ALS2 ... CTRL19999 CTRL20000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colData names(10): IID sex ... PC4 age</span>
<span class="r-in"><span><span class="co"># see ?recode for details</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># recode missing genotypes </span></span></span>
<span class="r-in"><span><span class="fu"><a href="recode.html">recode</a></span><span class="op">(</span><span class="va">GT</span>, imputeMethod <span class="op">=</span> <span class="st">"meanImpute"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rvat genoMatrix Object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> class: genoMatrix </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> dim: 42 25000 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> metadata(11): gdb gdbId ... geneticModel imputeMethod</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> assays(1): GT</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rownames(42): 1268 1270 ... 1315 1316</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rowData names(6): ploidy w ... effectAllele otherAllele</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colnames(25000): ALS1 ALS2 ... CTRL19999 CTRL20000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colData names(10): IID sex ... PC4 age</span>
<span class="r-in"><span><span class="fu"><a href="recode.html">recode</a></span><span class="op">(</span><span class="va">GT</span>, imputeMethod <span class="op">=</span> <span class="st">"missingToRef"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rvat genoMatrix Object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> class: genoMatrix </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> dim: 42 25000 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> metadata(11): gdb gdbId ... geneticModel imputeMethod</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> assays(1): GT</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rownames(42): 1268 1270 ... 1315 1316</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rowData names(6): ploidy w ... effectAllele otherAllele</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colnames(25000): ALS1 ALS2 ... CTRL19999 CTRL20000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colData names(10): IID sex ... PC4 age</span>
<span class="r-in"><span><span class="co"># see ?recode for details</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># generate aggregate (burden) scores</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># by default, scores will be added to the colData</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/stats/aggregate.html" class="external-link">aggregate</a></span><span class="op">(</span><span class="fu"><a href="recode.html">recode</a></span><span class="op">(</span><span class="va">GT</span>, imputeMethod <span class="op">=</span> <span class="st">"meanImpute"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rvat genoMatrix Object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> class: genoMatrix </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> dim: 42 25000 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> metadata(11): gdb gdbId ... geneticModel imputeMethod</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> assays(1): GT</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rownames(42): 1268 1270 ... 1315 1316</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rowData names(6): ploidy w ... effectAllele otherAllele</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colnames(25000): ALS1 ALS2 ... CTRL19999 CTRL20000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colData names(11): IID sex ... age aggregate</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># set `returnGT = FALSE` to return aggregates as a vector instead</span></span></span>
<span class="r-in"><span><span class="va">aggregate</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/aggregate.html" class="external-link">aggregate</a></span><span class="op">(</span><span class="fu"><a href="recode.html">recode</a></span><span class="op">(</span><span class="va">GT</span>, imputeMethod <span class="op">=</span> <span class="st">"meanImpute"</span><span class="op">)</span>, returnGT <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">aggregate</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         ALS1         ALS2         ALS3         ALS4         ALS5         ALS6 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4.267668e-05 0.000000e+00 1.248082e-04 0.000000e+00 8.505593e-05 9.297136e-05 </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Update cohort and variant info in genoMatrix --------------------------</span></span></span>
<span class="r-in"><span><span class="va">gdb</span> <span class="op">&lt;-</span> <span class="fu"><a href="gdb.html">gdb</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/rvatData/man/rvat_example.html" class="external-link">rvat_example</a></span><span class="op">(</span><span class="st">"rvatData.gdb"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">anno</span> <span class="op">&lt;-</span> <span class="fu"><a href="getAnno.html">getAnno</a></span><span class="op">(</span><span class="va">gdb</span>, <span class="st">"varInfo"</span>, fields <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"VAR_id"</span>, <span class="st">"CADDphred"</span>, <span class="st">"PolyPhen"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">updateGT</span><span class="op">(</span><span class="va">GT</span>, anno <span class="op">=</span> <span class="va">anno</span> <span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rvat genoMatrix Object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> class: genoMatrix </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> dim: 42 25000 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> metadata(11): gdb gdbId ... geneticModel imputeMethod</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> assays(1): GT</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rownames(42): 1268 1270 ... 1315 1316</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rowData names(4): ploidy w CADDphred PolyPhen</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colnames(25000): ALS1 ALS2 ... CTRL19999 CTRL20000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colData names(10): IID sex ... PC4 age</span>
<span class="r-in"><span><span class="va">pheno</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">GT</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">updateGT</span><span class="op">(</span><span class="va">GT</span>, SM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">GT</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rvat genoMatrix Object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> class: genoMatrix </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> dim: 42 25000 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> metadata(11): gdb gdbId ... geneticModel imputeMethod</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> assays(1): GT</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rownames(42): 1268 1270 ... 1315 1316</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rowData names(6): ploidy w ... effectAllele otherAllele</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colnames(25000): ALS1 ALS2 ... CTRL19999 CTRL20000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colData names(4): IID sex pheno pop</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Get variant carriers</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># retrieve a data.frame that lists the samples carrying each respective </span></span></span>
<span class="r-in"><span><span class="co"># variant in the genoMatrix. Additional variant and sample info can be included</span></span></span>
<span class="r-in"><span><span class="co"># using the `rowDataFields` and `colDataFields` respectively.</span></span></span>
<span class="r-in"><span><span class="va">carriers</span> <span class="op">&lt;-</span> <span class="fu">getCarriers</span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="va">GT</span>,</span></span>
<span class="r-in"><span>  rowDataFields <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"REF"</span>, <span class="st">"ALT"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  colDataFields <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"superPop"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">carriers</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   VAR_id     IID genotype superPop REF ALT</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1   1268 ALS1811        1      AFR   C   T</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2   1268 ALS2109        1      AFR   C   T</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3   1268 ALS2155        1      SAS   C   T</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4   1268 ALS3076        1      AFR   C   T</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5   1268 ALS4396        1      AFR   C   T</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6   1268 ALS4476        1      EAS   C   T</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Perform rare variant tests:</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># burden test (firth)</span></span></span>
<span class="r-in"><span><span class="va">rvb</span> <span class="op">&lt;-</span> <span class="fu"><a href="assocTest.html">assocTest</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="va">GT</span>,</span></span>
<span class="r-in"><span>  pheno <span class="op">=</span> <span class="st">"pheno"</span>,</span></span>
<span class="r-in"><span>  covar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sex"</span>, <span class="st">"PC1"</span>, <span class="st">"PC2"</span>, <span class="st">"PC3"</span>, <span class="st">"PC4"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  test <span class="op">=</span> <span class="st">"firth"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Keeping 25000/25000 available samples for analysis.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 42/42 variants are retained for analysis</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># single variant tests</span></span></span>
<span class="r-in"><span><span class="va">sv</span> <span class="op">&lt;-</span> <span class="fu"><a href="assocTest.html">assocTest</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="va">GT</span>,</span></span>
<span class="r-in"><span>  pheno <span class="op">=</span> <span class="st">"pheno"</span>,</span></span>
<span class="r-in"><span>  covar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sex"</span>, <span class="st">"PC1"</span>, <span class="st">"PC2"</span>, <span class="st">"PC3"</span>, <span class="st">"PC4"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  test <span class="op">=</span> <span class="st">"scoreSPA"</span>,</span></span>
<span class="r-in"><span>  singlevar <span class="op">=</span> <span class="cn">TRUE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Keeping 25000/25000 available samples for analysis.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 42/42 variants are retained for analysis</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>33/42 variants have less than 2 carriers, tests will be skipped for these variants.</span>
<span class="r-in"><span><span class="co"># see ?assocTest for details</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Kevin Kenna, Paul Hop.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer></div>

  


  

  </body></html>

