<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>The genoMatrix class for storing genotype data and associated sample and variant info. — genoMatrix • rvat</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="The genoMatrix class for storing genotype data and associated sample and variant info. — genoMatrix"><meta property="og:description" content="The genoMatrix class is specifically designed to represent genotype data and associated sample and variant info.
It extends the Bioconductor SummarizedExperiment class to accommodate genotype data."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rvat</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/basics.html">Basics</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/basics.html">Basics</a>
    </li>
    <li>
      <a href="../articles/association_testing.html">Association testing</a>
    </li>
    <li>
      <a href="../articles/explore_visualize_results.html">Visualization</a>
    </li>
    <li>
      <a href="../articles/cmdline.html">Command-line tool</a>
    </li>
  </ul></li>
<li>
  <a href="../articles/faq.html">F.A.Q.</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/kkenna/rvat" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>The genoMatrix class for storing genotype data and associated sample and variant info.</h1>
    
    <div class="hidden name"><code>genoMatrix.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>The genoMatrix class is specifically designed to represent genotype data and associated sample and variant info.
It extends the Bioconductor <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html" class="external-link">SummarizedExperiment</a></code> class to accommodate genotype data.</p>
    </div>


    <div id="construct-a-genomatrix">
    <h2>Construct a genoMatrix</h2>
    

<p><code><a href="getGT.html">getGT()</a></code>: Construct a genoMatrix, see <code><a href="getGT.html">getGT()</a></code> for details.</p>
    </div>
    <div id="accessors">
    <h2>Accessors</h2>
    

<p>The accessors are fully described in <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">SummarizedExperiment::SummarizedExperiment</a></code>, and
include <code>assays(x)</code>, <code>rowData(x)</code>, <code>colData(x)</code>, <code>metadata(x)</code>, <code>dim(x)</code>, <code>nrow(x)</code>, <code>ncol(x)</code>, <code>colnames(x)</code> and <code>rownames(x)</code>.</p>
    </div>
    <div id="getters">
    <h2>Getters</h2>
    

<p>In the following code snippets, x is a genoMatrix object</p><ul><li><p><code>getAF(x)</code>: Returns alternate allele frequencies. These will be equal to the minor allele frequences after applying <code>flipToMinor()</code>.</p></li>
<li><p><code>getMAF(x)</code>: Returns variant minor allele frequencies.</p></li>
<li><p><code>getAC(x)</code>: Returns alternate allele counts.  These will be equal to the minor allele counts after applying <code>flipToMinor()</code>.</p></li>
<li><p><code>getMAC(x)</code>: Returns minor allele counts.</p></li>
<li><p><code>getCR(x, var=TRUE)</code>: Returns call-rates for variants or samples.
<code>var=TRUE</code> by default, set to <code>FALSE</code> to return sample call-rates.</p></li>
<li><p><code>getNCarriers(x)</code>: Returns number of carriers of the alternate allele for each variant.
Note that when geneticModel = 'recessive' it will return the number of homozygous carriers.</p></li>
<li><p><code>summariseGeno(x)</code>: Returns a per variant summary of genotype counts and hwe testing.</p></li>
<li><p><code>getCarriers(x, VAR_id=NULL,colDataFields=NULL,rowDataFields=NULL,groupBy=NULL)</code>:
Return sample IDs for carriers of each of the variants in the genoMatrix.
<code>VAR_id</code> can be specified to return for the specified subset of variants.
<code>colDataFields</code> and <code>rowDataFields</code> can be specified include additional fields from <code>colData(x)</code> or <code>rowData(x)</code> in the output.
The <code>groupBy</code> parameter can be set to calculate carrier frequency among groups (such as cohort or phenotype), multiple groups can be set.
The <code>aggregate</code> parameter can be set to TRUE to return mean burden scores among the groupings.</p></li>
</ul><p>note: Variant ploidy (diploid, XnonPAR, YnonPAR) are handled according to sample sex.
Samples for which sex are not provided are excluded during AF calculation at non-diploid variants.</p>
    </div>
    <div id="subsetting">
    <h2>Subsetting</h2>
    


<p>Subsetting a genoMatrix object is equivalent to subsetting a SummarizedExperiment object,
and is fully described in fully described in <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">SummarizedExperiment::SummarizedExperiment</a></code>.
Please see the example section for examples.</p>
    </div>
    <div id="recode">
    <h2>Recode</h2>
    

<p>In the following code snippets, x is a genoMatrix object</p><ul><li><p><code>flipToMinor(x)</code>: Function to flip genotype dosages such that all GT values represent minor allele counts.</p></li>
<li><p><code>recode(x, geneticModel,imputeMethod,weights,MAFweights)</code>: Returns a recoded genoMatrix object, genetic model, imputation, and weights can be recoded. See <code><a href="recode.html">recode()</a></code> for details.</p></li>
<li><p><code>updateGT(x, SM = NULL, anno = NULL)</code>: Safe replacement of the colData or rowData table within a genoMatrix with a new table.</p></li>
</ul></div>
    <div id="rare-variant-testing">
    <h2>Rare variant testing</h2>
    

<p>In the following code snippets, x is a genoMatrix object</p><ul><li><p><code><a href="https://rdrr.io/pkg/S4Vectors/man/aggregate-methods.html" class="external-link">aggregate()</a></code>: Returns per sample aggregate dosage for a genoMatrix object.
The genoMatrix shouldn't contain missing values, use <code><a href="recode.html">recode()</a></code> to impute missing values.
Aggregation is dependent on <code>geneticModel</code>, <code>MAFweights</code> and <code>weights</code>, all can be set using the <code><a href="recode.html">recode()</a></code> method.
By default, an updated genoMatrix is returned with an <code>aggregate</code> field in <code>colData</code>. Set <code>returnGT</code> to <code>FALSE</code> to return a vector of aggregates.</p></li>
<li><p><code><a href="assocTest.html">assocTest()</a></code>: Perform aggregate (burden) and single variant assocation test. See <code><a href="assocTest.html">assocTest()</a></code> for details.</p></li>
</ul></div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="assocTest.html">assocTest</a></code></p>
<p><code><a href="recode.html">recode</a></code></p></div>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Kevin Kenna, Paul Hop.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

