<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Load genotypes from a gdb. — getGT • rvat</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Load genotypes from a gdb. — getGT"><meta property="og:description" content="Method to retrieve a genoMatrix for variants specified by varSet or a vector of VAR_ids,
and samples as specified by the cohort parameter.
The checkPloidy parameter can be set to GRCh37 (or hg19) or GRCh38 (or hg38) to
assign variant ploidy (diploid,XnonPAR,YnonPAR). If not specified, the genome build in the gdb will be used, if available (included in the genomeBuild parameter was set in buildGdb)."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rvat</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.3.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/basics.html">Getting started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/basics.html">Getting started</a>
    </li>
    <li>
      <a href="../articles/association_testing.html">Association testing</a>
    </li>
    <li>
      <a href="../articles/explore_visualize_results.html">Visualization</a>
    </li>
    <li>
      <a href="../articles/variant_annotation.html">Variant annotation</a>
    </li>
    <li>
      <a href="../articles/GSA.html">Gene set analysis</a>
    </li>
    <li>
      <a href="../articles/spatial_clustering.html">Spatial clustering</a>
    </li>
    <li>
      <a href="../articles/cmdline.html">Command-line tool</a>
    </li>
  </ul></li>
<li>
  <a href="../articles/changelog.html">Changelog</a>
</li>
<li>
  <a href="../articles/faq.html">F.A.Q.</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/KennaLab/rvat" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Load genotypes from a gdb.</h1>
    
    <div class="hidden name"><code>getGT.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Method to retrieve a <code><a href="genoMatrix.html">genoMatrix</a></code> for variants specified by <code>varSet</code> or a vector of VAR_ids,
and samples as specified by the <code>cohort</code> parameter.
The <code>checkPloidy</code> parameter can be set to <code>GRCh37</code> (or <code>hg19</code>) or <code>GRCh38</code> (or <code>hg38</code>) to
assign variant ploidy (diploid,XnonPAR,YnonPAR). If not specified, the genome build in the <code><a href="gdb.html">gdb</a></code> will be used, if available (included in the <code>genomeBuild</code> parameter was set in <code><a href="buildGdb.html">buildGdb</a></code>).</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">getGT</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  varSet <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  VAR_id <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ranges <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  cohort <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  anno <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  annoFields <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  includeVarInfo <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  checkPloidy <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  varSetName <span class="op">=</span> <span class="st">"unnamed"</span>,</span>
<span>  unit <span class="op">=</span> <span class="st">"unnamed"</span>,</span>
<span>  padding <span class="op">=</span> <span class="fl">250</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  strict <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>object</dt>
<dd><p>an object of class <code><a href="gdb.html">gdb</a></code></p></dd>


<dt>varSet</dt>
<dd><p><code><a href="varSet.html">varSet</a></code> object. If specified, the <code>VAR_id</code>, <code>varSetName</code> and <code>unit</code> parameters will be ignored.</p></dd>


<dt>VAR_id</dt>
<dd><p>Character vector containing target VAR_id.</p></dd>


<dt>ranges</dt>
<dd><p>Extract variants within specified ranges.
Ranges can be specified as a data.frame, including at least 'CHROM','start', and 'end' columns, or
can be a <code><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GenomicRanges::GRanges</a></code> object.</p></dd>


<dt>cohort</dt>
<dd><p>Optional use of cohort data previously uploaded to the gdb (see <code>uploadCohort</code>).
If a valid cohort name is provided, then the uploaded data for this cohort is used to filter and annotate the returned genoMatrix object.
If a dataframe is provided, then this is assumed to conform to the SM table constraints required for genoMatrix objects (see genoMatrix).</p></dd>


<dt>anno</dt>
<dd><p>Optional use of variant annotation data previously uplodated to the gdb (see <code>uploadAnno</code>).
If a valid annotation table name is provided, the variant annotations will be included in the <code>rowData</code> of the <code>genoMatrix</code>.
Note that currently only annotation tables that include one row per variant can be included.
The <code>annoFields</code> parameter can be used to retain only specified fields from the annotation table.</p></dd>


<dt>annoFields</dt>
<dd><p>A vector of field names to retain if the <code>anno</code> parameter is set.</p></dd>


<dt>includeVarInfo</dt>
<dd><p>Include variant info ('var' table from the gdb) in the <code>genoMatrix</code>? Defaults to <code>FALSE</code>.
Note that setting this parameter to <code>TRUE</code> will override the <code>anno</code>/<code>annoFields</code> parameters.</p></dd>


<dt>checkPloidy</dt>
<dd><p>Version of the human genome to use when assigning variant ploidy (diploid, XnonPAR, YnonPAR).
Accepted inputs are GRCh37, hg19, GRCh38, hg38.
If not specified, the genome build in the <code><a href="gdb.html">gdb</a></code> will be used, if available (included in the <code>genomeBuild</code> parameter was set in <code><a href="buildGdb.html">buildGdb</a></code>).
Otherwise, if the genome build is not included in the gdb metadata, and no value is provided, then all variants are assigned the default ploidy of "diploid"</p></dd>


<dt>varSetName</dt>
<dd><p>Optional name for the set of variants, for example: 'missense or 'LOF' (ignored if <code>varSet</code> is specified.)</p></dd>


<dt>unit</dt>
<dd><p>Optional 'unit' name, for example: 'SOD1' or 'ENSG00000142168' (ignored if <code>varSet</code> is specified.)</p></dd>


<dt>padding</dt>
<dd><p>Number of basepairs to extend the search region beyond the specified genomic ranges to capture variants where the reference allele (REF) overlaps the input ranges,
but the POS of the variant falls outside the ranges. This accounts for variants where the REF allele spans multiple base pairs.</p></dd>


<dt>verbose</dt>
<dd><p>Should the method be verbose? Defaults to <code>TRUE</code>.</p></dd>


<dt>strict</dt>
<dd><p>Should strict checks be performed? Defaults to <code>TRUE</code>. Strict tests currently includes
checking whether supplied varSetFile/varSetList/varSet was generated from the same gdb as specified in <code>object</code>.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A <code><a href="genoMatrix.html">genoMatrix</a></code> object.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">rvatData</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">gdb</span> <span class="op">&lt;-</span> <span class="fu"><a href="gdb.html">gdb</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/rvatData/man/rvat_example.html" class="external-link">rvat_example</a></span><span class="op">(</span><span class="st">"rvatData.gdb"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># retrieve genotypes of a set of variants based on their VAR_ids</span></span></span>
<span class="r-in"><span><span class="va">varinfo</span> <span class="op">&lt;-</span> <span class="fu"><a href="getAnno.html">getAnno</a></span><span class="op">(</span><span class="va">gdb</span>, table <span class="op">=</span> <span class="st">"varinfo"</span>, where <span class="op">=</span> <span class="st">"gene_name = 'SOD1' and ModerateImpact = 1"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">GT</span> <span class="op">&lt;-</span> <span class="fu">getGT</span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="va">gdb</span>,</span></span>
<span class="r-in"><span>  VAR_id <span class="op">=</span> <span class="va">varinfo</span><span class="op">$</span><span class="va">VAR_id</span>,</span></span>
<span class="r-in"><span>  cohort <span class="op">=</span> <span class="st">"pheno"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Retrieved genotypes for 38 variants</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># retrieve genotypes of a set of variants in a varSet</span></span></span>
<span class="r-in"><span><span class="va">varsetfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="varSetFile.html">varSetFile</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/rvatData/man/rvat_example.html" class="external-link">rvat_example</a></span><span class="op">(</span><span class="st">"rvatData_varsetfile.txt.gz"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">varset</span> <span class="op">&lt;-</span> <span class="fu"><a href="getVarSet.html">getVarSet</a></span><span class="op">(</span><span class="va">varsetfile</span>, unit <span class="op">=</span> <span class="st">"NEK1"</span>, varSetName <span class="op">=</span> <span class="st">"High"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">GT</span> <span class="op">&lt;-</span> <span class="fu">getGT</span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="va">gdb</span>,</span></span>
<span class="r-in"><span>  varSet <span class="op">=</span> <span class="va">varset</span>,</span></span>
<span class="r-in"><span>  cohort <span class="op">=</span> <span class="st">"pheno"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Retrieved genotypes for 33 variants</span>
<span class="r-in"><span><span class="co"># see ?varSetFile and ?getVarSet for more details</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># retrieve genotypes for a genomic interval</span></span></span>
<span class="r-in"><span><span class="va">GT</span> <span class="op">&lt;-</span> <span class="fu">getGT</span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="va">gdb</span>,</span></span>
<span class="r-in"><span>  ranges <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>CHROM <span class="op">=</span> <span class="st">"chr21"</span>, start <span class="op">=</span> <span class="fl">31659666</span>, end <span class="op">=</span> <span class="fl">31668931</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  cohort <span class="op">=</span> <span class="st">"pheno"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Retrieved genotypes for 49 variants</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># the `anno` parameter can be specified to include variant annotations in the rowData of the genoMatrix</span></span></span>
<span class="r-in"><span><span class="va">GT</span> <span class="op">&lt;-</span> <span class="fu">getGT</span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="va">gdb</span>,</span></span>
<span class="r-in"><span>  ranges <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>CHROM <span class="op">=</span> <span class="st">"chr21"</span>, start <span class="op">=</span> <span class="fl">31659666</span>, end <span class="op">=</span> <span class="fl">31668931</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  cohort <span class="op">=</span> <span class="st">"pheno"</span>,</span></span>
<span class="r-in"><span>  anno <span class="op">=</span> <span class="st">"varInfo"</span>,</span></span>
<span class="r-in"><span>  annoFields <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"VAR_id"</span>, <span class="st">"CHROM"</span>, <span class="st">"POS"</span>, <span class="st">"REF"</span>, <span class="st">"ALT"</span>, <span class="st">"HighImpact"</span>, <span class="st">"ModerateImpact"</span>, <span class="st">"Synonymous"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Retrieved genotypes for 49 variants</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">GT</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DataFrame with 6 rows and 9 columns</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           ploidy         w       CHROM         POS         REF         ALT</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      &lt;character&gt; &lt;numeric&gt; &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;character&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1268     diploid         1       chr21    31659783           C           T</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1269     diploid         1       chr21    31659796           G           A</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1270     diploid         1       chr21    31659799           G           C</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1271     diploid         1       chr21    31659819        GCAT           G</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1272     diploid         1       chr21    31659820           C           A</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1273     diploid         1       chr21    31659828           A           G</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       HighImpact ModerateImpact  Synonymous</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      &lt;character&gt;    &lt;character&gt; &lt;character&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1268           0              1           0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1269           0              0           1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1270           0              1           0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1271           0              1           0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1272           0              0           1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1273           0              1           0</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># the includeVarInfo parameter is a shorthand for include the "var" table</span></span></span>
<span class="r-in"><span><span class="va">GT</span> <span class="op">&lt;-</span> <span class="fu">getGT</span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="va">gdb</span>,</span></span>
<span class="r-in"><span>  ranges <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>CHROM <span class="op">=</span> <span class="st">"chr21"</span>, start <span class="op">=</span> <span class="fl">31659666</span>, end <span class="op">=</span> <span class="fl">31668931</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  cohort <span class="op">=</span> <span class="st">"pheno"</span>,</span></span>
<span class="r-in"><span>  includeVarInfo <span class="op">=</span> <span class="cn">TRUE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Retrieved genotypes for 49 variants</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">GT</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DataFrame with 6 rows and 11 columns</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           ploidy         w       CHROM       POS          ID         REF</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      &lt;character&gt; &lt;numeric&gt; &lt;character&gt; &lt;integer&gt; &lt;character&gt; &lt;character&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1268     diploid         1       chr21  31659783 rs121912442           C</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1269     diploid         1       chr21  31659796 rs772764888           G</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1270     diploid         1       chr21  31659799           .           G</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1271     diploid         1       chr21  31659819           .        GCAT</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1272     diploid         1       chr21  31659820 rs200454724           C</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1273     diploid         1       chr21  31659828 rs768029813           A</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>              ALT        QUAL      FILTER                   INFO      FORMAT</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      &lt;character&gt; &lt;character&gt; &lt;character&gt;            &lt;character&gt; &lt;character&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1268           T     15317.2           . AC=6;AN=47612;AF=0.0..          GT</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1269           A     14053.5           . AC=1;AN=45320;AF=2.2..          GT</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1270           C      732.18           . AC=1;AN=49466;AF=2.0..          GT</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1271           G      241.15           . AC=1;AN=44404;AF=2.2..          GT</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1272           A     1218.18           . AC=1;AN=48150;AF=2.0..          GT</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1273           G     15989.6           . AC=5;AN=49638;AF=0.0..          GT</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># The `checkPloidy` parameter can be set to the version of the human genome to use</span></span></span>
<span class="r-in"><span><span class="co"># to assign variant ploidy. (diploid, XnonPAR, YnonPAR). Accepted inputs are GRCh37, hg19, GRCh38, hg38.</span></span></span>
<span class="r-in"><span><span class="co"># We recommend, however, to set the genome build when building the gdb: the genome build will theb</span></span></span>
<span class="r-in"><span><span class="co"># be included in the gdb metadata and used automatically. see ?buildGdb for details</span></span></span>
<span class="r-in"><span><span class="va">varinfo</span> <span class="op">&lt;-</span> <span class="fu"><a href="getAnno.html">getAnno</a></span><span class="op">(</span><span class="va">gdb</span>, table <span class="op">=</span> <span class="st">"varinfo"</span>, where <span class="op">=</span> <span class="st">"gene_name = 'UBQLN2' and ModerateImpact = 1"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">GT</span> <span class="op">&lt;-</span> <span class="fu">getGT</span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="va">gdb</span>,</span></span>
<span class="r-in"><span>  VAR_id <span class="op">=</span> <span class="va">varinfo</span><span class="op">$</span><span class="va">VAR_id</span>,</span></span>
<span class="r-in"><span>  cohort <span class="op">=</span> <span class="st">"pheno"</span>,</span></span>
<span class="r-in"><span>  includeVarInfo <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>  checkPloidy <span class="op">=</span> <span class="st">"GRCh38"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Ploidy of non-pseudoautosomal regions of the sex chromosomes are being set based on build GRCh38</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Retrieved genotypes for 24 variants</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># see ?genoMatrix for more details on the genoMatric class.</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Kevin Kenna, Paul Hop.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer></div>

  


  

  </body></html>

