<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Load genotypes from a gdb. — getGT • rvat</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Load genotypes from a gdb. — getGT"><meta property="og:description" content="Method to retrieve a genoMatrix for variants specified by varSet or a vector of VAR_ids,
and samples as specified by the cohort parameter.
The checkPloidy parameter can be set to GRCh37 (or hg19) or GRCh38 (or hg38) to
assign variant ploidy (diploid,XnonPAR,YnonPAR)."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rvat</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/basics.html">Basics</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/basics.html">Basics</a>
    </li>
    <li>
      <a href="../articles/association_testing.html">Association testing</a>
    </li>
    <li>
      <a href="../articles/explore_visualize_results.html">Visualization</a>
    </li>
    <li>
      <a href="../articles/cmdline.html">Command-line tool</a>
    </li>
  </ul></li>
<li>
  <a href="../articles/faq.html">F.A.Q.</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/kkenna/rvat" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Load genotypes from a gdb.</h1>
    
    <div class="hidden name"><code>getGT.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Method to retrieve a <code><a href="genoMatrix.html">genoMatrix</a></code> for variants specified by <code>varSet</code> or a vector of VAR_ids,
and samples as specified by the <code>cohort</code> parameter.
The <code>checkPloidy</code> parameter can be set to <code>GRCh37</code> (or <code>hg19</code>) or <code>GRCh38</code> (or <code>hg38</code>) to
assign variant ploidy (diploid,XnonPAR,YnonPAR).</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">getGT</span><span class="op">(</span>
  <span class="va">object</span>,
  varSet <span class="op">=</span> <span class="cn">NULL</span>,
  VAR_id <span class="op">=</span> <span class="cn">NULL</span>,
  ranges <span class="op">=</span> <span class="cn">NULL</span>,
  cohort <span class="op">=</span> <span class="cn">NULL</span>,
  checkPloidy <span class="op">=</span> <span class="cn">NULL</span>,
  varSetName <span class="op">=</span> <span class="st">"unnamed"</span>,
  unit <span class="op">=</span> <span class="st">"unnamed"</span>,
  verbose <span class="op">=</span> <span class="cn">TRUE</span>
<span class="op">)</span>

<span class="co"># S4 method for gdb</span>
<span class="fu">getGT</span><span class="op">(</span>
  <span class="va">object</span>,
  varSet <span class="op">=</span> <span class="cn">NULL</span>,
  VAR_id <span class="op">=</span> <span class="cn">NULL</span>,
  ranges <span class="op">=</span> <span class="cn">NULL</span>,
  cohort <span class="op">=</span> <span class="cn">NULL</span>,
  checkPloidy <span class="op">=</span> <span class="cn">NULL</span>,
  varSetName <span class="op">=</span> <span class="st">"unnamed"</span>,
  unit <span class="op">=</span> <span class="st">"unnamed"</span>,
  verbose <span class="op">=</span> <span class="cn">TRUE</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>object</dt>
<dd><p>an object of class <code><a href="gdb.html">gdb</a></code></p></dd>


<dt>varSet</dt>
<dd><p><code><a href="varSet.html">varSet</a></code> object. If specified, the <code>VAR_id</code>, <code>varSetName</code> and <code>unit</code> parameters will be ignored.</p></dd>


<dt>VAR_id</dt>
<dd><p>Charaocter vector containing target VAR_id.</p></dd>


<dt>ranges</dt>
<dd><p>Extract variants within specified ranges.
Ranges can be specified as a data.frame, including at least 'CHROM','start', and 'end' columns, or
can be a <code><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GenomicRanges::GRanges</a></code> object.</p></dd>


<dt>cohort</dt>
<dd><p>Optional use of cohort data previously uploaded to the gdb (see uploadCohort).
If a valid cohort name is provided, then the uploaded data for this cohort is used to filter and annotate the returned genoMatrix object. If a dataframe is provided, then this is assumed to conform to the SM table constraints required for genoMatrix objects (see genoMatrix).</p></dd>


<dt>checkPloidy</dt>
<dd><p>Version of the human genome to use when assigning variant ploidy (diploid,XnonPAR,YnonPAR).
Accepted inputs are GRCh37, hg19, GRCh38, hg38.
If no value is provided then all variants are assigned the default ploidy of "diploid"</p></dd>


<dt>varSetName</dt>
<dd><p>Optional name for the set of variants, for example: 'missense or 'LOF' (ignored if <code>varSet</code> is specified.)</p></dd>


<dt>unit</dt>
<dd><p>Optional 'unit' name, for example: 'SOD1' or 'ENSG00000142168' (ignored if <code>varSet</code> is specified.)</p></dd>


<dt>verbose</dt>
<dd><p>Should the method be verbose? Defaults to <code>TRUE</code>.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A <code><a href="genoMatrix.html">genoMatrix</a></code> object.</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Kevin Kenna, Paul Hop.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

