<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ rvat</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rvat</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.3.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/basics.html">Getting started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/basics.html">Getting started</a>
    </li>
    <li>
      <a href="../articles/association_testing.html">Association testing</a>
    </li>
    <li>
      <a href="../articles/explore_visualize_results.html">Visualization</a>
    </li>
    <li>
      <a href="../articles/variant_annotation.html">Variant annotation</a>
    </li>
    <li>
      <a href="../articles/GSA.html">Gene set analysis</a>
    </li>
    <li>
      <a href="../articles/spatial_clustering.html">Spatial clustering</a>
    </li>
    <li>
      <a href="../articles/cmdline.html">Command-line tool</a>
    </li>
  </ul></li>
<li>
  <a href="../articles/changelog.html">Changelog</a>
</li>
<li>
  <a href="../articles/faq.html">F.A.Q.</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/KennaLab/rvat" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="all-functions">All functions <a href="#all-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="ACAT.html">ACAT()</a></code> </p>
        </td>
        <td><p>Combine P-values using the ACAT method.</p></td>
      </tr><tr><td>
          <p><code><a href="aggregate-gdb-method.html">aggregate(<i>&lt;gdb&gt;</i>)</a></code> </p>
        </td>
        <td><p>Aggregate genotypes into a single (burden) score for each individual</p></td>
      </tr><tr><td>
          <p><code><a href="aggregateFile.html">aggregateFile</a></code> <code><a href="aggregateFile.html">listSamples</a></code> <code><a href="aggregateFile.html">getUnit</a></code> <code><a href="aggregateFile.html">aggregateFile-class</a></code> <code><a href="aggregateFile.html">show,aggregateFile-method</a></code> <code><a href="aggregateFile.html">listUnits,aggregateFile-method</a></code> <code><a href="aggregateFile.html">listSamples,aggregateFile-method</a></code> <code><a href="aggregateFile.html">metadata,aggregateFile-method</a></code> <code><a href="aggregateFile.html">getGdbId,aggregateFile-method</a></code> <code><a href="aggregateFile.html">getRvatVersion,aggregateFile-method</a></code> <code><a href="aggregateFile.html">getUnit,aggregateFile-method</a></code> </p>
        </td>
        <td><p>Class to manage interactions with an aggregateFile</p></td>
      </tr><tr><td>
          <p><code><a href="aggregateFileList.html">aggregateFileList</a></code> <code><a href="aggregateFileList.html">aggregateFileList-class</a></code> <code><a href="aggregateFileList.html">show,aggregateFileList-method</a></code> <code><a href="aggregateFileList.html">listUnits,aggregateFileList-method</a></code> <code><a href="aggregateFileList.html">listSamples,aggregateFileList-method</a></code> <code><a href="aggregateFileList.html">length,aggregateFileList-method</a></code> <code><a href="aggregateFileList.html">metadata,aggregateFileList-method</a></code> </p>
        </td>
        <td><p>Class to facilitate merging aggregateFiles</p></td>
      </tr><tr><td>
          <p><code><a href="assocTest-aggregateFile.html">assocTest(<i>&lt;aggregateFile&gt;</i>)</a></code> </p>
        </td>
        <td><p>assocTest-aggregateFile</p></td>
      </tr><tr><td>
          <p><code><a href="assocTest-gdb.html">assocTest(<i>&lt;gdb&gt;</i>)</a></code> </p>
        </td>
        <td><p>assocTest-gdb</p></td>
      </tr><tr><td>
          <p><code><a href="assocTest-genoMatrix.html">assocTest(<i>&lt;genoMatrix&gt;</i>)</a></code> </p>
        </td>
        <td><p>assocTest-genoMatrix</p></td>
      </tr><tr><td>
          <p><code><a href="assocTest.html">assocTest</a></code> </p>
        </td>
        <td><p>Perform association tests on binary or quantitative traits</p></td>
      </tr><tr><td>
          <p><code><a href="buildCorMatrix.html">buildCorMatrix()</a></code> </p>
        </td>
        <td><p>buildCorMatrix</p></td>
      </tr><tr><td>
          <p><code><a href="buildGdb.html">buildGdb()</a></code> </p>
        </td>
        <td><p>Create a gdb file.</p></td>
      </tr><tr><td>
          <p><code><a href="buildGeneSet.html">buildGeneSet()</a></code> </p>
        </td>
        <td><p>Build a geneSetList or geneSetFile</p></td>
      </tr><tr><td>
          <p><code><a href="buildResamplingFile.html">buildResamplingFile()</a></code> </p>
        </td>
        <td><p>Build a resampling matrix</p></td>
      </tr><tr><td>
          <p><code><a href="buildVarSet-data.frame.html">buildVarSet(<i>&lt;data.frame&gt;</i>)</a></code> </p>
        </td>
        <td><p>buildVarSet-data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="buildVarSet-gdb.html">buildVarSet(<i>&lt;gdb&gt;</i>)</a></code> </p>
        </td>
        <td><p>buildVarSet-gdb</p></td>
      </tr><tr><td>
          <p><code><a href="buildVarSet.html">buildVarSet</a></code> </p>
        </td>
        <td><p>buildVarSet</p></td>
      </tr><tr><td>
          <p><code><a href="collapseAggregateFiles.html">collapseAggregateFiles()</a></code> </p>
        </td>
        <td><p>Collapse multiple aggregate files</p></td>
      </tr><tr><td>
          <p><code><a href="concatGdb.html">concatGdb()</a></code> </p>
        </td>
        <td><p>Concatenate gdb databases</p></td>
      </tr><tr><td>
          <p><code><a href="densityPlot.html">densityPlot()</a></code> </p>
        </td>
        <td><p>Density plot</p></td>
      </tr><tr><td>
          <p><code><a href="dropTable.html">dropTable()</a></code> </p>
        </td>
        <td><p>Drop table from gdb</p></td>
      </tr><tr><td>
          <p><code><a href="exposition_pipe.html">%$%</a></code> </p>
        </td>
        <td><p>Exposition pipe</p></td>
      </tr><tr><td>
          <p><code><a href="extractRanges.html">extractRanges()</a></code> </p>
        </td>
        <td><p>Extract variants from a gdb based on a set of ranges</p></td>
      </tr><tr><td>
          <p><code><a href="fitNullModelGSA.html">fitNullModelGSA()</a></code> </p>
        </td>
        <td><p>fitNullModelGSA</p></td>
      </tr><tr><td>
          <p><code><a href="forestplot.html">forestplot()</a></code> </p>
        </td>
        <td><p>forestplot Makes a forest plot of the effect size of all occurrences of the chosen gene set</p></td>
      </tr><tr><td>
          <p><code><a href="gdb.html">gdb</a></code> <code><a href="gdb.html">listAnno</a></code> <code><a href="gdb.html">listCohort</a></code> <code><a href="gdb.html">getRvatVersion</a></code> <code><a href="gdb.html">getGdbId</a></code> <code><a href="gdb.html">getGdbPath</a></code> <code><a href="gdb.html">getGenomeBuild</a></code> <code><a href="gdb.html">getCreationDate</a></code> <code><a href="gdb.html">gdb-class</a></code> <code><a href="gdb.html">show,gdb-method</a></code> <code><a href="gdb.html">close,gdb-method</a></code> <code><a href="gdb.html">listAnno,gdb-method</a></code> <code><a href="gdb.html">listCohort,gdb-method</a></code> <code><a href="gdb.html">getRvatVersion,gdb-method</a></code> <code><a href="gdb.html">getGdbId,gdb-method</a></code> <code><a href="gdb.html">getGdbPath,gdb-method</a></code> <code><a href="gdb.html">getCreationDate,gdb-method</a></code> <code><a href="gdb.html">getGenomeBuild,gdb-method</a></code> <code><a href="gdb.html">dropTable,gdb-method</a></code> </p>
        </td>
        <td><p>Connect and interact with a gdb</p></td>
      </tr><tr><td>
          <p><code><a href="geneSet.html">geneSet</a></code> <code><a href="geneSet.html">geneSet-class</a></code> <code><a href="geneSet.html">metadata,geneSet-method</a></code> <code><a href="geneSet.html">length,geneSet-method</a></code> <code><a href="geneSet.html">as.data.frame,geneSet-method</a></code> <code><a href="geneSet.html">listUnits,geneSet-method</a></code> <code><a href="geneSet.html">listWeights,geneSet-method</a></code> </p>
        </td>
        <td><p>Class that represents a geneSet</p></td>
      </tr><tr><td>
          <p><code><a href="geneSetAssoc.html">geneSetAssoc()</a></code> </p>
        </td>
        <td><p>geneSetAssoc</p></td>
      </tr><tr><td>
          <p><code><a href="geneSetFile.html">geneSetFile()</a></code> </p>
        </td>
        <td><p>Class to manage interactions with a geneSetFile</p></td>
      </tr><tr><td>
          <p><code><a href="geneSetList.html">geneSetList</a></code> <code><a href="geneSetList.html">listMetadata</a></code> <code><a href="geneSetList.html">dropUnits</a></code> <code><a href="geneSetList.html">remapIDs</a></code> <code><a href="geneSetList.html">geneSetList-class</a></code> <code><a href="geneSetList.html">names,geneSetList-method</a></code> <code><a href="geneSetList.html">length,geneSetList-method</a></code> <code><a href="geneSetList.html">lengths,geneSetList-method</a></code> <code><a href="geneSetList.html">show,geneSetList-method</a></code> <code><a href="geneSetList.html">metadata,geneSetList-method</a></code> <code><a href="geneSetList.html">listMetadata,geneSetList-method</a></code> <code><a href="geneSetList.html">listGeneSets,geneSetList-method</a></code> <code><a href="geneSetList.html">listUnits,geneSetList-method</a></code> <code><a href="geneSetList.html">[[,geneSetList,ANY,missing-method</a></code> <code><a href="geneSetList.html">[,geneSetList,ANY,ANY,ANY-method</a></code> <code><a href="geneSetList.html">sort,geneSetList-method</a></code> <code><a href="geneSetList.html">as.data.frame,geneSetList-method</a></code> <code><a href="geneSetList.html">as.list,geneSetList-method</a></code> <code><a href="geneSetList.html">dropUnits,geneSetList-method</a></code> <code><a href="geneSetList.html">remapIDs,geneSetList-method</a></code> <code><a href="geneSetList.html">write,geneSetList-method</a></code> </p>
        </td>
        <td><p>Class to manage multiple geneSets</p></td>
      </tr><tr><td>
          <p><code><a href="genoMatrix.html">genoMatrix</a></code> <code><a href="genoMatrix.html">getAF</a></code> <code><a href="genoMatrix.html">getMAF</a></code> <code><a href="genoMatrix.html">getAC</a></code> <code><a href="genoMatrix.html">getMAC</a></code> <code><a href="genoMatrix.html">getNCarriers</a></code> <code><a href="genoMatrix.html">getCR</a></code> <code><a href="genoMatrix.html">updateGT</a></code> <code><a href="genoMatrix.html">flipToMinor</a></code> <code><a href="genoMatrix.html">getCarriers</a></code> <code><a href="genoMatrix.html">genoMatrix-class</a></code> <code><a href="genoMatrix.html">show,genoMatrix-method</a></code> <code><a href="genoMatrix.html">getAF,genoMatrix-method</a></code> <code><a href="genoMatrix.html">getMAF,genoMatrix-method</a></code> <code><a href="genoMatrix.html">getAC,genoMatrix-method</a></code> <code><a href="genoMatrix.html">getMAC,genoMatrix-method</a></code> <code><a href="genoMatrix.html">getNCarriers,genoMatrix-method</a></code> <code><a href="genoMatrix.html">getCR,genoMatrix-method</a></code> <code><a href="genoMatrix.html">summariseGeno,genoMatrix-method</a></code> <code><a href="genoMatrix.html">[,genoMatrix,ANY,ANY,ANY-method</a></code> <code><a href="genoMatrix.html">updateGT,genoMatrix-method</a></code> <code><a href="genoMatrix.html">flipToMinor,genoMatrix-method</a></code> <code><a href="genoMatrix.html">recode,genoMatrix-method</a></code> <code><a href="genoMatrix.html">aggregate,genoMatrix-method</a></code> <code><a href="genoMatrix.html">getCarriers,genoMatrix-method</a></code> </p>
        </td>
        <td><p>The genoMatrix class for storing genotype data and associated sample and variant info.</p></td>
      </tr><tr><td>
          <p><code><a href="getAnno.html">getAnno()</a></code> </p>
        </td>
        <td><p>Get an annotation table from a gdb</p></td>
      </tr><tr><td>
          <p><code><a href="getCohort.html">getCohort()</a></code> </p>
        </td>
        <td><p>Get an cohort table from a gdb</p></td>
      </tr><tr><td>
          <p><code><a href="getGT.html">getGT()</a></code> </p>
        </td>
        <td><p>Load genotypes from a gdb.</p></td>
      </tr><tr><td>
          <p><code><a href="getGeneSet.html">getGeneSet()</a></code> </p>
        </td>
        <td><p>Get geneset(s) from a geneSetList/geneSetFile</p></td>
      </tr><tr><td>
          <p><code><a href="getRanges.html">getRanges()</a></code> </p>
        </td>
        <td><p>getRanges</p></td>
      </tr><tr><td>
          <p><code><a href="getVarSet.html">getVarSet()</a></code> </p>
        </td>
        <td><p>Retrieve varSet records from a varSetFile or varSetList</p></td>
      </tr><tr><td>
          <p><code><a href="gsaResult.html">gsaResult()</a></code> </p>
        </td>
        <td><p>Create a <code>gsaResult-class</code></p></td>
      </tr><tr><td>
          <p><code><a href="listGeneSets.html">listGeneSets()</a></code> </p>
        </td>
        <td><p>Return names of geneSets included in a geneSetFile or geneSetList</p></td>
      </tr><tr><td>
          <p><code><a href="listUnits.html">listUnits()</a></code> </p>
        </td>
        <td><p>Return the units included in an varSetFile/varSetList/aggregateFile</p></td>
      </tr><tr><td>
          <p><code><a href="listVarSets.html">listVarSets()</a></code> </p>
        </td>
        <td><p>Return names of varsets included in a varSetFile or varSetList</p></td>
      </tr><tr><td>
          <p><code><a href="listWeights.html">listWeights()</a></code> </p>
        </td>
        <td><p>Return names of weights included in a varSet or geneSet</p></td>
      </tr><tr><td>
          <p><code><a href="manhattan.html">manhattan()</a></code> </p>
        </td>
        <td><p>Generate manhattan plot</p></td>
      </tr><tr><td>
          <p><code><a href="mapToCDS.html">mapToCDS()</a></code> </p>
        </td>
        <td><p>mapToCDS</p></td>
      </tr><tr><td>
          <p><code><a href="mapVariants.html">mapVariants()</a></code> </p>
        </td>
        <td><p>mapVariants</p></td>
      </tr><tr><td>
          <p><code><a href="mergeAggregateFiles.html">mergeAggregateFiles()</a></code> </p>
        </td>
        <td><p>Merge multiple aggregate files</p></td>
      </tr><tr><td>
          <p><code><a href="mutationPlot.html">mutationPlot()</a></code> </p>
        </td>
        <td><p>Mutation Plot</p></td>
      </tr><tr><td>
          <p><code><a href="nullModelGSA-class.html">listUnits(<i>&lt;nullModelGSA&gt;</i>)</a></code> <code><a href="nullModelGSA-class.html">length(<i>&lt;nullModelGSA&gt;</i>)</a></code> <code><a href="nullModelGSA-class.html">show(<i>&lt;nullModelGSA&gt;</i>)</a></code> </p>
        </td>
        <td><p>An S4 class to store and handle null models for mixed linear model gene set analysis.</p></td>
      </tr><tr><td>
          <p><code><a href="pipe.html">%&gt;%</a></code> </p>
        </td>
        <td><p>Pipe</p></td>
      </tr><tr><td>
          <p><code><a href="qqplot.html">qqplot()</a></code> </p>
        </td>
        <td><p>Generate a qq-plot</p></td>
      </tr><tr><td>
          <p><code><a href="readGMT.html">readGMT()</a></code> </p>
        </td>
        <td><p>Read a gmt-file</p></td>
      </tr><tr><td>
          <p><code><a href="readResults.html">readResults()</a></code> </p>
        </td>
        <td><p>Read association results</p></td>
      </tr><tr><td>
          <p><code><a href="recode.html">recode()</a></code> </p>
        </td>
        <td><p>Recode genotypes</p></td>
      </tr><tr><td>
          <p><code><a href="resamplingFile.html">resamplingFile</a></code> <code><a href="resamplingFile.html">resamplingFile-class</a></code> <code><a href="resamplingFile.html">show,resamplingFile-method</a></code> </p>
        </td>
        <td><p>Generate and manage permutations</p></td>
      </tr><tr><td>
          <p><code><a href="rvatResult.html">rvatResult</a></code> <code><a href="rvatResult.html">topResult</a></code> <code><a href="rvatResult.html">rvatResult-class</a></code> <code><a href="rvatResult.html">rvbResult-class</a></code> <code><a href="rvatResult.html">singlevarResult-class</a></code> <code><a href="rvatResult.html">gsaResult-class</a></code> <code><a href="rvatResult.html">writeResult,rvatResult-method</a></code> <code><a href="rvatResult.html">getGdbId,rvatResult-method</a></code> <code><a href="rvatResult.html">getGenomeBuild,rvatResult-method</a></code> <code><a href="rvatResult.html">merge,rvatResult,data.frame-method</a></code> <code><a href="rvatResult.html">merge,rvatResult,DataFrame-method</a></code> <code><a href="rvatResult.html">topResult,rvatResult-method</a></code> <code><a href="rvatResult.html">qqplot,rvatResult-method</a></code> <code><a href="rvatResult.html">manhattan,rvatResult-method</a></code> <code><a href="rvatResult.html">rvbResult</a></code> <code><a href="rvatResult.html">summary,rvbResult-method</a></code> <code><a href="rvatResult.html">ACAT,rvatResult-method</a></code> <code><a href="rvatResult.html">singlevarResult</a></code> <code><a href="rvatResult.html">summary,singlevarResult-method</a></code> <code><a href="rvatResult.html">summary,gsaResult-method</a></code> </p>
        </td>
        <td><p>Class for handling and visualizing results generated using rvat.</p></td>
      </tr><tr><td>
          <p><code><a href="rvatViewer.html">rvatViewer()</a></code> </p>
        </td>
        <td><p>rvatViewer</p></td>
      </tr><tr><td>
          <p><code><a href="spatialClust.html">spatialClust()</a></code> </p>
        </td>
        <td><p>Generate variant sets based on spatial clustering</p></td>
      </tr><tr><td>
          <p><code><a href="subsetGdb.html">subsetGdb()</a></code> </p>
        </td>
        <td><p>Generate subset of gdb, retaining all tables.</p></td>
      </tr><tr><td>
          <p><code><a href="summariseGeno.html">summariseGeno(<i>&lt;gdb&gt;</i>)</a></code> </p>
        </td>
        <td><p>Return variant summaries</p></td>
      </tr><tr><td>
          <p><code><a href="uploadAnno.html">uploadAnno()</a></code> </p>
        </td>
        <td><p>Upload variant annotation into gdb.</p></td>
      </tr><tr><td>
          <p><code><a href="uploadCohort.html">uploadCohort()</a></code> </p>
        </td>
        <td><p>Upload cohort table into gdb.</p></td>
      </tr><tr><td>
          <p><code><a href="varSet.html">varSet</a></code> <code><a href="varSet.html">listVars</a></code> <code><a href="varSet.html">varSet-class</a></code> <code><a href="varSet.html">mapToCDS,gdb-method</a></code> <code><a href="varSet.html">show,varSet-method</a></code> <code><a href="varSet.html">metadata,varSet-method</a></code> <code><a href="varSet.html">getGdbId,varSet-method</a></code> <code><a href="varSet.html">getRvatVersion,varSet-method</a></code> <code><a href="varSet.html">listVars,varSet-method</a></code> <code><a href="varSet.html">listWeights,varSet-method</a></code> <code><a href="varSet.html">as.data.frame,varSet-method</a></code> </p>
        </td>
        <td><p>Class that represents a set of variants and weights</p></td>
      </tr><tr><td>
          <p><code><a href="varSetFile.html">varSetFile</a></code> <code><a href="varSetFile.html">varSetFile-class</a></code> <code><a href="varSetFile.html">show,varSetFile-method</a></code> <code><a href="varSetFile.html">metadata,varSetFile-method</a></code> <code><a href="varSetFile.html">getGdbId,varSetFile-method</a></code> <code><a href="varSetFile.html">getRvatVersion,varSetFile-method</a></code> <code><a href="varSetFile.html">length,varSetFile-method</a></code> <code><a href="varSetFile.html">listUnits,varSetFile-method</a></code> <code><a href="varSetFile.html">listVarSets,varSetFile-method</a></code> <code><a href="varSetFile.html">getVarSet,varSetFile-method</a></code> </p>
        </td>
        <td><p>Class to manage interactions with a varSetFile</p></td>
      </tr><tr><td>
          <p><code><a href="varSetList.html">varSetList</a></code> <code><a href="varSetList.html">collapseVarSetList</a></code> <code><a href="varSetList.html">varSetList-class</a></code> <code><a href="varSetList.html">show,varSetList-method</a></code> <code><a href="varSetList.html">length,varSetList-method</a></code> <code><a href="varSetList.html">[[,varSetList,ANY,missing-method</a></code> <code><a href="varSetList.html">[,varSetList,ANY,ANY,ANY-method</a></code> <code><a href="varSetList.html">listUnits,varSetList-method</a></code> <code><a href="varSetList.html">listVarSets,varSetList-method</a></code> <code><a href="varSetList.html">getGdbId,varSetList-method</a></code> <code><a href="varSetList.html">getRvatVersion,varSetList-method</a></code> <code><a href="varSetList.html">getVarSet,varSetList-method</a></code> <code><a href="varSetList.html">write,varSetList-method</a></code> <code><a href="varSetList.html">as.data.frame,varSetList-method</a></code> <code><a href="varSetList.html">collapseVarSetList,varSetList-method</a></code> <code><a href="varSetList.html">metadata,varSetList-method</a></code> </p>
        </td>
        <td><p>Class to manage multiple varSets</p></td>
      </tr><tr><td>
          <p><code><a href="vcfInfo2Table.html">vcfInfo2Table()</a></code> </p>
        </td>
        <td><p>Convert vcf info field to table format</p></td>
      </tr><tr><td>
          <p><code><a href="writeResult.html">writeResult()</a></code> </p>
        </td>
        <td><p>Write an rvatResult to disk.</p></td>
      </tr><tr><td>
          <p><code><a href="writeVcf.html">writeVcf()</a></code> </p>
        </td>
        <td><p>writeVcf</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Kevin Kenna, Paul Hop.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer></div>

  


  

  </body></html>

