<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>mapVariants — mapVariants • rvat</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="mapVariants — mapVariants"><meta property="og:description" content="Method to map the variants in a gdb to a set of ranges or features.
The input can be a set of ranges (CHROM, start, end), a bed-file or a gff/gtf-file.
Variants in the gdb will be mapped onto those ranges and annotated with the features/columns
included in the input file.
For example, variants can be easily mapped upon genomic features downloaded in gff format from ensembl.
The output can be written to disk  (output parameter) or directly uploaded to the gdb (uploadName parameter)."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rvat</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.3.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/basics.html">Getting started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/basics.html">Getting started</a>
    </li>
    <li>
      <a href="../articles/association_testing.html">Association testing</a>
    </li>
    <li>
      <a href="../articles/explore_visualize_results.html">Visualization</a>
    </li>
    <li>
      <a href="../articles/variant_annotation.html">Variant annotation</a>
    </li>
    <li>
      <a href="../articles/GSA.html">Gene set analysis</a>
    </li>
    <li>
      <a href="../articles/spatial_clustering.html">Spatial clustering</a>
    </li>
    <li>
      <a href="../articles/cmdline.html">Command-line tool</a>
    </li>
  </ul></li>
<li>
  <a href="../articles/changelog.html">Changelog</a>
</li>
<li>
  <a href="../articles/faq.html">F.A.Q.</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/KennaLab/rvat" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>mapVariants</h1>
    
    <div class="hidden name"><code>mapVariants.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Method to map the variants in a <code><a href="gdb.html">gdb</a></code> to a set of ranges or features.
The input can be a set of ranges (CHROM, start, end), a bed-file or a gff/gtf-file.
Variants in the gdb will be mapped onto those ranges and annotated with the features/columns
included in the input file.
For example, variants can be easily mapped upon genomic features downloaded in gff format from ensembl.
The output can be written to disk  (<code>output</code> parameter) or directly uploaded to the <code><a href="gdb.html">gdb</a></code> (<code>uploadName</code> parameter).</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">mapVariants</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  ranges <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  gff <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  bed <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  bedCols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">character</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  fields <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  uploadName <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  output <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  sep <span class="op">=</span> <span class="st">"\t"</span>,</span>
<span>  skipIndexes <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  overWrite <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>object</dt>
<dd><p>a <code><a href="gdb.html">gdb</a></code> object</p></dd>


<dt>ranges</dt>
<dd><p>Can be 1) a data.frame, including at least 'CHROM','start', and 'end' columns.
2) a <code><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GenomicRanges::GRanges</a></code> object. 3) a filepath to a ranges file containing at least 'CHROM','start', and 'end' columns.
Separator can be specified using the <code>sep</code> parameter (defaults to <code>\\t</code>).</p></dd>


<dt>gff</dt>
<dd><p>Path to a gff- or gtf-file.</p></dd>


<dt>bed</dt>
<dd><p>Path to a bed-file. Specify extra columns using the <code>bedCols</code> parameter.</p></dd>


<dt>bedCols</dt>
<dd><p>A character vector of names of the extra columns to read from the BED-file.
Optionally the vector can be a named vector to indicate the classes of the columns (i.e. c("gene_id" = "character", "gene_name"="character")).
If not named, all extra columns will be read as character columns (see examples).</p></dd>


<dt>fields</dt>
<dd><p>Feature fields to keep. Defaults to <code>NULL</code> in which case all fields are kept.</p></dd>


<dt>uploadName</dt>
<dd><p>Name of table to upload to the gdb.
If not specified, either specifiy <code>output</code> to
write the results to disk, or otherwise the results will be returned in the R session.</p></dd>


<dt>output</dt>
<dd><p>Optionally, an output file path. Can be used instead of <code>uploadName</code> to write the results to disk.</p></dd>


<dt>sep</dt>
<dd><p>Field separator, relevant if <code>ranges</code> is a filepath. Defaults to <code>\\t</code>.</p></dd>


<dt>skipIndexes</dt>
<dd><p>Flag indicating whether to skip indexing of imported table.
Relevant if <code>uploadName</code> is specified, and thus the output table is imported in the gdb.
Defaults to <code>FALSE</code>.</p></dd>


<dt>overWrite</dt>
<dd><p>if <code>uploadName</code> is specified, should an existing table in the gdb with the same name be overwitten?
Defaults to <code>FALSE</code>.</p></dd>


<dt>verbose</dt>
<dd><p>Should the method be verbose? Defaults to <code>TRUE</code>.</p></dd>

</dl></div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">rvatData</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">rtracklayer</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicRanges" class="external-link">GenomicRanges</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">gdb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rvatData/man/create_example_gdb.html" class="external-link">create_example_gdb</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># map variants to gene models</span></span></span>
<span class="r-in"><span><span class="va">ranges</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  seqnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chr21"</span>, <span class="st">"chr4"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  ranges <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    start <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">31659666</span>, <span class="fl">169369704</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    end <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">31668931</span>, <span class="fl">169612632</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="op">)</span>,</span></span>
<span class="r-in"><span>  gene_name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SOD1"</span>, <span class="st">"NEK1"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">mapVariants</span><span class="op">(</span><span class="va">gdb</span>,</span></span>
<span class="r-in"><span>            ranges <span class="op">=</span> <span class="va">ranges</span>,</span></span>
<span class="r-in"><span>            uploadName <span class="op">=</span> <span class="st">"gene"</span>,</span></span>
<span class="r-in"><span>            verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># similarly, ranges can be a data.frame</span></span></span>
<span class="r-in"><span><span class="va">ranges</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  CHROM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chr21"</span>, <span class="st">"chr4"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  start <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">31659666</span>, <span class="fl">169369704</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  end <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">31668931</span>, <span class="fl">169612632</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  gene_name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SOD1"</span>, <span class="st">"NEK1"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">mapVariants</span><span class="op">(</span><span class="va">gdb</span>,</span></span>
<span class="r-in"><span>            ranges <span class="op">=</span> <span class="va">ranges</span>,</span></span>
<span class="r-in"><span>            uploadName <span class="op">=</span> <span class="st">"gene"</span>,</span></span>
<span class="r-in"><span>            verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span></span>
<span class="r-in"><span>            overWrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Table 'gene' already exists, it will be overwritten (as `overWrite=TRUE`)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Table 'gene' removed from gdb</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># often you'd want to map variants to a large set of ranges, such as ensembl models</span></span></span>
<span class="r-in"><span><span class="co"># mapVariants supports several file formats, including gff/gtf, bed and ranges</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># map variants using a gtf file</span></span></span>
<span class="r-in"><span><span class="va">gtffile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span>fileext <span class="op">=</span> <span class="st">".gtf"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">rtracklayer</span><span class="fu">::</span><span class="fu">export</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/makeGRangesFromDataFrame.html" class="external-link">makeGRangesFromDataFrame</a></span><span class="op">(</span><span class="va">ranges</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                    con <span class="op">=</span> <span class="va">gtffile</span>, </span></span>
<span class="r-in"><span>                    format <span class="op">=</span> <span class="st">"gtf"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">mapVariants</span><span class="op">(</span><span class="va">gdb</span>,</span></span>
<span class="r-in"><span>            gff <span class="op">=</span> <span class="va">gtffile</span>,</span></span>
<span class="r-in"><span>            uploadName <span class="op">=</span> <span class="st">"gene"</span>,</span></span>
<span class="r-in"><span>            verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span></span>
<span class="r-in"><span>            overWrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Table 'gene' already exists, it will be overwritten (as `overWrite=TRUE`)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Table 'gene' removed from gdb</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># map variants using a bed file</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">bedfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span>fileext <span class="op">=</span> <span class="st">".bed"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">rtracklayer</span><span class="fu">::</span><span class="fu">export</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/makeGRangesFromDataFrame.html" class="external-link">makeGRangesFromDataFrame</a></span><span class="op">(</span><span class="va">ranges</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                    con <span class="op">=</span> <span class="va">bedfile</span>, </span></span>
<span class="r-in"><span>                    format <span class="op">=</span> <span class="st">"bed"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">mapVariants</span><span class="op">(</span><span class="va">gdb</span>,</span></span>
<span class="r-in"><span>            bed <span class="op">=</span> <span class="va">bedfile</span>,</span></span>
<span class="r-in"><span>            uploadName <span class="op">=</span> <span class="st">"gene"</span>,</span></span>
<span class="r-in"><span>            verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span></span>
<span class="r-in"><span>            overWrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Table 'gene' already exists, it will be overwritten (as `overWrite=TRUE`)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Table 'gene' removed from gdb</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># see the variant annotation tutorial on the rvat website for more details</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Kevin Kenna, Paul Hop.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer></div>

  


  

  </body></html>

