<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Mutation Plot — mutationPlot • rvat</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Mutation Plot — mutationPlot"><meta property="og:description" content="Generates a mutation plot visualizing variant-level and gene-level association results along a transcript structure.
Optionally, custom tracks such as protein domains or mutation clusters can be overlaid."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rvat</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.3.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/basics.html">Getting started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/basics.html">Getting started</a>
    </li>
    <li>
      <a href="../articles/association_testing.html">Association testing</a>
    </li>
    <li>
      <a href="../articles/explore_visualize_results.html">Visualization</a>
    </li>
    <li>
      <a href="../articles/variant_annotation.html">Variant annotation</a>
    </li>
    <li>
      <a href="../articles/GSA.html">Gene set analysis</a>
    </li>
    <li>
      <a href="../articles/spatial_clustering.html">Spatial clustering</a>
    </li>
    <li>
      <a href="../articles/cmdline.html">Command-line tool</a>
    </li>
  </ul></li>
<li>
  <a href="../articles/changelog.html">Changelog</a>
</li>
<li>
  <a href="../articles/faq.html">F.A.Q.</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/KennaLab/rvat" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Mutation Plot</h1>
    
    <div class="hidden name"><code>mutationPlot.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Generates a mutation plot visualizing variant-level and gene-level association results along a transcript structure.
Optionally, custom tracks such as protein domains or mutation clusters can be overlaid.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">mutationPlot</span><span class="op">(</span></span>
<span>  <span class="va">singleVar</span>,</span>
<span>  <span class="va">cds</span>,</span>
<span>  customTracks <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  rvbGene <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  cdsGapSize <span class="op">=</span> <span class="fl">30</span>,</span>
<span>  cdsLimits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="fl">0.05</span><span class="op">)</span>, <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="fl">0.05</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  pointRange <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.25</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>  impactScale <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  trackSpacing <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  trackHeight <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  trackOrder <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  splitByTrackType <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  panelsizes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>  interactive <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  svHoverFields <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  trackHoverFields <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  cdsHoverFields <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>singleVar</dt>
<dd><p>A <code>singlevarResult</code> or data.frame containing single variant association results.
Required columns include <code>POS</code> (variant position, CDS coordinates), <code>P</code> (p-value), and <code>OR</code> (odds ratio).
An optional <code>impact</code> column can be included to represent variant impact, which will be mapped to different point shapes.</p></dd>


<dt>cds</dt>
<dd><p>A <code>GRanges</code> or <code>IRanges</code> object representing the coding sequence (CDS) regions of the transcript.</p></dd>


<dt>customTracks</dt>
<dd><p>An optional <code>rvbResult</code> or data.frame containing rare variant association statistics for custom tracks.
Should include <code>start</code> and <code>end</code> columns (CDS coordinates).
A <code>trackType</code> column is required if <code>splitByTrackType = TRUE</code>.
Additional columns can be included for hover information (see <code>trackHoverFields</code>).</p></dd>


<dt>rvbGene</dt>
<dd><p>An optional data.frame or <code>rvbResult</code> object containing gene-level association results.
Should contain a <code>P</code> column for the gene-level p-value.</p></dd>


<dt>cdsGapSize</dt>
<dd><p>The size of the gap to introduce between CDS exons, in base pairs. Defaults to 30.</p></dd>


<dt>cdsLimits</dt>
<dd><p>Specifies the y-axis limits for the CDS track (a vector of length 2).  Defaults to <code>c(-0.05, 0.05)</code> (nominal significance).</p></dd>


<dt>pointRange</dt>
<dd><p>Specifies the minimum and maximum size of the points representing variants (vector of length 2).
Point size corresponds to the absolute log of the odds ratio (<code>OR</code>). Defaults to <code>c(0.25, 3)</code>.</p></dd>


<dt>impactScale</dt>
<dd><p>Named character vector mapping impact levels to point shapes.
For example: <code>c("HIGH" = 24, "MODERATE" = 21)</code>. Only used if <code>singleVar</code> contains an <code>impact</code> column.</p></dd>


<dt>trackSpacing</dt>
<dd><p>Spacing between custom tracks. Defaults to 1.</p></dd>


<dt>trackHeight</dt>
<dd><p>Height of each custom track. Defaults to 1.</p></dd>


<dt>trackOrder</dt>
<dd><p>Optional vector of track names, specifying the order in which the track are plotted (if <code>splitByTrackType = TRUE</code>)</p></dd>


<dt>splitByTrackType</dt>
<dd><p>Should tracks be grouped by type (TRUE/FALSE)?
If <code>TRUE</code>, the <code>customTracks</code> data frame must contain a <code>trackType</code> column. Defaults to <code>FALSE</code>.</p></dd>


<dt>panelsizes</dt>
<dd><p>Relative sizes of the custom tracks vs. the other tracks. Defaults to <code>c(1,3)</code>.</p></dd>


<dt>interactive</dt>
<dd><p>Should the plot be interactive (using plotly)? Defaults to <code>FALSE</code>.</p></dd>


<dt>svHoverFields</dt>
<dd><p>If <code>interactive = TRUE</code>, an optional character vector that specifies which columns from <code>singleVar</code> to include in the hover information for single variant points.</p></dd>


<dt>trackHoverFields</dt>
<dd><p>If <code>interactive = TRUE</code>, an optional character vector that specifies which columns from <code>customTracks</code> to include in the hover information for custom tracks.</p></dd>


<dt>cdsHoverFields</dt>
<dd><p>If <code>interactive = TRUE</code>, an optional character vector that specifies which columns from the <code>cds</code> data frame to include in hover information for CDS regions.</p></dd>

</dl></div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Kevin Kenna, Paul Hop.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer></div>

  


  

  </body></html>

