<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Return variant summaries — summariseGeno • rvat</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Return variant summaries — summariseGeno"><meta property="og:description" content="Returns a per variant summary of genotype counts, frequencies, call-rates and hwe testing.
Note, the gdb implementation is described here, summariseGeno can also be run directly on a
genoMatrix object as described in the genoMatrix documentation."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rvat</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.3.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/basics.html">Getting started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/basics.html">Getting started</a>
    </li>
    <li>
      <a href="../articles/association_testing.html">Association testing</a>
    </li>
    <li>
      <a href="../articles/explore_visualize_results.html">Visualization</a>
    </li>
    <li>
      <a href="../articles/variant_annotation.html">Variant annotation</a>
    </li>
    <li>
      <a href="../articles/GSA.html">Gene set analysis</a>
    </li>
    <li>
      <a href="../articles/spatial_clustering.html">Spatial clustering</a>
    </li>
    <li>
      <a href="../articles/cmdline.html">Command-line tool</a>
    </li>
  </ul></li>
<li>
  <a href="../articles/changelog.html">Changelog</a>
</li>
<li>
  <a href="../articles/faq.html">F.A.Q.</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/KennaLab/rvat" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Return variant summaries</h1>
    
    <div class="hidden name"><code>summariseGeno.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Returns a per variant summary of genotype counts, frequencies, call-rates and hwe testing.
Note, the <code><a href="gdb.html">gdb</a></code> implementation is described here, <code>summariseGeno</code> can also be run directly on a
<code><a href="genoMatrix.html">genoMatrix</a></code> object as described in the <code><a href="genoMatrix.html">genoMatrix</a></code> documentation.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="co"># S4 method for gdb</span></span>
<span><span class="fu">summariseGeno</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  cohort <span class="op">=</span> <span class="st">"SM"</span>,</span>
<span>  varSet <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  VAR_id <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  pheno <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  memlimit <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  geneticModel <span class="op">=</span> <span class="st">"allelic"</span>,</span>
<span>  checkPloidy <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  keep <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  output <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  splitBy <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  minCallrateVar <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  maxCallrateVar <span class="op">=</span> <span class="cn">Inf</span>,</span>
<span>  minCallrateSM <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  maxCallrateSM <span class="op">=</span> <span class="cn">Inf</span>,</span>
<span>  minMAF <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  maxMAF <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  minMAC <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  maxMAC <span class="op">=</span> <span class="cn">Inf</span>,</span>
<span>  minCarriers <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  maxCarriers <span class="op">=</span> <span class="cn">Inf</span>,</span>
<span>  minCarrierFreq <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  maxCarrierFreq <span class="op">=</span> <span class="cn">Inf</span>,</span>
<span>  strict <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>object</dt>
<dd><p>a <code><a href="gdb.html">gdb</a></code> object</p></dd>


<dt>cohort</dt>
<dd><p>If a valid cohort name is provided, then the uploaded data for this cohort is used to filter and annotate the genotypes
If not specified, all samples in the gdb will be loaded.</p></dd>


<dt>varSet</dt>
<dd><p>a <code><a href="varSetList.html">varSetList</a></code> or <code><a href="varSetFile.html">varSetFile</a></code> object. Alternatively the VAR_id parameter can be specified.</p></dd>


<dt>VAR_id</dt>
<dd><p>A list of VAR_ids, alternatively the varSet parameter can be specified.
The <code>memlimit</code> argument controls how many variants to analyze at a time.</p></dd>


<dt>pheno</dt>
<dd><p>colData field to test as response variable, although not used within this method,
this can be useful to filter samples which have missing data for the response variable.</p></dd>


<dt>memlimit</dt>
<dd><p>Maximum number of variants to load at once (if <code>VAR_id</code> is specified).</p></dd>


<dt>geneticModel</dt>
<dd><p>Which genetic model to apply? ('allelic', 'recessive' or 'dominant').
Defaults to <code>allelic</code>.</p></dd>


<dt>checkPloidy</dt>
<dd><p>Version of the human genome to use when assigning variant ploidy (diploid, XnonPAR, YnonPAR).
Accepted inputs are GRCh37, hg19, GRCh38, hg38.
If not specified, the genome build in the <code><a href="gdb.html">gdb</a></code> will be used, if available (included if the <code>genomeBuild</code> parameter was set in <code><a href="buildGdb.html">buildGdb</a></code>).
Otherwise, if the genome build is not included in the gdb metadata, and no value is provided, then all variants are assigned the default ploidy of "diploid"</p></dd>


<dt>keep</dt>
<dd><p>vector of sample IDs to keep, defaults to <code>NULL</code>, in which case all samples are kept.</p></dd>


<dt>output</dt>
<dd><p>Output file path for results.
Defaults to <code>NULL</code>, in which case results are not written.</p></dd>


<dt>splitBy</dt>
<dd><p>Split variant summaries by labels indicated in the specified field.</p></dd>


<dt>minCallrateVar</dt>
<dd><p>Minimum genotype rate for variant retention.</p></dd>


<dt>maxCallrateVar</dt>
<dd><p>Maximum genotype rate for variant retention.</p></dd>


<dt>minCallrateSM</dt>
<dd><p>Minimum genotype rate for sample retention.</p></dd>


<dt>maxCallrateSM</dt>
<dd><p>Maximum genotype rate for sample retention.</p></dd>


<dt>minMAF</dt>
<dd><p>Minimum minor allele frequency for variant retention.</p></dd>


<dt>maxMAF</dt>
<dd><p>Maximum minor allele frequency for variant retention.</p></dd>


<dt>minMAC</dt>
<dd><p>Minimum minor allele count for variant retention.</p></dd>


<dt>maxMAC</dt>
<dd><p>Maximum minor allele count for variant retention.</p></dd>


<dt>minCarriers</dt>
<dd><p>Minimum carrier count for variant retention.</p></dd>


<dt>maxCarriers</dt>
<dd><p>Maximum carrier count for variant retention.</p></dd>


<dt>minCarrierFreq</dt>
<dd><p>Minimum carrier frequency for variant retention.</p></dd>


<dt>maxCarrierFreq</dt>
<dd><p>Maximum carrier frequency for variant retention.</p></dd>


<dt>strict</dt>
<dd><p>Should strict checks be performed? Defaults to <code>TRUE</code>.
Strict checks currently includes checking whether supplied varSetFile/varSetList was generated from the same gdb as specified in <code>object</code>.</p></dd>


<dt>verbose</dt>
<dd><p>Should the function be verbose? (TRUE/FALSE), defaults to <code>TRUE</code>.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>Returns a <code>data.frame</code> with the following columns:</p><ul><li><p><code>VAR_id</code>:  VAR_id of the respective variant.</p></li>
<li><p><code>AF</code>:  Allele frequency</p></li>
<li><p><code>callRate</code>:  callRate</p></li>
<li><p><code>geno0</code>:  Number of samples with genotype='0'.
When <code>geneticModel</code>='allelic' or 'dominant' this is the number of individuals that are
homozygous for the reference allele.</p></li>
<li><p><code>geno1</code>:  Number of samples with genotype='1'.
When <code>geneticModel</code>='allelic' this is the number of individuals that are
heterozygous for the reference allele.
When <code>geneticModel</code> = 'dominant' this represents the number of individuals who carry at least one
alternate allele.
When <code>geneticModel</code> = 'recessive' this represents the number of individuals who are homozygous
for the alternate allele.</p></li>
<li><p><code>geno2</code>: When <code>geneticModel</code> = 'allelic', the number of individuals who are homozygous
for the alternate allele.</p></li>
</ul></div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">rvatData</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">gdb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rvatData/man/create_example_gdb.html" class="external-link">create_example_gdb</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># generate for variant summaries for list of variants</span></span></span>
<span class="r-in"><span><span class="va">sumgeno</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">summariseGeno</span><span class="op">(</span><span class="va">gdb</span>,</span></span>
<span class="r-in"><span>              cohort <span class="op">=</span> <span class="st">"pheno"</span>,</span></span>
<span class="r-in"><span>              VAR_id <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">100</span>,</span></span>
<span class="r-in"><span>              output <span class="op">=</span> <span class="va">sumgeno</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Analysing chunk1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Retrieved genotypes for 100 variants</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Analysing unit chunk1; varSet none</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># generate for variant summaries for varSetFile</span></span></span>
<span class="r-in"><span><span class="va">varsetfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="varSetFile.html">varSetFile</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/rvatData/man/rvat_example.html" class="external-link">rvat_example</a></span><span class="op">(</span><span class="st">"rvatData_varsetfile.txt.gz"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">varsets</span> <span class="op">&lt;-</span> <span class="fu"><a href="getVarSet.html">getVarSet</a></span><span class="op">(</span><span class="va">varsetfile</span>, unit <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SOD1"</span>, <span class="st">"FUS"</span><span class="op">)</span>, varSetName <span class="op">=</span> <span class="st">"High"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">summariseGeno</span><span class="op">(</span><span class="va">gdb</span>,</span></span>
<span class="r-in"><span>              cohort <span class="op">=</span> <span class="st">"pheno"</span>,</span></span>
<span class="r-in"><span>              varSet <span class="op">=</span> <span class="va">varsets</span>,</span></span>
<span class="r-in"><span>              output <span class="op">=</span> <span class="va">sumgeno</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Analysing FUS</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Retrieved genotypes for 1 variants</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Analysing unit FUS; varSet High</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Analysing SOD1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Retrieved genotypes for 4 variants</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Analysing unit SOD1; varSet High</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># variant summaries can be generated for subgroups using the `splitBy` parameter.</span></span></span>
<span class="r-in"><span><span class="co"># this will result in an additional column in the output for the subgroups</span></span></span>
<span class="r-in"><span><span class="fu">summariseGeno</span><span class="op">(</span><span class="va">gdb</span>,</span></span>
<span class="r-in"><span>              cohort <span class="op">=</span> <span class="st">"pheno"</span>,</span></span>
<span class="r-in"><span>              VAR_id <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">100</span>,</span></span>
<span class="r-in"><span>              splitBy <span class="op">=</span> <span class="st">"pheno"</span>,</span></span>
<span class="r-in"><span>              output <span class="op">=</span> <span class="va">sumgeno</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Analysing chunk1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Retrieved genotypes for 100 variants</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Analysing unit chunk1; varSet none</span>
<span class="r-in"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="va">sumgeno</span>, header <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># contains 'pheno' column</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   VAR_id pheno           AF callRate geno0 geno1 geno2 hweP</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1      1     1 0.0000000000   0.9180  4590     0     0    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2      2     1 0.0000000000   0.9826  4913     0     0    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3      3     1 0.0000000000   1.0000  5000     0     0    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4      4     1 0.0000000000   0.8512  4256     0     0    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5      5     1 0.0001010305   0.9898  4948     1     0    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6      6     1 0.0004021717   0.9946  4969     4     0    1</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># summariseGeno can be ran directly on a genoMatrix</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">GT</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">sumgeno</span> <span class="op">&lt;-</span> <span class="fu">summariseGeno</span><span class="op">(</span><span class="va">GT</span><span class="op">)</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Kevin Kenna, Paul Hop.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer></div>

  


  

  </body></html>

